{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4d5ca2ef-0b8f-4f3e-a81b-9c3a0fc3dc07",
   "metadata": {},
   "source": [
    "## Refine groHMM transcription units overlapping RefSeq genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f49be998-3a2b-4fda-a12a-07787293ae71",
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths(\"/home/mahat/.conda/envs/r422/lib/R/library\")\n",
    ".libPaths()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a782e79-5ee9-41d0-8985-f53f82554741",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "    library(rtracklayer)\n",
    "    library(groHMM)\n",
    "    library(TxDb.Mmusculus.UCSC.mm10.knownGene)\n",
    "    library(org.Mm.eg.db)\n",
    "    library(plyranges)\n",
    "    library(tidyverse)\n",
    "});"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b6a44c7-34c7-4a7e-b170-ee8941775e45",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(\n",
    "    mc.cores=getCores(8),\n",
    "    repr.plot.width=3,\n",
    "    repr.plot.height=3,\n",
    "    jupyter.plot_mimetypes = \"image/svg+xml\"\n",
    ");\n",
    "theme_set(theme_classic());"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a07f0ddb-790e-42a6-b4ef-3a54fa26da86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# see http://genome.ucsc.edu/FAQ/FAQformat#format9\n",
    "header_names = c(\n",
    "    \"ID\", \"tx_name\", \"chr\", \"strand\", \"start\", \"end\",\n",
    "    \"cdsStart\", \"cdsEnd\", \"Exons\", \"xStarts\", \"xEnds\",\n",
    "    \"score\", \"name\", \"startStat\", \"endStat\", \"xFrames\"\n",
    ");\n",
    "#ref_transcripts = read_delim(\"http://s3.amazonaws.com/igv.org.genomes/mm10/ncbiRefSeq.sorted.txt.gz\",\n",
    "#ref_transcripts = read_delim(\"http://hgdownload.cse.ucsc.edu/goldenPath/mm10/database/refGene.txt.gz\",\n",
    "ref_transcripts = read_delim(\"/net/bmc-lab2/data/lab/scgroseq/group/genes_enhancer_list/ncbiRefSeq.sorted.txt.gz\",\n",
    "                            col_names = header_names, show_col_types=F,\n",
    "                            delim = \"\\t\") %>%\n",
    "    mutate(xStarts=NULL, xEnds=NULL, xFrames=NULL) %>%\n",
    "    mutate(name=paste0(\"GN-\", name)) %>%\n",
    "    filter(nchar(chr) <= 5) %>%\n",
    "    GRanges;\n",
    "names(ref_transcripts) = ref_transcripts$name;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4c65bc46-0be0-48e6-8d4b-85ef03070fa0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "length(ref_transcripts)\n",
    "\n",
    "ref_mrna = ref_transcripts %>%\n",
    "    filter( substr(tx_name, 0, 2) %in% c(\"NM\", \"XM\")  );\n",
    "\n",
    "ref_ncrna = ref_transcripts %>%\n",
    "    filter( !substr(tx_name, 0, 2) %in% c(\"NM\", \"XM\") );\n",
    "\n",
    "# remove ncRNA overlapping mRNA genes to avoid mapping issues\n",
    "ref_ncrna = subsetByOverlaps( ref_ncrna, ref_mrna, invert=T );\n",
    "ref_transcripts = c(ref_mrna, ref_ncrna);\n",
    "\n",
    "length(ref_transcripts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4a3ad17f-a9c5-404f-b761-c7c5e307ac8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ref_mrna = ref_mrna %>%\n",
    "    group_by(name) %>%\n",
    "    reduce_ranges_directed();\n",
    "names(ref_mrna) = ref_mrna$name;\n",
    "ref_mrna\n",
    "#saveRDS(ref_mrna, \"/net/bmc-lab2/data/lab/scgroseq/group/notebooks/data/UCSC_mm10_ncbiRefSeq_mRNA.rds\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f6c2260-8245-46eb-93fe-f4e3cfdd3385",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# discard empty chromosomes\n",
    "seqlevels(ref_transcripts) = seqlevelsInUse(ref_transcripts);\n",
    "\n",
    "# group genes by ID and reduce by strand-specific overlaps\n",
    "# this provides the full list of genes, regardless of thier dREG stauts\n",
    "# will be used to identify missed genes to add back to the glist\n",
    "rgConsensus = ref_mrna;\n",
    "names(rgConsensus) = rgConsensus$name;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c68cda64-2880-4de9-ac57-3abcd10fa624",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load dREG peak calls and convert to GRanges\n",
    "# this file is generated by calling dREG (Charles Danko Lab, Cornell) on bulk PROseq data, using the server: https://dreg.dnasequence.org/\n",
    "dREG = read.table(\"../data/PROseq_mES_BRsComb.dREG.peak.full.bed\", header=F, stringsAsFactors=F);\n",
    "colnames(dREG) = c(\"chr\", \"start\", \"end\", \"score\", \"pval\", \"center\");\n",
    "dREG = GRanges(dREG)\n",
    "dREG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ff0d4a0-d7f7-4b02-928c-483747b817e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# define promoter regions from refseq genes +- 500 bp\n",
    "refseq_promoters = promoters( ref_transcripts, upstream = 500, downstream = 500 );\n",
    "active_promoters = findOverlaps( refseq_promoters, dREG );\n",
    "\n",
    "# overlap with dREG to identify which promoters are active\n",
    "hasActivePro = queryHits(active_promoters) %>% unique();\n",
    "active_isoforms = ref_transcripts[ hasActivePro ];\n",
    "active_isoforms;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b520ea37-8e7b-4777-b928-e2564b036096",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use PROseq data to define active transcription units\n",
    "PROseq = readRDS(\"/net/bmc-lab2/data/lab/scgroseq/group/notebooks/data/PROv2p8_consolidated.rds\");\n",
    "PROseq = PROseq %>%\n",
    "    filter(mapq >= 3 & miRQC & umiQC) %>%\n",
    "    resize(width=1, fix=\"end\") %>%\n",
    "    subsetByOverlaps(active_isoforms);\n",
    "#PROseq\n",
    "\n",
    "active_isoforms$pol_density = countOverlaps(active_isoforms, PROseq) / width(active_isoforms);\n",
    "#hist(log(active_isoforms$pol_density))\n",
    "\n",
    "active_genes = active_isoforms %>%\n",
    "    arrange(desc(pol_density)) %>%\n",
    "    filter(!duplicated(name)) %>%\n",
    "    mutate(dREG=T);\n",
    "active_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e3466c6e-7857-4fc1-ae70-29efb278d8af",
   "metadata": {},
   "outputs": [],
   "source": [
    "# re-add inactive genes\n",
    "missed_genes = subsetByOverlaps( rgConsensus, active_genes, invert=T );\n",
    "missed_genes$dREG=F;\n",
    "\n",
    "all_genes = c(active_genes, missed_genes) %>%\n",
    "    filter(!duplicated(name));\n",
    "length(all_genes)\n",
    "names(all_genes) = all_genes$name;\n",
    "all_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b149df2-1091-46f9-a25f-8b7dc33a0d68",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Prioritize largest genes, then remove smaller overlaps.\n",
    "filtered_genes = all_genes[ order(-width(all_genes)) ]\n",
    "hits = findOverlaps( filtered_genes, filtered_genes );\n",
    "# after sorting by size, earlier index = larger gene\n",
    "hits = hits[ hits@from <  hits@to ];\n",
    "# remove overlapping regions with later index = smaller gene\n",
    "filtered_genes = filtered_genes[-unique(hits@to)] %>%\n",
    "    sort() %>%\n",
    "    select(c(dREG, name));\n",
    "\n",
    "names(filtered_genes) = filtered_genes$name;\n",
    "summary(filtered_genes$dREG)\n",
    "filtered_genes"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee4c1415-3410-4ba2-bd7c-9826f284c9d5",
   "metadata": {},
   "source": [
    "## Refine groHMM using active genes\n",
    "1. Find overlaps between groHMM and active genes\n",
    "2. Split groHMM TUs with 3 kbp gap if they \"read through\" into a second active gene\n",
    "3. Extend genes using refined groHMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "64641bab-6c05-42f1-b167-327081135dd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load optimized groHMM transcription unit calls\n",
    "HMM_transcripts = read_bed(file=\"/net/bmc-lab2/data/lab/scgroseq/group/notebooks/data/groHMM_mES_BRsComb_LP-50_UTS10.bed\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94e821db-43c5-4be8-bc6a-af64eb0d65fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "getGeneID = function( regions, genes ) {\n",
    "    hits = findOverlaps( regions, genes );\n",
    "    geneIDs = rep(NA, length(regions));\n",
    "    geneIDs[hits@from] = genes$name[ hits@to ];\n",
    "    return(geneIDs);\n",
    "}\n",
    "\n",
    "HMM_transcripts$name = getGeneID( HMM_transcripts, filtered_genes );"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c05fd14a-1487-4e6e-8c42-7750f0234946",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "hits = findOverlaps( HMM_transcripts, filtered_genes );\n",
    "# check frequency distribution of overlaps\n",
    "numHits = table(queryHits(hits));\n",
    "numHits = as.data.frame( numHits, stringsAsFactors = F );\n",
    "colnames(numHits) = c(\"index\", \"count\");\n",
    "numHits$index = as.integer(numHits$index);\n",
    "head(numHits);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "956c3549-1faa-4722-928e-2b21f4e73869",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot( numHits, aes(x=count)) +\n",
    "    geom_histogram(binwidth=1) +\n",
    "    xlim(0, 8) +\n",
    "    ggtitle(\"Genes per groHMM unit\") +\n",
    "    xlab(\"Number of genes\") +\n",
    "    ylab(\"Number of units\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f6af552-6081-476d-b830-afbb0204dfe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create 100 bp window upstream of active reference genes\n",
    "activeGpro = subsetByOverlaps( filtered_genes, HMM_transcripts ) %>%\n",
    "    promoters(upstream=100, downstream=0);\n",
    "\n",
    "# subtract promoters from TUs = split TUs between genes\n",
    "HMM_transcripts = subtract( HMM_transcripts, activeGpro ) %>% unlist();\n",
    "\n",
    "# update gene assignments\n",
    "HMM_transcripts$name = getGeneID( HMM_transcripts, filtered_genes );"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5bcd3499-daf1-4d42-93d5-049ed5449c7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Group/merge HMMs overlapping the same gene.\n",
    "HMM_transcripts = HMM_transcripts %>%\n",
    "    as.data.frame %>%\n",
    "    filter( !is.na(name) ) %>%\n",
    "    group_by(name) %>%\n",
    "    mutate( start=min(start), end=max(end) ) %>%\n",
    "    ungroup() %>%\n",
    "    filter(!duplicated(name)) %>%\n",
    "    GRanges;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ae31848-c31e-40a2-b3b9-ff797398a695",
   "metadata": {},
   "outputs": [],
   "source": [
    "plUnits = HMM_transcripts[ strand(HMM_transcripts) == \"+\" ];\n",
    "mnUnits = HMM_transcripts[ strand(HMM_transcripts) == \"-\" ];\n",
    "\n",
    "# set gene 5' end according to reference annotation\n",
    "# plus strand 5' = start coordinate\n",
    "plUnits = data.frame(\n",
    "        chr = seqnames(plUnits),\n",
    "#        start = start(ref_transcripts[ as.character(plUnits$name) ]),\n",
    "        start = start(filtered_genes[ as.character(plUnits$name) ]),\n",
    "        end = end(plUnits),\n",
    "        strand=\"+\",\n",
    "        name=plUnits$name\n",
    "    ) %>%\n",
    "        filter( end > start ) %>%\n",
    "        GRanges();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c76192f-9343-4a62-9ed6-c0a0abfd6e96",
   "metadata": {},
   "outputs": [],
   "source": [
    "# minus strand 5' = end coordinate\n",
    "mnUnits = data.frame(\n",
    "        chr = seqnames(mnUnits),\n",
    "        start = start(mnUnits),\n",
    "#        end = end(ref_transcripts[ as.character(mnUnits$name) ]),\n",
    "        end = end(filtered_genes[ as.character(mnUnits$name) ]),\n",
    "        strand=\"-\",\n",
    "        name=mnUnits$name\n",
    "    ) %>%\n",
    "        filter( end > start ) %>%\n",
    "        GRanges();\n",
    "\n",
    "refinedUnits = c( plUnits, mnUnits );"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03abfdaa-c5ba-4b96-921a-35053c6d23f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# merge overlapping genes with same name\n",
    "refinedUnits = refinedUnits %>%\n",
    "    group_by(name) %>%\n",
    "    reduce_ranges_directed() %>%\n",
    "    filter(!duplicated(name)) %>%\n",
    "    sort();\n",
    "\n",
    "refinedUnits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3febc5b5-ea10-4ed6-b04c-e4a3df4d662f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reset any genes shortened by HMM\n",
    "xgene = refinedUnits %>% filter( name %in% rgConsensus$name );\n",
    "#xrefg = rgConsensus[ as.character(xgene$name) ];\n",
    "xrefg = filtered_genes[ as.character(xgene$name) ];\n",
    "shortg = xgene[ width(xrefg) > width(xgene) ]$name;\n",
    "refinedUnits = refinedUnits %>%\n",
    "    filter( ! name %in% shortg ) %>%\n",
    "    append( xrefg[ as.character(shortg) ] ) %>%\n",
    "    sort;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c967c95-8bdb-4db3-90ac-4ed56478872f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reset any genes missed by HMM\n",
    "missed_genes = subsetByOverlaps( rgConsensus, refinedUnits, invert=T );\n",
    "refinedUnits = c(refinedUnits, missed_genes);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "943aa430-e3f7-4bb8-b6e7-7b0bc121acca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for any overlaps among features and remove smaller\n",
    "refinedUnits = refinedUnits[order(-width(refinedUnits))];\n",
    "hits = findOverlaps( refinedUnits, refinedUnits );\n",
    "hits = hits[ hits@from < hits@to ];\n",
    "refinedUnits[ hits@from ]$name\n",
    "refinedUnits[ hits@to ]$name\n",
    "refinedUnits = refinedUnits[ -unique(hits@to) ] %>%\n",
    "    sort;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52caa6ab-37d0-437d-95be-48f60757c21b",
   "metadata": {},
   "outputs": [],
   "source": [
    "names(refinedUnits) = refinedUnits$name;\n",
    "\n",
    "refinedUnits  = refinedUnits %>%\n",
    "    filter( names %in% names(filtered_genes) ) %>%\n",
    "    filter( !duplicated(names) ) %>%\n",
    "    mutate( dREG = filtered_genes[names]$dREG );\n",
    "refinedUnits\n",
    "\n",
    "write_bed(refinedUnits, file=\"../data/groHMM_dREG_refinedGenes_mES_mm10.bed\");\n",
    "write_rds(refinedUnits, file=\"../data/groHMM_dREG_refinedGenes_mES_mm10.rds\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c507a497",
   "metadata": {},
   "outputs": [],
   "source": [
    "e = evaluateHMMInAnnotations(refinedUnits, filtered_genes)\n",
    "e$eval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ff5d8fa-cbbc-4dd8-92ec-4bcfc446f267",
   "metadata": {},
   "outputs": [],
   "source": [
    "# match reference genes to refined units\n",
    "sum(refinedUnits$name %in% filtered_genes$name);\n",
    "sum(filtered_genes$name %in% refinedUnits$name);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4164aba2-af7a-4224-a67d-0a7c57fc84cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "#xgene = refinedUnits %>% filter( gene %in% rgConsensus$name );\n",
    "#xrefg = rgConsensus[ as.character(xgene$name) ];\n",
    "xgene = refinedUnits %>% filter( name %in% filtered_genes$name );\n",
    "xrefg = filtered_genes[ as.character(xgene$name) ];\n",
    "\n",
    "gstats = data.frame(\n",
    "    oldw = width(xrefg),\n",
    "    neww = width(xgene),\n",
    "    deltaw = width(xgene) - width(xrefg)\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee37d59e-4a63-4308-ac2e-c5aa425ac3db",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot( gstats, aes(x=oldw) ) + \n",
    "    geom_histogram(breaks=(0:50)*1000) +\n",
    "    ylim(0, 1500) +\n",
    "    ggtitle(\"Reference gene widths\") +\n",
    "    xlab(\"width\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3942d936-eded-4753-9039-6a70321ea48d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot( gstats, aes(x=neww) ) + \n",
    "    geom_histogram(breaks=(0:50)*1000) +\n",
    "    ylim(0, 1500) +\n",
    "    ggtitle(\"Refined gene widths\") +\n",
    "    xlab(\"width\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29c8583d-6250-42e4-a01e-f7d61a50d993",
   "metadata": {},
   "outputs": [],
   "source": [
    "ggplot( gstats, aes(x=deltaw) ) + \n",
    "    geom_histogram(breaks=(0:30)*1000) +\n",
    "    ggtitle(\"Change in gene widths\") +\n",
    "    ylim(0, 1500) +\n",
    "    xlab(\"width\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9e2dad3-0266-4a17-8cab-89b65ccf4085",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene list of extensions >20kbp\n",
    "xgene[ gstats$deltaw > 20000 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9612157-b912-46e8-8f3e-afae4b2d2ea1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# % of all consensus genes\n",
    "100 * length(refinedUnits) / length(filtered_genes)\n",
    "length(refinedUnits)\n",
    "length(filtered_genes)\n",
    "# limit to protein-coding genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "225fd6d9-5c30-4370-aabe-c10343957d77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# % of genome that is transcribed\n",
    "txn = refinedUnits %>% reduce_ranges_directed();\n",
    "100 * sum(width(txn)) / 5.9E9"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d885fc61-0cbd-447e-adf7-df903a45d0e7",
   "metadata": {},
   "source": [
    "### Save dREG filtered genes\n",
    "Same length and names as the groHMM + dREG filtered genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e631215d-2ec6-4cda-96a0-77d2fbc4e375",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered_genes = filtered_genes %>%\n",
    "    filter( name %in% names(refinedUnits) );\n",
    "filtered_genes\n",
    "\n",
    "write_rds(filtered_genes, file=\"../data/dREG_refinedGenes_mES_mm10.rds\");\n",
    "write_bed(filtered_genes, file=\"../data/dREG_refinedGenes_mES_mm10.bed\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19256276-d837-4653-9506-f840ec6c7fac",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
