{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4d5ca2ef-0b8f-4f3e-a81b-9c3a0fc3dc07",
   "metadata": {},
   "source": [
    "## Refine groHMM transcription units overlapping candidate enhancers (gene-distal dREG peaks)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5ad7614-cd9e-4eaf-b2a2-df0fd990785e",
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths(\"/home/mahat/.conda/envs/r422/lib/R/library\")\n",
    ".libPaths()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a7f3d8a-5aa1-47db-8238-ae7e5ba4d88e",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressPackageStartupMessages({\n",
    "    library(rtracklayer)\n",
    "    library(plyranges)\n",
    "    library(tidyverse)\n",
    "    library(foreach)\n",
    "});"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b5f3cf5-36f7-4963-b87b-62b038a7efa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(\n",
    "    repr.plot.width=3,\n",
    "    repr.plot.height=3,\n",
    "    jupyter.plot_mimetypes = \"image/svg+xml\"\n",
    ");\n",
    "theme_set(theme_classic());"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3464965a-4bce-4597-b46f-d144766b8cd1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "## load genes:\n",
    "dREGrefingedGenes =          readRDS(\"../data/dREG_refinedGenes_mES_mm10.rds\");\n",
    "names(dREGrefingedGenes)=NULL;\n",
    "dREGrefingedGenes\n",
    "groHMMrefingedGenes = readRDS(\"../data/groHMM_dREG_refinedGenes_mES_mm10.rds\");\n",
    "names(groHMMrefingedGenes)=NULL;\n",
    "groHMMrefingedGenes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb447dbf-6031-4950-87e5-0d1f983f3162",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "inactivegenes = groHMMrefingedGenes  %>%\n",
    "    filter( !dREG ) %>%\n",
    "    select( name );\n",
    "inactivegenes\n",
    "\n",
    "activegenes = groHMMrefingedGenes  %>%\n",
    "    filter( dREG ) %>%\n",
    "    select( name );\n",
    "activegenes\n",
    "\n",
    "summary(width(activegenes))\n",
    "summary(width(inactivegenes))\n",
    "\n",
    "# # specify maximum gene length to be used for removing overlapping dREG enhancers\n",
    "# # No need to remove dREG after 30kb since we dont use gene signal after 30kb\n",
    "# geneLength = 30000;\n",
    "# truncgenes = activegenes;\n",
    "\n",
    "# longf = which( width(truncgenes) > geneLength );\n",
    "# truncgenes[longf] = truncgenes[longf] %>%\n",
    "#     resize( width = geneLength, fix=\"start\" );\n",
    "# summary(width(truncgenes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c050353-8875-480f-881c-dea941a8e47e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load raw HMM calls\n",
    "raw_HMM = read_bed(file=\"../data/groHMM_mES_BRsComb_LP-50_UTS10.bed\");\n",
    "raw_HMM$score=NULL;\n",
    "raw_HMM$name=NULL;\n",
    "summary(width(raw_HMM))\n",
    "\n",
    "# trim long raw HMM calls to max of 10 kb\n",
    "# removes unnecessary overlap with dREG\n",
    "# results in longest stitched dREG to 5 kb\n",
    "longf = which( width(raw_HMM) >= 5000 );\n",
    "raw_HMM[longf] = raw_HMM[longf] %>%\n",
    "     resize( width = 5000, fix=\"start\" );\n",
    "summary(width(raw_HMM))\n",
    "raw_HMM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c68cda64-2880-4de9-ac57-3abcd10fa624",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load dREG peak calls and convert to GRanges\n",
    "# this file is generated by calling dREG (Charles Danko Lab, Cornell) on bulk PROseq data, using the server: https://dreg.dnasequence.org/\n",
    "dREG = read.table(\"../data/PROseq_mES_BRsComb.dREG.peak.full.bed\", header=F, stringsAsFactors=F);\n",
    "colnames(dREG) = c(\"chr\", \"start\", \"end\", \"score\", \"pval\", \"center\");\n",
    "dREG = GRanges(dREG)\n",
    "dREG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dfc8635f-5401-4996-b1f1-0e64f26b6c19",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(raw_HMM);\n",
    "length(dREG);\n",
    "\n",
    "# remove enhancers within 500bp of active genes\n",
    "# (invert=T: keep regions that DO NOT overlap active genes)\n",
    "HMMenh  = subsetByOverlaps( raw_HMM, activegenes, invert=T, ignore.strand=T, maxgap = 1500 );\n",
    "dREGenh = subsetByOverlaps(    dREG, activegenes, invert=T, ignore.strand=T, maxgap = 1500 );\n",
    "length(HMMenh);\n",
    "length(dREGenh);\n",
    "\n",
    "# DEACTIVATING THIS FOR NOW\n",
    "# # add enhancers on non-gene strand supported by HMM and dREG\n",
    "# HMM_nts = truncgenes;\n",
    "# strand(HMM_nts) = ifelse( strand(truncgenes) == \"+\", \"-\", \"+\" );\n",
    "# geneenh = subsetByOverlaps( raw_HMM, dREG, ignore.strand=T, maxgap=500 ) %>%\n",
    "#     subsetByOverlaps( HMM_nts );\n",
    "\n",
    "# HMMenh = append(HMMenh, geneenh);\n",
    "\n",
    "# length(geneenh);\n",
    "# length(HMMenh);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1026e7b3-ef8b-43f1-93a4-2267a51a5de2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get distance from each enhancer to nearest active gene\n",
    "enh_dist = as.data.frame( distanceToNearest( dREGenh, activegenes, ignore.strand=T ) );\n",
    "\n",
    "ggplot( enh_dist, aes(x=distance)) +\n",
    "    geom_histogram(binwidth=500) +\n",
    "    ggtitle(\"Enhancer-gene distance\") +\n",
    "    xlim(0, 5E4) +\n",
    "    # ylim(0, 600) +\n",
    "    xlab(\"Distance to nearest active gene\") +\n",
    "    ylab(\"Number of enhancers\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f8fc306-c8e9-487d-8306-9c29a522b63c",
   "metadata": {},
   "outputs": [],
   "source": [
    "hits = findOverlaps( dREGenh, HMMenh, maxgap=500 );\n",
    "length(unique(hits@from))\n",
    "length(unique(hits@to))\n",
    "# inspect examples of groHMM/dreg enhancers that aren't in agreement"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73776fe3-75c8-41e9-b129-5f94443f5317",
   "metadata": {},
   "outputs": [],
   "source": [
    "# this block was not here before, I added it:\n",
    "HMMenh = unique(HMMenh[hits@to])\n",
    "summary(width(HMMenh))\n",
    "summary(width(dREGenh))\n",
    "HMMenh"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19d6f3c3-4c5a-455c-a624-86b0aaf998da",
   "metadata": {},
   "outputs": [],
   "source": [
    "# resize dREG-specific enhancers to average of HMMenh,\n",
    "# append all HMMenh and merge,\n",
    "# filter against genes again after extension,\n",
    "# add enhancer names\n",
    "\n",
    "## In this version, not requiring dREGenh to not overlap HMMenh\n",
    "enhancers = subsetByOverlaps(dREGenh, HMMenh, invert=T, ignore.strand=T, maxgap=500) %>%\n",
    "    resize( width=3000, fix=\"center\" ) %>%\n",
    "    append( HMMenh ) %>%\n",
    "    mutate( strand=\"*\" ) %>%\n",
    "    reduce_ranges() %>%\n",
    "    subsetByOverlaps( activegenes, invert=T, ignore.strand=T ) %>%\n",
    "    filter( seqnames != \"chrM\" );\n",
    "summary(width(enhancers));\n",
    "enhancers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "14a6411d-1177-4b3a-aeef-7b5c5ab09a09",
   "metadata": {},
   "outputs": [],
   "source": [
    "# resize short enhancers to 3kbp\n",
    "short_enh = enhancers %>%\n",
    "    filter( width < 3000 ) %>%\n",
    "    resize( width=3000, fix=\"center\" );\n",
    "\n",
    "# merge overlaps and add back into set\n",
    "enhancers = enhancers %>%\n",
    "    subsetByOverlaps( short_enh, invert=T ) %>%\n",
    "    append( short_enh ) %>%\n",
    "    reduce_ranges() %>%\n",
    "    mutate( name=paste0(seqnames, \"-\", start) );\n",
    "enhancers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "08781067-126c-4bac-a599-edd939ca2ab8",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = append(groHMMrefingedGenes, enhancers) %>%\n",
    "    sort();\n",
    "features\n",
    "write_bed(features, file=\"../data/groHMM_dREG_refinedFeatures_mES_mm10.bed\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9ac1082-6687-4285-8296-00fa20a793e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = append(groHMMrefingedGenes, enhancers) %>%\n",
    "    sort();\n",
    "features\n",
    "write_bed(features, file=\"../data/groHMM_dREG_refinedFeatures_mES_mm10.bed\");\n",
    "\n",
    "dREGfeatures = append(dREGrefingedGenes, enhancers) %>%\n",
    "    sort();\n",
    "dREGfeatures\n",
    "write_bed(dREGfeatures,    file=\"../data/dREG_refinedFeatures_mES_mm10.bed\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "05a2ed53-4cae-4fa9-8d82-629ca5efcb7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(enhancers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2be43a8d-512a-40d1-acca-d4320890e6b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "enh_size = data.frame( w=width(enhancers) );\n",
    "\n",
    "ggplot( enh_size, aes(x=w)) +\n",
    "    geom_histogram(binwidth=100) +\n",
    "    ggtitle(\"Enhancer widths\") +\n",
    "    xlim(0, 7000) +\n",
    "    xlab(\"Width\") +\n",
    "    ylab(\"Number of enhancers\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6f585b3-14ed-41d0-996f-ba5167bc2155",
   "metadata": {},
   "outputs": [],
   "source": [
    "# estimate % of genome for transcribed enhancers\n",
    "sum(width(enhancers))/3E7"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "accb0758-6efd-4f5d-8911-43fc8edc1907",
   "metadata": {},
   "outputs": [],
   "source": [
    "# check for any overlaps among features\n",
    "# ensure no overlaps to avoid double-counting reads\n",
    "hits = findOverlaps( features, features );\n",
    "ovpairs = hits@from != hits@to;\n",
    "sum( ovpairs );\n",
    "features[ hits@from[ ovpairs ] ];\n",
    "features[ hits@to[ ovpairs ] ];\n",
    "\n",
    "## THE OVERLAP IS ONLY WITH INACTIVE GENES. WITHOUT INCACTIVE GENES, THERE ARE NO OVERLAPPING FEATURES.\n",
    "summary(features[ hits@from[ ovpairs ] ]$dREG)\n",
    "summary(features[ hits@to[ ovpairs ] ]$dREG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "370b6c8f-697b-4f84-9055-1b27589fbd6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = features %>%\n",
    "    filter(substr(name, 0, 3) == 'GN-');\n",
    "summary(width(test));\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fac99953-9500-4146-9219-af5fc40d66e0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6a813ecd-6a11-4141-acfd-fe0cc3eacdda",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
