{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "beb1a6a4-fb5c-41b9-9ffc-c5a1bdc1b9b3",
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths(\"/home/mahat/.conda/envs/r422/lib/R/library\")\n",
    ".libPaths()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9db58c45-23df-41b8-bf99-cdfc9afb5872",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "    library(tidyverse)\n",
    "    library(ggExtra)\n",
    "    library(matrixStats)\n",
    "    library(plyranges)\n",
    "    library(viridis)\n",
    "    library(data.table)\n",
    "    library(ggridges)\n",
    "    library(ggbeeswarm)\n",
    "    library(ggpointdensity)\n",
    "    library(doParallel)\n",
    "});\n",
    "source(\"./scGRO_functions.r\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a623f44-ed47-4336-a09c-e230051a735b",
   "metadata": {},
   "outputs": [],
   "source": [
    "options(\n",
    "    repr.plot.width=4,\n",
    "    repr.plot.height=4,\n",
    "    jupyter.plot_mimetypes = \"image/svg+xml\"\n",
    ");\n",
    "theme_set(theme_classic() +\n",
    "    theme(\n",
    "        axis.title.x = element_text(color=\"black\", size=14, face=\"bold\"), \n",
    "        axis.title.y = element_text(color=\"black\", size=14, face=\"bold\"),\n",
    "        axis.text = element_text(color=\"black\", size=12, face=\"bold\"),\n",
    "        plot.title = element_text(face=\"bold\", size=14, hjust = 0.5),\n",
    "        axis.line = element_blank(),\n",
    "        # axis.ticks = element_blank()\n",
    "        panel.border = element_rect(colour = \"grey\", fill=NA, linewidth=1)\n",
    "    )\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0b9f8d5e-127d-4c0e-9167-7aefa54ba044",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get equation and r^2 as string\n",
    "# https://groups.google.com/forum/#!topic/ggplot2/1TgH-kG5XMA\n",
    "\n",
    "lm_eqn = function(x, y) {\n",
    "    m = lm(y ~ x);\n",
    "    eq = substitute(italic(y) == a + b %.% italic(x)*\",\"~~italic(r)^2~\"=\"~r2, \n",
    "         list(a = format(unname(coef(m)[1]), digits = 2),\n",
    "              b = format(unname(coef(m)[2]), digits = 2),\n",
    "             r2 = format(summary(m)$r.squared, digits = 2)))\n",
    "    as.character(as.expression(eq));\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16757113-4d68-4bba-b457-cd5c9c9b0323",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load genes with dREG status for expression filter\n",
    "genesWithdREGstatus = readRDS(\"../data/dREG_refinedGenes_mES_mm10.rds\");\n",
    "genesWithdREGstatus\n",
    "unexpressed = genesWithdREGstatus %>% \n",
    "    filter(dREG == F);\n",
    "# unexpressed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b697173d-6329-4bca-9a78-b7ff44c6efde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify feature length (half at the start and other half at the end)\n",
    "# to be trimmed to eliminate the effect of paused Pol II at TSS and TES\n",
    "trimEndLength =  1000;\n",
    "# specify maximum gene length to be used for G-E correlation:\n",
    "maxGeneLengthAfterTrim = 10000;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22f4fd9a-cc8b-46c6-99ab-fe447aeada02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load genes\n",
    "features = read_bed(\"../data/dREG_refinedFeatures_mES_mm10_OSNcustomEnhancers_SEs.bed\");\n",
    "# features = read_bed(\"../data/mES_BRsComb_dREGfiltered_features_customized_OSNenhancersPlusSEs_v1.bed\");\n",
    "# features = read_bed(\"../data/groHMM_mES_BRsComb_LP-50_UTS10_features_customized_OSNenhancersPlusSEs_v2.bed\");\n",
    "names(features) = features$name;\n",
    "features$name=NULL;\n",
    "features$score=NULL;\n",
    "length(features)\n",
    "\n",
    "genes = features %>%\n",
    "    filter( substr(names, 0, 3) == \"GN-\" ) %>%\n",
    "    # filter( names %in% names(genesWithdREGstatus[genesWithdREGstatus$dREG == T]) | \n",
    "    #     !names %in% names(genesWithdREGstatus) ) %>%\n",
    "    # removing features from chrM - resizing chrM dREG results in negative granges:\n",
    "    filter( seqnames != \"chrM\" ) %>%\n",
    "    # filter genes for minimum length:\n",
    "    filter( width >= maxGeneLengthAfterTrim + trimEndLength ) %>%\n",
    "    anchor_center() %>%\n",
    "    mutate(width = width - trimEndLength);\n",
    "# truncate long genes\n",
    "longf = which( width(genes) > maxGeneLengthAfterTrim );\n",
    "genes[longf] = genes[longf] %>%\n",
    "    resize( width = maxGeneLengthAfterTrim, fix=\"start\" );\n",
    "summary(width(genes))\n",
    "genes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ac8cbe9-811a-48a2-8df4-1736a15ce1b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load groHMM-extended genes and enhancers\n",
    "dREG = read.table(\"../data/PROseq_mES_BRsComb.dREG.peak.full.bed\", header=F, stringsAsFactors=F);\n",
    "colnames(dREG) = c(\"chr\", \"leftEnd\", \"rightEnd\", \"score\", \"pval\", \"center\");\n",
    "dREG = dREG %>%\n",
    "    filter(chr != \"chrM\") %>%\n",
    "    mutate(start = center - 1500) %>%\n",
    "    mutate(end = center + 1500) %>%\n",
    "    ## removing dREG peaks with score lower than the 1st quantile\n",
    "    # filter( score > summary(score)[2] );\n",
    "    GRanges() %>%\n",
    "    # filter out dREG peaks that overlap with genes\n",
    "    subsetByOverlaps( features[ substr(names(features), 0, 3) == \"GN-\" ], invert = TRUE );\n",
    "\n",
    "names(dREG) = paste0(dREG@seqnames, \"-\", dREG$center)\n",
    "dREG$leftEnd = NULL;\n",
    "dREG$rightEnd = NULL;\n",
    "dREG$pval = NULL;\n",
    "\n",
    "# mark overlapping dREG peaks:\n",
    "hits = findOverlaps( dREG );\n",
    "dREG$overlappingEnh = NA;\n",
    "# assign ovelapping dREG peaks\n",
    "dREG$overlappingEnh[hits@from] = names(dREG)[hits@to];\n",
    "# filter peak to only leave one among overlapping enhancers with highest score:\n",
    "dREG = dREG %>%\n",
    "    # group overlapping enhancers and keep one with max score\n",
    "    group_by( overlappingEnh ) %>%\n",
    "    filter( score == max(score) ) %>%\n",
    "    ungroup();\n",
    "dREG\n",
    "\n",
    "# make dREG anti-sense\n",
    "dREGantis = dREG %>%\n",
    "    # removing the 250 nt at the center as paused peak\n",
    "    mutate(end = center - 251) %>%\n",
    "    mutate(strand = \"-\") \n",
    "# names(dREGantis) = paste0(names(dREGantis), \"_as\")\n",
    "\n",
    "# make dREGsense\n",
    "dREGsense = dREG %>%\n",
    "    # removing the 250 nt at the center as paused peak\n",
    "    mutate(start = center + 251) %>%\n",
    "    mutate(strand = \"+\")\n",
    "# names(dREGsense) = paste0(names(dREGsense), \"_s\")\n",
    "\n",
    "enhancers = c(dREGantis, dREGsense) %>%\n",
    "    sort();\n",
    "summary(width(enhancers))\n",
    "enhancers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67b570a7-2729-493b-9e4e-740ad642cba8",
   "metadata": {},
   "outputs": [],
   "source": [
    "features = c(genes, enhancers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "49600903-b52e-4608-a3c0-5193ecf946ee",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = readRDS( \"../data/scGROv2p8_mapq3qc_max10kbp_filtered_counts.rds\" );\n",
    "dim(counts)\n",
    "hist(colSums(counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7df1b34c-7c17-4eec-94f0-3d7910f3c0e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # filter cells for a tighter range of expression\n",
    "# counts = counts[, colSums(counts) > 500]\n",
    "# dim(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29cd00db-057b-462e-bc05-8e9ce908c83a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot mean vs variance\n",
    "data.frame( cmean=rowMeans(counts), cvar=apply(counts, 1, var) ) %>%\n",
    "    ggplot(aes(x=cmean, y=cvar)) +\n",
    "    geom_pointdensity(size=0.5) +\n",
    "    scale_color_viridis_c() +\n",
    "    ggtitle(\"scGROseq counts\") +\n",
    "    # scale_x_log10() + \n",
    "    # scale_y_log10() +\n",
    "    xlab(\"Mean\") +\n",
    "    ylab(\"Variance\") +\n",
    "    xlim(0, 1) +\n",
    "    ylim(0, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b39a870d-0074-4f1b-8ee8-a11804837f21",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fraction of reads mapping to unexpressed genes\n",
    "dim(counts[rownames(counts) %in% names(unexpressed),])\n",
    "sum(rowSums(counts[rownames(counts) %in% names(unexpressed),]))/sum(sum(counts))*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7eea13e-a510-4473-9242-42e3696118fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "scGRO  = readRDS(\"../data/scGROv2p8_consolidated.rds\");\n",
    "scGRO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9aa45961-dd88-4b78-9ec0-843b33461615",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge experiment ID and cell barcode to create unique cell ID across experiments\n",
    "scGRO = scGRO %>%\n",
    "    filter( umiQC & plateQC & cellQC & countQC & miRQC ) %>%\n",
    "    mutate(cellID = factor( paste( Exp, Plate, Cell, sep=\"-\") ) ) %>%\n",
    "    # filter( !(Exp == \"Exp236\" & Plate %in% paste0(\"c0\", 5:8)) ) %>%\n",
    "    # filter( !(Exp == \"Exp260b\") ) %>%\n",
    "    resize(width=1, fix=\"end\") %>%\n",
    "    select(cellID) %>%\n",
    "    filter(cellID %in% colnames(counts)) %>%\n",
    "    subsetByOverlaps(features);\n",
    "names(scGRO) = NULL;\n",
    "scGRO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f43591d-a7d2-4793-92a4-40379ce95f1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "scGRO$cellID = droplevels(scGRO$cellID);\n",
    "\n",
    "# mark features for each read\n",
    "hits = findOverlaps( scGRO, features );\n",
    "scGRO$feature = NA;\n",
    "scGRO$feature[hits@from] = names(features)[hits@to];\n",
    "scGRO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "493d2405-4ac8-4d95-8a4e-9563232bb42d",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(unique(scGRO$feature))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "079bdfd0-32c6-445d-a367-bb2af7f3ca51",
   "metadata": {},
   "outputs": [],
   "source": [
    "# estimate burst rate genome-wide\n",
    "#burst_window_size = 30000;\n",
    "pol_ii_speed = 2500*60; # per hour\n",
    "total_cells=n_distinct(scGRO$cellID)\n",
    "nalleles=2*total_cells;\n",
    "# not true capture rate of scGRO, but for regions of genome used in this analysis\n",
    "# Most of the drop outs are paused Pol II. although the capture efficiency looks like 5% based on transcribing Pol II, the gene body capture efficiency is likely 10%\n",
    "capture_efficiency = 0.10;\n",
    "gene_txn_time = width(features)/pol_ii_speed; # number of hours to transcribe this feature\n",
    "names(gene_txn_time) = names(features);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26069691-a3ca-492c-8e52-1df5a46b4de8",
   "metadata": {},
   "outputs": [],
   "source": [
    "bursts_10k = scGRO %>%\n",
    "    #mutate( burstID = paste(cellID, floor(start/burst_window_size)) ) %>%\n",
    "    as.data.table %>%\n",
    "    select(-end, -width) %>%\n",
    "    group_by(feature) %>%\n",
    "    summarise( N_reads = n(), N_bursts=n_distinct(cellID) ) %>% #, N_bursts=n_distinct(burstID) ) %>%\n",
    "    mutate( chr = as.character(seqnames(features[feature])) ) %>%\n",
    "    mutate( start = start(features[feature]), end = end(features[feature]) ) %>%\n",
    "    # Enhancers had two 2500 window in sense and antisense. Total length is 2*2500\n",
    "    mutate( length = ifelse( substr(feature, 0, 3) == \"GN-\", width(features[feature]), 2*width(features[feature])) ) %>%\n",
    "    mutate( meanExpPerKb = N_reads/total_cells/length*1000 ) %>%\n",
    "    mutate( burst_size = N_reads/N_bursts ) %>%\n",
    "    # adjust for capture efficiency\n",
    "    mutate( burst_size = 1+(burst_size-1)/capture_efficiency ) %>%\n",
    "   # mutate( burst_rate = N_bursts/gene_txn_time[feature]/nalleles ) %>%\n",
    "    mutate( burst_rate = N_bursts/gene_txn_time[feature]/ifelse( chr %in% c(\"chrX\", \"chrY\"), total_cells, 2*total_cells) ) %>%\n",
    "    mutate( detect_rate = burst_size*capture_efficiency ) %>%\n",
    "    mutate( burst_rate = burst_rate/ifelse(detect_rate>1, 1, detect_rate) ) %>%\n",
    "    select(feature, chr, start, end, length, meanExpPerKb, N_reads, N_bursts, burst_rate, burst_size );\n",
    "dim(bursts_10k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11b746d5-5593-40f0-a7eb-6d7d247cd055",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# restrict to first 5kbp of features\n",
    "features_5k = features;\n",
    "longf = width(features_5k) >= 5000;\n",
    "\n",
    "features_5k[longf] = features_5k[longf] %>%\n",
    "    resize(width=5000, fix=\"start\");\n",
    "\n",
    "scGRO_5k = scGRO %>%\n",
    "    subsetByOverlaps(features_5k);\n",
    "\n",
    "burst_window_size_5k = 5000;\n",
    "gene_txn_time_5k = width(features_5k)/pol_ii_speed;\n",
    "names(gene_txn_time_5k) = names(features_5k);\n",
    "\n",
    "bursts_5k = scGRO_5k %>%\n",
    "    #mutate( burstID = paste(cellID, floor(start/burst_window_size)) ) %>%\n",
    "    as.data.table %>%\n",
    "    select(-end, -width) %>%\n",
    "    group_by(feature) %>%\n",
    "    summarise( N_reads_5k=n(), N_bursts_5k=n_distinct(cellID) ) %>% #, N_bursts=n_distinct(burstID) ) %>%\n",
    "    mutate( chr=as.character(seqnames(features[feature])) ) %>%\n",
    "    mutate( start = start(features[feature]), end = end(features[feature]) ) %>%\n",
    "    # Enhancers had two 2500 window in sense and antisense. Total length is 2*2500\n",
    "    mutate( length = ifelse( substr(feature, 0, 3) == \"GN-\", width(features[feature]), 2*width(features[feature])) ) %>%\n",
    "    mutate( meanExpPerKb_5k = N_reads_5k/total_cells/length*1000 ) %>%\n",
    "    mutate( burst_size_5k = N_reads_5k/N_bursts_5k ) %>%\n",
    "    # adjust for capture efficiency\n",
    "    mutate( burst_size_5k = 1+(burst_size_5k-1)/capture_efficiency ) %>%\n",
    "    # mutate( burst_rate_5k = N_bursts_5k/gene_txn_time[feature]/nalleles ) %>%\n",
    "    mutate( burst_rate_5k = N_bursts_5k/gene_txn_time[feature]/ifelse( chr %in% c(\"chrX\", \"chrY\"), total_cells, 2*total_cells) ) %>%\n",
    "    mutate( detect_rate   = burst_size_5k*capture_efficiency ) %>%\n",
    "    mutate( burst_rate_5k = burst_rate_5k/ifelse(detect_rate>1, 1, detect_rate) ) %>%\n",
    "    select(feature, N_reads_5k, N_bursts_5k, burst_rate_5k, burst_size_5k );\n",
    "dim(bursts_5k)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26b43933-22c7-4998-be6a-f253db68a41f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#join 10k and 5K bursting kinetics:\n",
    "bursts = left_join( bursts_10k, bursts_5k, by=\"feature\" ) %>%\n",
    "    mutate( Type = ifelse(substr(feature, 0, 3) == \"GN-\", \"Gene\", \"Enhancer\") ) %>%\n",
    "    mutate( Type = factor(Type, levels = c(\"Gene\", \"Enhancer\")) ) %>%\n",
    "    na.omit() %>% # without = 24083\n",
    "    arrange(desc(N_bursts));\n",
    "dim(bursts)\n",
    "head(bursts)\n",
    "\n",
    "fwrite(bursts, file=\"../data/scGROv2p8_max10kbp_max5kbp_burst_rate_0p10_captureEfficiency.csv\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db3ad6f0-4247-43d5-a31b-d4ba713b8cbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "summaryTable = bursts %>%\n",
    "    mutate( Type = ifelse(substr(feature, 0, 2) == \"GN\", \"Gene\", \"Enhancer\") ) %>%\n",
    "    mutate( Type = factor(Type, levels = c(\"Gene\", \"Enhancer\")) );\n",
    "\n",
    "# summary of burst frequency:\n",
    "br10 = summaryTable %>% group_by(Type) %>%\n",
    "    summarise(min(burst_rate), median(burst_rate), mean(burst_rate), max(burst_rate))\n",
    "#summary of time between bursts:\n",
    "bi10 = summaryTable %>% group_by(Type) %>%\n",
    "    summarise(min(1/burst_rate), median(1/burst_rate), mean(1/burst_rate), max(1/burst_rate))\n",
    "# summary of burst size:\n",
    "bs10 = summaryTable %>% group_by(Type) %>%\n",
    "    summarise(min(burst_size), median(burst_size), mean(burst_size), max(burst_size))\n",
    "\n",
    "br5 = summaryTable %>% group_by(Type) %>%\n",
    "    summarise(min(burst_rate_5k), median(burst_rate_5k), mean(burst_rate_5k), max(burst_rate_5k))\n",
    "bi5 = summaryTable %>% group_by(Type) %>%\n",
    "    summarise(min(1/burst_rate_5k), median(1/burst_rate_5k), mean(1/burst_rate_5k), max(1/burst_rate_5k))\n",
    "bs5 = summaryTable %>% group_by(Type) %>%\n",
    "    summarise(min(burst_size_5k), median(burst_size_5k), mean(burst_size_5k), max(burst_size_5k))\n",
    "\n",
    "brAll = rbind(br10, setNames(bi10, names(br10)), setNames(bs10, names(br10)), \n",
    "              setNames(br5, names(br10)), setNames(bi5, names(br10)), setNames(bs5, names(br10)))\n",
    "brAll\n",
    "fwrite(brAll, file=\"../data/scGROv2p8_burst_kinetics_0p10_captureEfficiency_burstWindow.csv\");"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "119a695a-322c-4572-b32b-f6725aa62f8a",
   "metadata": {},
   "source": [
    "##### capture efficiency = 0.1 \n",
    "## 10 kb truncated:\n",
    "Type\tmin(burst_rate)\tmedian(burst_rate)\tmean(burst_rate)\tmax(burst_rate)\n",
    "Gene\t0.009491268\t      1.0208386\t            1.278291\t       18.36560\n",
    "Enhancer\t0.011389522\t  0.6833713\t            1.180078\t       83.99151\n",
    "\n",
    "Type\tmin(1/burst_rate)\tmedian(1/burst_rate)\tmean(1/burst_rate)\tmax(1/burst_rate)\n",
    "Gene\t   0.05444961\t         0.9795868\t             2.312488\t        105.36\n",
    "Enhancer   0.01190597\t         1.4633333\t             1.978099\t        87.80\n",
    "\n",
    "Type\tmin(burst_size)\tmedian(burst_size)\tmean(burst_size)\tmax(burst_size)\n",
    "Gene\t      1\t              1\t                 1.231106\t        6.033829\n",
    "Enhancer\t  1\t              1\t                 1.044373\t        70.950363\n",
    "## 5 kb truncated:\n",
    "Type\tmin(burst_rate_5k)\tmedian(burst_rate_5k)\tmean(burst_rate_5k)\tmax(burst_rate_5k)\n",
    "Gene\t    0.00284738\t         0.5694761\t             0.7902558\t         18.33713\n",
    "Enhancer\t0.01138952\t         0.6833713\t             1.1800781\t         83.99151\n",
    "A tibble: 2 × 5\n",
    "Type\tmin(1/burst_rate_5k)\tmedian(1/burst_rate_5k)\tmean(1/burst_rate_5k)\tmax(1/burst_rate_5k)\n",
    "Gene\t     0.05453416\t                1.756000\t          4.123305\t               351.2\n",
    "Enhancer\t 0.01190597\t                1.463333\t          1.978099\t               87.8\n",
    "A tibble: 2 × 5\n",
    "Type\tmin(burst_size_5k)\tmedian(burst_size_5k)\tmean(burst_size_5k)\tmax(burst_size_5k)\n",
    "Gene\t         1\t                 1\t                    1.120727\t      11.00000\n",
    "Enhancer\t     1\t                 1\t                    1.044373\t      70.95036\n",
    "\n",
    "\n",
    "##### capture efficiency = 0.05\n",
    "## 10 kb truncated:\n",
    "Type\tmin(burst_rate)\tmedian(burst_rate)\tmean(burst_rate)\tmax(burst_rate)\n",
    "Gene\t    0.01035411\t    1.825775\t       2.151973\t           36.73121\n",
    "Enhancer\t0.01138952\t    1.366743\t       2.266185\t           141.41944\n",
    "\n",
    "Type\tmin(1/burst_rate)\tmedian(1/burst_rate)\tmean(1/burst_rate)\tmax(1/burst_rate)\n",
    "Gene\t   0.027224806\t         0.5477127\t             1.116992\t         96.58\n",
    "Enhancer   0.007071164\t         0.7316667\t             1.011312\t         87.80\n",
    "\n",
    "Type\tmin(burst_size)\tmedian(burst_size)\tmean(burst_size)\tmax(burst_size)\n",
    "Gene\t       1\t           1\t               1.468401\t        11.06766\n",
    "Enhancer\t   1\t           1\t               1.088745\t        140.90073\n",
    "## 5 kb truncated:\n",
    "Type\tmin(burst_rate_5k)\tmedian(burst_rate_5k)\tmean(burst_rate_5k)\tmax(burst_rate_5k)\n",
    "Gene\t    0.00284738\t           1.082005\t               1.416489\t        36.67426\n",
    "Enhancer\t0.01138952\t           1.366743\t               2.266185\t        141.41944\n",
    "\n",
    "Type\tmin(1/burst_rate_5k)\tmedian(1/burst_rate_5k)\tmean(1/burst_rate_5k)\tmax(1/burst_rate_5k)\n",
    "Gene\t     0.027267081\t            0.9242105\t           2.144820\t              351.2\n",
    "Enhancer\t 0.007071164\t            0.7316667\t           1.011312\t              87.8\n",
    "\n",
    "Type\tmin(burst_size_5k)\tmedian(burst_size_5k)\tmean(burst_size_5k)\tmax(burst_size_5k)\n",
    "Gene\t         1\t                 1\t                    1.241454\t      21.0000\n",
    "Enhancer\t     1\t                 1\t                    1.088745\t      140.9007"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8eaacff1-fde0-400a-80ad-3e1cfbb35159",
   "metadata": {},
   "outputs": [],
   "source": [
    "bkg_bursts = bursts %>%\n",
    "    filter(feature %in% unexpressed$name) %>%\n",
    "    arrange(-burst_rate)\n",
    "# head(bkg_bursts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "877a8d22-df56-403c-86ac-502f9813f2ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the burst size and burst frequence ranked genes for GSEA:\n",
    "rankedList = bursts %>%\n",
    "    filter( substr(feature, 0, 3) == \"GN-\" ) %>%\n",
    "    mutate( name = sub(\"GN-\", \"\", feature) ) %>%\n",
    "    arrange(desc(burst_rate));\n",
    "dim(rankedList)\n",
    "rankedList[1:2, ]\n",
    "out = data.frame(\n",
    "    Gene = rankedList$name)\n",
    "    # BurstFreq = test$burst_rate );\n",
    "out[1:10,]\n",
    "fwrite(out, file=\"../data/BurstFrequency_rankedList.txt\", sep='\\t', col.names=F);\n",
    "\n",
    "rankedList = bursts %>%\n",
    "    filter( substr(feature, 0, 3) == \"GN-\" ) %>%\n",
    "    mutate( name = sub(\"GN-\", \"\", feature) ) %>%\n",
    "    arrange(desc(burst_size));\n",
    "dim(rankedList)\n",
    "rankedList[1:2, ]\n",
    "out = data.frame(\n",
    "    Gene = rankedList$name)\n",
    "    # BurstFreq = test$burst_rate );\n",
    "out[1:10,]\n",
    "fwrite(out, file=\"../data/BurstSize_rankedList.txt\", sep='\\t', col.names=F);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb9ee682-513e-48c9-9603-076d2a8c9272",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Burst size before accounting for capture efficiency:\n",
    "bursts %>%\n",
    "    ggplot(aes(x=N_reads/N_bursts)) +\n",
    "    geom_histogram(binwidth=0.01) +\n",
    "    scale_y_log10() +\n",
    "    xlim(0.99, 1.5) +\n",
    "    xlab(\"Average reads/ feature / active cell\") +\n",
    "    ylab(\"Number of features\")\n",
    "\n",
    "# Burst size after accounting for capture efficiency:\n",
    "bursts %>%\n",
    "    ggplot(aes(x=burst_size)) +\n",
    "    geom_histogram(binwidth=0.25) +\n",
    "    scale_y_log10() +\n",
    "    xlim(0, 10) +\n",
    "    xlab(\"Estimated burst size\") +\n",
    "    ylab(\"Number of features\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3576dd2-5710-40d6-921d-a58313cec0c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "bs = bursts %>%\n",
    "    filter(Type == \"Gene\") %>%\n",
    "    ggplot(aes(x = burst_size)) +\n",
    "    geom_histogram(bins = 30, color = \"white\", fill = \"#39568C\") +\n",
    "    xlim(0, 7.5) +\n",
    "    scale_y_log10() +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Number of features\") \n",
    "bs\n",
    "ggsave(bs, filename = \"../plots/Burst_size_Genes.pdf\", width=4, height=4);\n",
    "\n",
    "bs = bursts %>%\n",
    "    ggplot(aes(x = burst_size, fill=Type)) +\n",
    "    geom_histogram(bins = 16, color = \"white\") +\n",
    "    xlim(0, 7.5) +\n",
    "    scale_y_log10() +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    # facet title:\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "bs\n",
    "ggsave(bs, filename = \"../plots/Burst_size_Genes+Enhancers_facet.pdf\", width=8, height=4);\n",
    "\n",
    "bs = bursts %>%\n",
    "    ggplot(aes(x = burst_size, fill=Type)) +\n",
    "    geom_histogram(bins = 16, color = \"white\") +\n",
    "    xlim(0, 7.5) +\n",
    "    scale_y_log10() +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type, ncol = 1) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    # facet title:\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "bs\n",
    "ggsave(bs, filename = \"../plots/Burst_size_Genes+Enhancers_facet_vertical.pdf\", width=4, height=8);\n",
    "\n",
    "bs = bursts %>%\n",
    "    ggplot(aes(x = burst_size, fill = Type)) +\n",
    "    geom_histogram(bins = 16, position = \"identity\", alpha = 0.5) + # \n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(legend.position = c(0.8, 0.8)) +\n",
    "    xlim(0, 7.5) +\n",
    "    scale_y_log10() +\n",
    "    coord_cartesian(ylim = c(1.5, 10000)) +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\"))\n",
    "bs\n",
    "ggsave(bs, filename = \"../plots/Burst_size_Genes+Enhancers.pdf\", width=4, height=4);\n",
    "\n",
    "bursts %>%\n",
    "    ggplot(aes(x=burst_size, color=Type)) +\n",
    "    theme(legend.position = c(0.2, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=0.7) +\n",
    "    ggtitle(\"Burst size from 10kb gene body\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Number of features\");\n",
    "\n",
    "ks.test(burst_size ~ Type, bursts, alternative = \"greater\", exact=F);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "26f692d2-9308-4bd1-bbee-3ccf5c011d8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "br = bursts %>%\n",
    "    filter(Type == \"Gene\") %>%\n",
    "    ggplot(aes(x = burst_rate)) +\n",
    "    geom_histogram(bins = 30, color = \"white\", fill = \"#39568C\") +\n",
    "    scale_x_log10(limits = c(0.01, 20)) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Number of features\") \n",
    "br\n",
    "ggsave(br, filename = \"../plots/Burst_rate_Genes.pdf\", width=4, height=4);\n",
    "\n",
    "br = bursts %>%\n",
    "    ggplot(aes(x = burst_rate, fill=Type)) +\n",
    "    geom_histogram(bins = 16, color = \"white\") +\n",
    "    scale_x_log10(limits = c(0.01, 20)) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    # facet title:\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "br\n",
    "ggsave(br, filename = \"../plots/Burst_rate_Genes+Enhancers_facet.pdf\", width=8, height=4);\n",
    "\n",
    "br = bursts %>%\n",
    "    ggplot(aes(x = burst_rate, fill=Type)) +\n",
    "    geom_histogram(bins = 16, color = \"white\") +\n",
    "    scale_x_log10(limits = c(0.01, 20)) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type, ncol = 1) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    # facet title:\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "br\n",
    "ggsave(br, filename = \"../plots/Burst_rate_Genes+Enhancers_facet_vertical.pdf\", width=4, height=8);\n",
    "\n",
    "br = bursts %>%\n",
    "    ggplot(aes(x = burst_rate, fill = Type)) +\n",
    "    geom_histogram(bins = 16, position = \"identity\", alpha = 0.5) + # \n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(legend.position = c(0.2, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.01, 20)) +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\"))\n",
    "br\n",
    "ggsave(br, filename = \"../plots/Burst_rate_Genes+Enhancers.pdf\", width=4, height=4);\n",
    "\n",
    "bursts %>%\n",
    "    ggplot(aes(x=burst_rate, color=Type)) +\n",
    "    theme(legend.position = c(0.2, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.01, 20)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=0.7) +\n",
    "    ggtitle(\"Bursts frequency\") +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Number of features\");\n",
    "\n",
    "ks.test(burst_rate ~ Type, bursts, alternative = \"greater\", exact=F);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f6d3328-903c-4bf5-ba78-8c80127403e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "tbb = bursts %>%\n",
    "    filter(Type == \"Gene\") %>%\n",
    "    ggplot(aes(x = 1/burst_rate)) +\n",
    "    geom_histogram(bins = 30, color = \"white\", fill=\"#39568C\") +\n",
    "    scale_x_log10(limits = c(0.1, 20)) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Time until burst (hour)\") +\n",
    "    ylab(\"Number of features\")\n",
    "tbb\n",
    "ggsave(tbb, filename = \"../plots/Time_between_bursts_Genes.pdf\", width=4, height=4);\n",
    "\n",
    "tbb = bursts %>%\n",
    "    ggplot(aes(x = 1/burst_rate, fill=Type)) +\n",
    "    geom_histogram(bins = 16, color = \"white\") +\n",
    "    scale_x_log10(limits = c(0.1, 20)) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Time until burst (hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "tbb\n",
    "ggsave(tbb, filename = \"../plots/Time_between_bursts_Genes+Enhancers_facet.pdf\", width=8, height=4);\n",
    "\n",
    "tbb = bursts %>%\n",
    "    ggplot(aes(x = 1/burst_rate, fill=Type)) +\n",
    "    geom_histogram(bins = 16, color = \"white\") +\n",
    "    scale_x_log10(limits = c(0.1, 20)) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Time until burst (hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type, ncol = 1) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "tbb\n",
    "ggsave(tbb, filename = \"../plots/Time_between_bursts_Genes+Enhancers_facet_vertical.pdf\", width=4, height=8);\n",
    "\n",
    "bursts %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x = 1/burst_rate, fill = Type)) +\n",
    "    geom_histogram(bins = 16, alpha = 0.5) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(legend.position = c(0.8, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.1, 20)) +\n",
    "    # xlim(0, 10) +\n",
    "    # ggtitle(\"Time until burst (hour) from 10kb gene body\") +\n",
    "    xlab(\"Time until burst (hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\"))\n",
    "\n",
    "bursts %>%\n",
    "    ggplot(aes(x=1/burst_rate, color=Type)) +\n",
    "    theme(legend.position = c(0.2, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=0.7) +\n",
    "    # ggtitle(\"Time until burst (hour) from 10kb gene body\") +\n",
    "    xlab(\"Time until burst (hour)\") +\n",
    "    ylab(\"Number of features\");\n",
    "\n",
    "ks.test(1/burst_rate ~ Type, bursts, alternative = \"greater\", exact=F);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0e901ac1-4d22-474b-82a2-947458839c84",
   "metadata": {},
   "outputs": [],
   "source": [
    "subset = bursts$burst_rate_5k != bursts$burst_rate;\n",
    "txt = lm_eqn( bursts$burst_rate_5k[subset], bursts$burst_rate[subset] );\n",
    "dim(bursts)\n",
    "\n",
    "corr10v5 = bursts %>% filter( subset );\n",
    "dim(corr10v5)\n",
    "\n",
    "corr10v5 = bursts %>%\n",
    "    filter( subset ) %>%\n",
    "    ggplot( aes(x=burst_rate_5k, y=burst_rate) ) +\n",
    "    geom_pointdensity() +\n",
    "    annotate(geom=\"label\", label.size=NA, x=0, y=Inf, hjust=0, vjust=1, label=txt, parse=T, fill=NA) +\n",
    "    scale_x_log10() +\n",
    "    scale_y_log10() +\n",
    "    scale_color_viridis() +\n",
    "    theme(legend.position = c(0.875, 0.175),   legend.key.size = unit(5, 'mm'), \n",
    "          legend.title = element_text(size=0), legend.background = element_blank()) +\n",
    "    xlab(\"Burst frequency (5 kb burst window)\") +\n",
    "    ylab(\"Burst frequency (10 kb burst window)\")\n",
    "corr10v5\n",
    "ggsave(corr10v5, filename = \"../plots/Correlation_10kb_vs_5kb.pdf\", width=4, height=4);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bfc6ff77-cb8d-4722-8957-e7565c97d67a",
   "metadata": {},
   "outputs": [],
   "source": [
    "bs = bursts %>%\n",
    "    ggplot(aes(x = burst_size_5k, fill=Type)) +\n",
    "    geom_histogram(bins = 25, color = \"white\") +\n",
    "    xlim(0, 7.5) +\n",
    "    scale_y_log10() +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 5kb gene body\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    # facet title:\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "bs\n",
    "ggsave(bs, filename = \"../plots/Burst_size_5k_Genes+Enhancers_facet.pdf\", width=8, height=4);\n",
    "\n",
    "bs = bursts %>%\n",
    "    ggplot(aes(x = burst_size_5k, fill = Type)) +\n",
    "    geom_histogram(bins = 15, position = \"identity\", alpha = 0.5) + # \n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(legend.position = c(0.8, 0.8)) +\n",
    "    xlim(0, 7) +\n",
    "    scale_y_log10() +\n",
    "    coord_cartesian(ylim = c(1.5, 10000)) +\n",
    "    # ggtitle(\"Bursts from 5kb gene body\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\"))\n",
    "bs\n",
    "ggsave(bs, filename = \"../plots/Burst_size_5k_Genes+Enhancers.pdf\", width=4, height=4);\n",
    "\n",
    "bursts %>%\n",
    "    ggplot(aes(x=burst_size_5k, color=Type)) +\n",
    "    theme(legend.position = c(0.2, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=0.7) +\n",
    "    # ggtitle(\"Burst size from 5kb gene body\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Number of features\");\n",
    "\n",
    "ks.test(burst_size_5k ~ Type, bursts, alternative = \"greater\", exact=F);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6cdc5486-2b55-47b8-8de1-209949cc75a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "br = bursts %>%\n",
    "    ggplot(aes(x = burst_rate_5k, fill=Type)) +\n",
    "    geom_histogram(bins = 25, color = \"white\") +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    # facet title:\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "br\n",
    "ggsave(br, filename = \"../plots/Burst_rate_5k_Genes+Enhancers_facet.pdf\", width=8, height=4);\n",
    "\n",
    "br = bursts %>%\n",
    "    ggplot(aes(x = burst_rate_5k, fill = Type)) +\n",
    "    geom_histogram(bins = 25, position = \"identity\", alpha = 0.5) + # \n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(legend.position = c(0.2, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.01, 20)) +\n",
    "    # ggtitle(\"Bursts from 5kb gene body\") +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\"))\n",
    "br\n",
    "ggsave(br, filename = \"../plots/Burst_rate_5k_Genes+Enhancers.pdf\", width=4, height=4);\n",
    "\n",
    "bursts %>%\n",
    "    ggplot(aes(x=burst_rate_5k, color=Type)) +\n",
    "    theme(legend.position = c(0.2, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=0.7) +\n",
    "    # ggtitle(\"Bursts from 5kb gene body\") +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Number of features\");\n",
    "\n",
    "ks.test(burst_rate_5k ~ Type, bursts, alternative = \"greater\", exact=F);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4fe825b5-d67a-453b-96ca-3b193a5e2b9c",
   "metadata": {},
   "outputs": [],
   "source": [
    "tbb = bursts %>%\n",
    "    ggplot(aes(x = 1/burst_rate_5k, fill=Type)) +\n",
    "    geom_histogram(bins = 25, position = \"identity\", color = \"white\") +\n",
    "    scale_x_log10(limits = c(0.1, 100)) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    # ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    xlab(\"Time until burst (hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    facet_wrap(~Type) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 0, face=\"bold\"))\n",
    "tbb\n",
    "ggsave(tbb, filename = \"../plots/Time_between_bursts_5k_Genes+Enhancers_facet.pdf\", width=8, height=4);\n",
    "\n",
    "tbb = bursts %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x = 1/burst_rate_5k, fill = Type)) +\n",
    "    geom_histogram(bins = 25, position = \"identity\", alpha = 0.5) +\n",
    "    scale_fill_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    theme(legend.position = c(0.8, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.1, 100)) +\n",
    "    # xlim(0, 10) +\n",
    "    # ggtitle(\"Time until burst (hour) from 5kb gene body\") +\n",
    "    xlab(\"Time until burst (hour)\") +\n",
    "    ylab(\"Number of features\") +\n",
    "    guides(fill=guide_legend(title=\"\"))\n",
    "tbb\n",
    "ggsave(tbb, filename = \"../plots/Time_between_bursts_5k_Genes+Enhancers.pdf\", width=4, height=4);\n",
    "\n",
    "bursts %>%\n",
    "    ggplot(aes(x=1/burst_rate_5k, color=Type)) +\n",
    "    theme(legend.position = c(0.2, 0.8)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual(values=c(\"#39568C\",\"#1F968B\")) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=0.7) +\n",
    "    # ggtitle(\"Time until burst (hour) from 10kb gene body\") +\n",
    "    xlab(\"Time until burst (hour)\") +\n",
    "    ylab(\"Number of features\");\n",
    "\n",
    "ks.test(1/burst_rate_5k ~ Type, bursts, alternative = \"greater\", exact=F);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "587b7b6c-e921-46b5-9035-be41aacf9106",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter burst sizes of more than 1:\n",
    "test = bursts %>%\n",
    "    filter(burst_size > 1)\n",
    "\n",
    "out = test %>%\n",
    "    ggplot(aes(x=burst_rate, y=burst_size)) +\n",
    "    geom_pointdensity() +\n",
    "    scale_color_viridis() +\n",
    "    # annotate(geom=\"label\", label.size=NA, x=-Inf, y=Inf, hjust=0, vjust=1, label=txtG, parse=T, fill=NA) +\n",
    "    # annotate(geom=\"label\", label.size=NA, x=Inf, y=Inf, hjust=0, vjust=1, label=txtE, parse=T, fill=NA) +\n",
    "    # stat_smooth(method = 'lm') +\n",
    "    xlim(0, 10) +\n",
    "    ylim(0, 10) +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Burst size\") +\n",
    "    facet_wrap(~Type) +\n",
    "    theme(legend.position = c(0.875, 0.75),   legend.key.size = unit(5, 'mm'), \n",
    "          legend.title = element_text(size=0), legend.background = element_blank(),\n",
    "          strip.background = element_blank(), strip.text = element_text(size = 14, face=\"bold\"))\n",
    "out\n",
    "ggsave(out, filename = \"../plots/Burst_size_vs_burst_frequency_facet.pdf\", width=8, height=4);\n",
    "\n",
    "txt = lm_eqn(test[test$Type == \"Gene\", ]$burst_rate, test[test$Type == \"Gene\", ]$burst_size)\n",
    "out = test %>%\n",
    "    filter( Type == \"Gene\") %>%\n",
    "    ggplot(aes(x=burst_rate, y=burst_size)) +\n",
    "    geom_pointdensity() +\n",
    "    scale_color_viridis() +\n",
    "    annotate(geom=\"label\", label.size=NA, x=-Inf, y=Inf, hjust=0, vjust=1, label=txt, parse=T, fill=NA) +\n",
    "    # stat_smooth(method = 'lm') +\n",
    "    xlim(0, 10) +\n",
    "    ylim(0, 10) +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Burst size\") +\n",
    "    theme(legend.position = c(0.875, 0.75),   legend.key.size = unit(5, 'mm'), \n",
    "          legend.title = element_text(size=0), legend.background = element_blank())\n",
    "out\n",
    "ggsave(out, filename = \"../plots/Burst_size_vs_burst_frequency_Genes.pdf\", width=4, height=4);\n",
    "\n",
    "txt = lm_eqn(test[test$Type == \"Enhancer\", ]$burst_rate, test[test$Type == \"Enhancer\", ]$burst_size)\n",
    "out = test %>%\n",
    "    filter( Type == \"Enhancer\") %>%\n",
    "    ggplot(aes(x=burst_rate, y=burst_size)) +\n",
    "    geom_pointdensity() +\n",
    "    scale_color_viridis() +\n",
    "    annotate(geom=\"label\", label.size=NA, x=-Inf, y=Inf, hjust=0, vjust=1, label=txt, parse=T, fill=NA) +\n",
    "    # stat_smooth(method = 'lm') +\n",
    "    xlim(0, 10) +\n",
    "    ylim(0, 10) +\n",
    "    xlab(\"Burst frequency (per hour)\") +\n",
    "    ylab(\"Burst size\") +\n",
    "    theme(legend.position = c(0.875, 0.75),   legend.key.size = unit(5, 'mm'), \n",
    "          legend.title = element_text(size=0), legend.background = element_blank())\n",
    "out\n",
    "ggsave(out, filename = \"../plots/Burst_size_vs_burst_frequency_Enhancers.pdf\", width=4, height=4);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e51c8da-eed7-45c3-b5c2-5ab24455987d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# THIS LENGTH EFFECT ON BURST KINETICS IS ONLY WHEN GENES ARE NOT SELECTED FOR 10 KB:\n",
    "txt = lm_eqn(bursts[bursts$Type == \"Gene\", ]$burst_rate, bursts[bursts$Type == \"Gene\", ]$length)\n",
    "out = bursts %>%\n",
    "    filter( Type == \"Gene\") %>%\n",
    "    ggplot(aes(x=length, y=burst_rate)) +\n",
    "    # ggplot(aes(x=length, y=1+(N_reads/N_bursts-1)/capture_efficiency)) +\n",
    "    geom_pointdensity() +\n",
    "    scale_color_viridis() +\n",
    "    annotate(geom=\"label\", label.size=NA, x=-Inf, y=Inf, hjust=0, vjust=1, label=txt, parse=T, fill=NA) +\n",
    "    # stat_smooth(method = 'lm') +\n",
    "    # scale_x_log10(limits = c(1000, 30000)) +\n",
    "    xlim(0, maxGeneLengthAfterTrim) +\n",
    "    ylim(0, 40) +\n",
    "    xlab(\"Gene length\") +\n",
    "    ylab(\"Burst frequency (per hour)\") +\n",
    "    theme(legend.position = c(0.875, 0.75),    legend.key.size = unit(5, 'mm'), \n",
    "          legend.title = element_text(size=0), legend.background = element_blank())\n",
    "    # facet_wrap(~Type) +\n",
    "    # theme(strip.background = element_blank(),\n",
    "    #     strip.text = element_text(size = 14, face=\"bold\"))\n",
    "out\n",
    "# ggsave(out, filename = \"../plots/BurstFrequency_vs_geneLength_Genes.pdf\", width=4, height=4); # _whenGeneNotLimitedTo10kb\n",
    "\n",
    "txt = lm_eqn(bursts[bursts$Type == \"Gene\", ]$burst_size, bursts[bursts$Type == \"Gene\", ]$length)\n",
    "out = bursts %>%\n",
    "    filter( Type == \"Gene\") %>%\n",
    "    ggplot(aes(x=length, y=burst_size)) +\n",
    "    geom_pointdensity() +\n",
    "    scale_color_viridis() +\n",
    "    annotate(geom=\"label\", label.size=NA, x=-Inf, y=Inf, hjust=0, vjust=1, label=txt, parse=T, fill=NA) +\n",
    "    # stat_smooth(method = 'lm') +\n",
    "    # scale_x_log10(limits = c(1000, 30000)) +\n",
    "    xlim(0, maxGeneLengthAfterTrim) +\n",
    "    ylim(0, 8) +\n",
    "    xlab(\"Gene length\") +\n",
    "    ylab(\"Burst size\") +\n",
    "    theme(legend.position = c(0.875, 0.75),    legend.key.size = unit(5, 'mm'), \n",
    "          legend.title = element_text(size=0), legend.background = element_blank())\n",
    "    # facet_wrap(~Type) +\n",
    "    # theme(strip.background = element_blank(),\n",
    "    #     strip.text = element_text(size = 14, face=\"bold\"))\n",
    "out\n",
    "# ggsave(out, filename = \"../plots/BurstSize_vs_geneLength_Genes.pdf\", width=4, height=4); # _whenGeneNotLimitedTo10kb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b513218d-9076-48f9-990f-bd387ff4ccba",
   "metadata": {},
   "outputs": [],
   "source": [
    "bursts %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=Type, y = burst_rate, fill = Type)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.25, 0.5, 0.75)) +\n",
    "    #geom_jitter(height = 0, width = 0.4, alpha=0.1, size=0.5) +\n",
    "    scale_y_log10(limits = c(0.05, 5)) +\n",
    "    ggtitle(\"Bursts from 10kb gene body\") +\n",
    "    ylab(\"Bursts per hour\") +\n",
    "    guides(fill=guide_legend(title=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01b77ea9-ba2f-4937-8a5b-2c8de34ba156",
   "metadata": {},
   "outputs": [],
   "source": [
    "bursts %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=burst_rate_5k, y=burst_size_5k)) +\n",
    "    geom_pointdensity(show.legend = F) +\n",
    "    scale_color_viridis() +\n",
    "    annotate(geom = \"label\", x = 6, y = 11, label = txt, parse=T) +\n",
    "    #stat_smooth(method = 'lm') +\n",
    "    xlim(0, 10) +\n",
    "    ylim(0, 12) +\n",
    "    xlab(\"Estimated burst rate from 5kb \\n of gene body (per hour)\") +\n",
    "    ylab(\"Estimated burst size\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c108355-d89a-4978-90fb-2c2470c52984",
   "metadata": {},
   "outputs": [],
   "source": [
    "txt = lm_eqn( bursts$N_reads, bursts$burst_rate );\n",
    "\n",
    "bursts %>%\n",
    "    filter(Type == \"Gene\") %>%\n",
    "    ggplot( aes(x=burst_rate, y=N_reads) ) +\n",
    "    geom_pointdensity(show.legend = F) +\n",
    "    annotate(geom = \"label\", x = 0.3, y = 3e4, label = txt, parse=T) +\n",
    "    scale_x_log10() +\n",
    "    scale_y_log10() +\n",
    "    scale_color_viridis() +\n",
    "    xlab(\"Burst frequency from 10kbp gene body\") +\n",
    "    ylab(\"Number of reads in gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eae33782-9b01-49c7-88c9-0b0199405776",
   "metadata": {},
   "outputs": [],
   "source": [
    "txt = lm_eqn( bursts$N_bursts, bursts$N_reads );\n",
    "\n",
    "bursts %>%\n",
    "    filter(Type == \"Gene\") %>%\n",
    "    ggplot( aes(x=N_bursts, y=N_reads) ) +\n",
    "    geom_pointdensity(show.legend = F) +\n",
    "    annotate(geom = \"label\", x = 0.3, y = 3e4, label = txt, parse=T) +\n",
    "    xlim(10, 200) +\n",
    "    ylim(10, 200) +\n",
    "    scale_color_viridis() +\n",
    "    xlab(\"Number of cells transcribing\") +\n",
    "    ylab(\"Number of reads in gene\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20669576-8098-44c1-ae4d-6007424e76a9",
   "metadata": {},
   "outputs": [],
   "source": [
    "txt = lm_eqn( bursts$N_reads, bursts$burst_rate );\n",
    "\n",
    "bursts %>%\n",
    "    filter(Type == \"Enhancer\") %>%\n",
    "    ggplot( aes(x=burst_rate, y=N_reads) ) +\n",
    "    geom_pointdensity(show.legend = F) +\n",
    "    annotate(geom = \"label\", x = 0.3, y = 3e4, label = txt, parse=T) +\n",
    "    scale_x_log10() +\n",
    "    scale_y_log10() +\n",
    "    scale_color_viridis() +\n",
    "    xlab(\"Burst frequency\") +\n",
    "    ylab(\"Number of reads in enhancer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2490c227-bac7-4bc9-bade-6faf0238c456",
   "metadata": {},
   "outputs": [],
   "source": [
    "bursts %>%\n",
    "    ggplot( aes(x=burst_rate, y=N_bursts) ) +\n",
    "    geom_pointdensity(show.legend = F) +\n",
    "    scale_x_log10() +\n",
    "    scale_y_log10() +\n",
    "    scale_color_viridis() +\n",
    "    xlab(\"Burst frequency from 10kbp gene body\") +\n",
    "    ylab(\"Number of bursts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b027cc96-52ae-4f6a-a101-612de7a527a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "bursts %>%\n",
    "    filter(Type == \"Gene\") %>%\n",
    "    ggplot( aes(x=burst_rate, y=N_bursts) ) +\n",
    "    geom_pointdensity(show.legend = F) +\n",
    "    ylim(0, 1000) +\n",
    "    scale_x_log10() +\n",
    "    #scale_y_log10(limits=c(0.1, 100)) +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(\"Genes, 10kbp\") +\n",
    "    xlab(\"Bursts per hour\") +\n",
    "    ylab(\"Number of bursts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4eb3951-afde-4435-9862-67b5370b2c6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "bursts %>%\n",
    "    filter(Type == \"Enhancer\") %>%\n",
    "    ggplot( aes(x=burst_rate, y=N_bursts) ) +\n",
    "    geom_pointdensity(show.legend = F) +\n",
    "    scale_x_log10() +\n",
    "    ylim(0, 1000) +\n",
    "    #scale_y_log10(limits=c(0.1, 100)) +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(\"Enhancers\") +\n",
    "    xlab(\"Bursts per hour\") +\n",
    "    ylab(\"Number of bursting cells\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b174936d-7edc-49b8-a93c-8646ea3c69dd",
   "metadata": {},
   "source": [
    "## Promoter elements & burst frequencies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cbf7d43-2ee1-4781-8545-c1c40781552c",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata = read_bed(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/mm10_genes_with_TATA-box_in_promoters.bed\");\n",
    "names(tata) = paste0(\"GN-\", sub(\"_1\", \"\", tata$name));\n",
    "tata\n",
    "\"TATA\"\n",
    "length(tata)\n",
    "\n",
    "notata = read_bed(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/mm10_genes_withOUT_TATA-box_in_promoters.bed\");\n",
    "names(notata) = paste0(\"GN-\", sub(\"_1\", \"\", notata$name));\n",
    "\"No TATA\"\n",
    "length(notata)\n",
    "\n",
    "inr = read_bed(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/mm10_genes_with_Inr_in_promoters.bed\");\n",
    "names(inr) = paste0(\"GN-\", sub(\"_1\", \"\", inr$name));\n",
    "\"Inr\"\n",
    "length(inr)\n",
    "\n",
    "noinr = read_bed(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/mm10_genes_withOUT_Inr_in_promoters.bed\");\n",
    "names(noinr) = paste0(\"GN-\", sub(\"_1\", \"\", noinr$name));\n",
    "\"No Inr\"\n",
    "length(noinr)\n",
    "\n",
    "gc = read_bed(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/mm10_genes_with_GC-box_in_promoters.bed\");\n",
    "names(gc) = paste0(\"GN-\", sub(\"_1\", \"\", gc$name));\n",
    "\"GC\"\n",
    "length(gc)\n",
    "\n",
    "nogc = read_bed(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/mm10_genes_withOUT_GC-box_in_promoters.bed\");\n",
    "names(nogc) = paste0(\"GN-\", sub(\"_1\", \"\", nogc$name));\n",
    "\"No GC\"\n",
    "length(nogc)\n",
    "\n",
    "tatainr = read_bed(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/mm10_genes_with_TATA-box-AND-Inr_in_promoters.bed\");\n",
    "names(tatainr) = paste0(\"GN-\", sub(\"_1\", \"\", tatainr$name));\n",
    "\"TATA + Inr\"\n",
    "length(tatainr)\n",
    "\n",
    "notatainr = read_bed(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/mm10_genes_withOUT_TATA-box-AND-Inr_in_promoters.bed\");\n",
    "names(notatainr) = paste0(\"GN-\", sub(\"_1\", \"\", notatainr$name));\n",
    "\"No TATA + No Inr\"\n",
    "length(notatainr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a91f8d1-78a9-43a0-aec3-4b869365e4bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "PauseButton = readRDS(\"~/group/mES_PROseq_RNAseq_ATACseq/scRNAseq_mES_Larrson_Sandberg_2019/CoPRO_MEF_NonHS_PauseButtons.rds\") %>%\n",
    "    filter( PauseButton );\n",
    "PauseButton\n",
    "\n",
    "# define promoter regions of features\n",
    "gene_promoters = promoters(genes, upstream = 100, downstream = 400 );\n",
    "gene_promoters\n",
    "# active_promoters = findOverlaps( gene_promoters, pb);\n",
    "# active_promoters\n",
    "pb = subsetByOverlaps( gene_promoters, PauseButton, ignore.strand=F);\n",
    "pb\n",
    "nopb = subsetByOverlaps( gene_promoters, PauseButton, ignore.strand=F, invert=T );\n",
    "nopb\n",
    "# # overlap with dREG to identify which promoters are active\n",
    "# hasActivePro = queryHits(active_promoters) %>% unique();\n",
    "# active_isoforms = ref_transcripts[ hasActivePro ];\n",
    "# active_isoforms;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b53560ab-1c3e-44fd-89b4-edea68d86e8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "length(features[names(features) %in% names(tata)])\n",
    "summary(width(features[names(features) %in% names(tata)]))\n",
    "length(features[names(features) %in% names(notata)])\n",
    "summary(width(features[names(features) %in% names(notata)]))\n",
    "\n",
    "length(features[names(features) %in% names(tatainr)])\n",
    "summary(width(features[names(features) %in% names(tatainr)]))\n",
    "length(features[names(features) %in% names(notatainr)])\n",
    "summary(width(features[names(features) %in% names(notatainr)]))\n",
    "\n",
    "length(features[names(features) %in% names(pb)])\n",
    "summary(width(features[names(features) %in% names(pb)]))\n",
    "length(features[names(features) %in% names(nopb)])\n",
    "summary(width(features[names(features) %in% names(nopb)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e095a39-db78-4503-b138-64e8f4031429",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tata = allBursts[allBursts$feature %in% names(tata),];\n",
    "# dim(tata)\n",
    "# notata = allBursts[allBursts$feature %in% names(notata),];\n",
    "# dim(notata)\n",
    "# inr = allBursts[allBursts$feature %in% names(inr),];\n",
    "# dim(inr)\n",
    "# noinr = allBursts[allBursts$feature %in% names(noinr),];\n",
    "# dim(noinr)\n",
    "# gc = allBursts[allBursts$feature %in% names(gc),];\n",
    "# dim(gc)\n",
    "# nogc = allBursts[allBursts$feature %in% names(nogc),];\n",
    "# dim(nogc)\n",
    "# tatainr = allBursts[allBursts$feature %in% names(tatainr),];\n",
    "# dim(tatainr)\n",
    "# notatainr = allBursts[allBursts$feature %in% names(notatainr),];\n",
    "# dim(notatainr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce5c88d3-0766-4241-bf2d-951e08cba301",
   "metadata": {},
   "outputs": [],
   "source": [
    "burstsWpElement = bursts %>%\n",
    "    # mutate( pElement = ifelse(feature %in% names(tata), \"TATA\", \"NA\") );\n",
    "    mutate( GC = case_when(\n",
    "        feature %in% names(gc) ~ \"GC\",\n",
    "        feature %in% names(nogc) ~ \"No_GC\",\n",
    "         TRUE ~ NA_character_)\n",
    "    ) %>%\n",
    "    mutate( TATA = case_when(\n",
    "        feature %in% names(tata) ~ \"TATA\",\n",
    "        feature %in% names(notata) ~ \"No_TATA\",\n",
    "        TRUE ~ NA_character_)\n",
    "    ) %>%\n",
    "    mutate( Inr = case_when(\n",
    "        feature %in% names(inr) ~ \"Inr\",\n",
    "        feature %in% names(noinr) ~ \"No_Inr\",\n",
    "        TRUE ~ NA_character_)\n",
    "    ) %>%\n",
    "    mutate( TATA_Inr = case_when(\n",
    "        feature %in% names(tatainr) ~ \"TATA_Inr\",\n",
    "        feature %in% names(notatainr) ~ \"No_TATA_Inr\",\n",
    "        TRUE ~ NA_character_)  \n",
    "    ) %>%\n",
    "    mutate( PB = case_when(\n",
    "        feature %in% names(pb) ~ \"PB\",\n",
    "        feature %in% names(nopb) ~ \"No_PB\",\n",
    "        TRUE ~ NA_character_)\n",
    "    );\n",
    "burstsWpElement[1:5,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "601fbad2-ddff-4867-9f45-b2879d583308",
   "metadata": {},
   "outputs": [],
   "source": [
    "myorder = c(\"TATA\", \"No_TATA\", \"Inr\", \"No_Inr\", \"TATA_Inr\", \"No_TATA_Inr\", \"PB\", \"No_PB\");\n",
    "allElements = burstsWpElement %>%\n",
    "    pivot_longer(TATA:PB, names_to=\"Source\", values_to=\"Group\") %>%\n",
    "    filter(!is.na(Group)) %>%\n",
    "    mutate(Group=factor(Group, levels=myorder))\n",
    "allElements[1:5,]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d62d54c-6430-486f-96ac-fed20125e150",
   "metadata": {},
   "outputs": [],
   "source": [
    "tataenrich = allElements %>%\n",
    "    filter(burst_size > 1) %>%\n",
    "    ggplot(aes(x=Group, y = burst_size, fill = Group)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    # scale_fill_brewer(palette = \"Set3\") +\n",
    "    # theme(axis.text.x = element_text(angle=30, hjust = 1)) +\n",
    "    scale_fill_manual(values=c(\"#3e6790\",\"#6485a6\", \"#646673\", \"#868892\", \"#b0863d\", \"#bf9e63\", \"#2b6464\", \"#558383\")) +\n",
    "    scale_y_log10(limits = c(1, 4)) +\n",
    "    # ggtitle(\"Role of promoter elements in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Bursts size\");\n",
    "tataenrich\n",
    "ggsave(tataenrich, filename = \"../plots/Promoter_elements_burst_size.pdf\", width=6.5, height=4);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb480dd9-1f0c-4e09-90bd-9216002d9cad",
   "metadata": {},
   "outputs": [],
   "source": [
    "# t-test tests mean:\n",
    "t.test(burst_size ~ Group, allElements, subset = Group %in% c(\"TATA\", \"No_TATA\"));\n",
    "t.test(burst_size ~ Group, allElements, subset = Group %in% c(\"Inr\", \"No_Inr\"));\n",
    "t.test(burst_size ~ Group, allElements, subset = Group %in% c(\"TATA_Inr\", \"No_TATA_Inr\"));\n",
    "t.test(burst_size ~ Group, allElements, subset = Group %in% c(\"PB\", \"No_PB\"));\n",
    "\n",
    "# f-test for difference in variance:\n",
    "var.test(burst_size ~ Group, allElements, subset = Group %in% c(\"TATA\", \"No_TATA\"));\n",
    "var.test(burst_size ~ Group, allElements, subset = Group %in% c(\"Inr\", \"No_Inr\"));\n",
    "var.test(burst_size ~ Group, allElements, subset = Group %in% c(\"TATA_Inr\", \"No_TATA_Inr\"));\n",
    "var.test(burst_size ~ Group, allElements, subset = Group %in% c(\"PB\", \"No_PB\"));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3c7c1a5-fb0f-493c-a432-e8c0f2424c9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "tataenrich = allElements %>%\n",
    "    # filter(burst_size > 1) %>%\n",
    "    ggplot(aes(x=Group, y = burst_rate, fill = Group)) +\n",
    "    # geom_jitter(height = 0, width = 0.4, alpha=0.1, size=0.5, shape=21, color=\"grey\") +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    # scale_fill_brewer(palette = \"Set3\") +\n",
    "    # theme(axis.text.x = element_text(angle=30, hjust = 1)) +\n",
    "    # scale_fill_manual(values=c(\"#3e6790\",\"#6485a6\", \"#646673\", \"#868892\", \"#2b6464\", \"#558383\",\"#a38071\", \"#d6b3a4\",  \"#b0863d\", \"#bf9e63\")) +\n",
    "    scale_fill_manual(values=c(\"#3e6790\",\"#6485a6\", \"#646673\", \"#868892\", \"#b0863d\", \"#bf9e63\", \"#2b6464\", \"#558383\")) +\n",
    "    scale_y_log10(limits = c(0.05, 10)) +\n",
    "    # ggtitle(\"Role of promoter elements in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Burst frequency (per hour)\");\n",
    "tataenrich\n",
    "ggsave(tataenrich, filename = \"../plots/Promoter_elements_burst_frequency.pdf\", width=6.5, height=4);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d4eea69-fbae-4698-9971-beb36d1a3aec",
   "metadata": {},
   "outputs": [],
   "source": [
    "ks.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"TATA\", \"No_TATA\"), alternative = \"greater\", exact=F);\n",
    "wilcox.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"TATA\", \"No_TATA\"));\n",
    "\n",
    "ks.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"Inr\", \"No_Inr\"), alternative = \"greater\", exact=F);\n",
    "wilcox.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"Inr\", \"No_Inr\"));\n",
    "\n",
    "ks.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"TATA_Inr\", \"No_TATA_Inr\"), alternative = \"greater\", exact=F);\n",
    "wilcox.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"TATA_Inr\", \"No_TATA_Inr\"));\n",
    "\n",
    "ks.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"PB\", \"No_PB\"), alternative = \"greater\", exact=F);\n",
    "wilcox.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"PB\", \"No_PB\"));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc4b8686-3fa9-4ea3-8ba4-78698b0c9302",
   "metadata": {},
   "outputs": [],
   "source": [
    "# t-test tests mean:\n",
    "t.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"TATA\", \"No_TATA\"));\n",
    "t.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"Inr\", \"No_Inr\"));\n",
    "t.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"TATA_Inr\", \"No_TATA_Inr\"));\n",
    "t.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"PB\", \"No_PB\"));\n",
    "\n",
    "# f-test for difference in variance:\n",
    "var.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"TATA\", \"No_TATA\"));\n",
    "var.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"Inr\", \"No_Inr\"));\n",
    "var.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"TATA_Inr\", \"No_TATA_Inr\"));\n",
    "var.test(burst_rate ~ Group, allElements, subset = Group %in% c(\"PB\", \"No_PB\"));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7fd15215-eb79-4b85-885e-55fc8701e2a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "tataenrich = allElements %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=Group, y = 1/burst_rate, fill = Group)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    # scale_fill_brewer(palette = \"Set3\") +\n",
    "    # theme(axis.text.x = element_text(angle=30, hjust = 1)) +\n",
    "    scale_fill_manual(values=c(\"#3e6790\",\"#6485a6\", \"#646673\", \"#868892\", \"#b0863d\", \"#bf9e63\", \"#2b6464\", \"#558383\")) +\n",
    "    scale_y_log10(limits = c(0.1, 100)) +\n",
    "    # ggtitle(\"Role of promoter elements in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Time between bursts\");\n",
    "tataenrich\n",
    "ggsave(tataenrich, filename = \"../plots/Promoter_elements_burst_interval.pdf\", width=6.5, height=4);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2add7ea7-dcdf-4384-939c-6a1bbdc2d7d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "ks.test(burst_size ~ Group, allElements, subset = Group %in% c(\"TATA\", \"No_TATA\"), alternative = \"greater\", exact=F);\n",
    "wilcox.test(burst_size ~ Group, allElements, subset = Group %in% c(\"TATA\", \"No_TATA\"));\n",
    "\n",
    "ks.test(burst_size ~ Group, allElements, subset = Group %in% c(\"Inr\", \"No_Inr\"), alternative = \"greater\", exact=F);\n",
    "wilcox.test(burst_size ~ Group, allElements, subset = Group %in% c(\"Inr\", \"No_Inr\"));\n",
    "\n",
    "ks.test(burst_size ~ Group, allElements, subset = Group %in% c(\"TATA_Inr\", \"No_TATA_Inr\"), alternative = \"greater\", exact=F);\n",
    "wilcox.test(burst_size ~ Group, allElements, subset = Group %in% c(\"TATA_Inr\", \"No_TATA_Inr\"));\n",
    "\n",
    "ks.test(burst_size ~ Group, allElements, subset = Group %in% c(\"PB\", \"No_PB\"), alternative = \"greater\", exact=F);\n",
    "wilcox.test(burst_size ~ Group, allElements, subset = Group %in% c(\"PB\", \"No_PB\"));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a1aec72-4a0d-45fe-8bb9-3202a3826964",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_df = burstsWpElement[burstsWpElement$TATA %in% c(\"TATA\", \"No_TATA\"),];\n",
    "ptest = ks.test(burst_rate ~ TATA, tata_df, alternative = \"less\", exact=F);\n",
    "ptest\n",
    "tataenrich = tata_df %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=TATA, y = burst_rate, fill = TATA)) +\n",
    "    # geom_jitter(\n",
    "        # height = 0, width = 0.4, alpha=0.1, size=0.5, \n",
    "        # shape=21, color=\"grey\", position=position_jitter(0.2)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(0.01, 10)) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Bursts per hour\");\n",
    "\n",
    "tataenrich\n",
    "ggsave(tataenrich, filename = \"../plots/burstFreq_TATA.pdf\", width=4, height=4);\n",
    "\n",
    "tataenrich = tata_df %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=TATA, y = 1/burst_rate, fill = TATA)) +\n",
    "    # geom_jitter(\n",
    "        # height = 0, width = 0.4, alpha=0.1, size=0.5, \n",
    "        # shape=21, color=\"grey\", position=position_jitter(0.2)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(0.1, 100)) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Time until burst (hour)\");\n",
    "tataenrich\n",
    "ggsave(tataenrich, filename = \"../plots/TimeBetweenBursts_TATA.pdf\", width=4, height=4);\n",
    "\n",
    "tata_df %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=burst_rate, color=TATA)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"Bursts per hour\") +\n",
    "    ylab(\"Fraction of genes\");\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7b3a999-27e3-4320-a344-cd5e362c1833",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_df = burstsWpElement[burstsWpElement$TATA %in% c(\"TATA\", \"No_TATA\"),];\n",
    "ptest = ks.test(burst_rate_5k ~ TATA, tata_df, alternative = \"less\", exact=F);\n",
    "ptest\n",
    "tata_df %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=TATA, y = burst_rate_5k, fill = TATA)) +\n",
    "    # geom_jitter(\n",
    "        # height = 0, width = 0.4, alpha=0.1, size=0.5, \n",
    "        # shape=21, color=\"grey\", position=position_jitter(0.2)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(0.01, 10)) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Bursts per hour from 5K\");\n",
    "\n",
    "tata_df %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=burst_rate_5k, color=TATA)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"Bursts per hour from 5k\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f2389dc2-0410-4251-bca2-4696a60142ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptest = ks.test(burst_size ~ TATA, tata_df, alternative = \"less\", exact=F);\n",
    "ptest\n",
    "tataenrich = tata_df %>%\n",
    "    filter(burst_size > 1) %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=TATA, y = burst_size, fill = TATA)) +\n",
    "    # geom_jitter(\n",
    "    #     height = 0, width = 0.4, alpha=0.1, size=0.5, shape=21, color=\"grey\") +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(1, 5)) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Burst size\");\n",
    "tataenrich\n",
    "ggsave(tataenrich, filename = \"../plots/BurstSize_TATA.pdf\", width=4, height=4);\n",
    "\n",
    "tata_df %>%\n",
    "    filter(burst_size > 1) %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=burst_size, color=TATA)) +\n",
    "    scale_x_log10(limits = c(1, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82cfa47c-163d-4af7-9da7-ec534f62e50a",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptest = ks.test(burst_size_5k ~ TATA, tata_df, alternative = \"less\", exact=F);\n",
    "ptest\n",
    "tata_df %>%\n",
    "    filter(burst_size_5k > 1) %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=TATA, y = burst_size_5k, fill = TATA)) +\n",
    "    geom_jitter(\n",
    "        height = 0, width = 0.4, alpha=0.1, size=0.5, shape=21, color=\"grey\") +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(1, 5)) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Burst size from 5k\");\n",
    "\n",
    "tata_df %>%\n",
    "    filter(burst_size_5k > 1) %>%\n",
    "    # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "    ggplot(aes(x=burst_size_5k, color=TATA)) +\n",
    "    scale_x_log10(limits = c(1, 5)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of TATA element in Bursts\") +\n",
    "    xlab(\"Burst size from 5k\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93f6caab-fd3f-45f4-90d6-330990df7ff2",
   "metadata": {},
   "outputs": [],
   "source": [
    "ptest = ks.test(length ~ TATA, tata_df, alternative = \"less\", exact=F);\n",
    "ptest\n",
    "tata_df %>%\n",
    "# filter(N_bursts >= 5 & N_cells_on >= 10) %>%\n",
    "    ggplot(aes(x=TATA, y = length, fill = TATA)) +\n",
    "    # geom_jitter(\n",
    "        # height = 0, width = 0.4, alpha=0.1, size=0.5, \n",
    "        # shape=21, color=\"grey\", position=position_jitter(0.2)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(1, maxGeneLengthAfterTrim)) +\n",
    "    ggtitle(\"Gene length & TATA\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Length of gene\");\n",
    "\n",
    "tata_df %>%\n",
    "    ggplot(aes(x=length, color=TATA)) +\n",
    "    scale_x_log10(limits = c(1, maxGeneLengthAfterTrim)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Gene length & TATA\") +\n",
    "    xlab(\"Gene length\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa51d070-83a2-48ba-ac20-a53d0cee75fc",
   "metadata": {},
   "outputs": [],
   "source": [
    "inr_df = burstsWpElement[burstsWpElement$Inr %in% c(\"Inr\", \"No_Inr\"),];\n",
    "ptest = ks.test(burst_rate ~ Inr, inr_df, alternative = \"greater\", exact=F);\n",
    "ptest\n",
    "tataenrich = inr_df %>%\n",
    "    # filter(N_bursts >= 5 & N_cells_on >= 10) %>%\n",
    "    ggplot(aes(x=Inr, y = burst_rate, fill = Inr)) +\n",
    "    # geom_jitter(\n",
    "        # height = 0, width = 0.4, alpha=0.1, size=0.5, \n",
    "        # shape=21, color=\"grey\", position=position_jitter(0.2)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(0.01, 10)) +\n",
    "    ggtitle(\"Role of Inr element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Bursts per hour\");\n",
    "tataenrich\n",
    "ggsave(tataenrich, filename = \"../plots/BurstFrequency_Inr.pdf\", width=4, height=4);\n",
    "\n",
    "inr_df %>%\n",
    "    ggplot(aes(x=burst_rate, color=Inr)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of Inr element in Bursts\") +\n",
    "    xlab(\"Bursts per hour\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8f2ab762-20d2-45d0-84d3-31bcb3c79a9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_df = burstsWpElement[burstsWpElement$GC %in% c(\"GC\", \"No_GC\"),];\n",
    "ptest = ks.test(burst_rate ~ GC, gc_df, alternative = \"greater\", exact=F);\n",
    "ptest\n",
    "gc_df %>%\n",
    "    # filter(N_bursts >= 5 & N_cells_on >= 10) %>%\n",
    "    ggplot(aes(x=GC, y = burst_rate, fill = GC)) +\n",
    "    # geom_jitter(\n",
    "        # height = 0, width = 0.4, alpha=0.1, size=0.5, \n",
    "        # shape=21, color=\"grey\", position=position_jitter(0.2)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(0.01, 10)) +\n",
    "    ggtitle(\"Role of GC element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Bursts per hour\");\n",
    "\n",
    "gc_df %>%\n",
    "    ggplot(aes(x=burst_rate, color=GC)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of GC element in Bursts\") +\n",
    "    xlab(\"Bursts per hour\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d32d4f22-4a18-4714-912a-e05a6a7de7f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_inr_df = burstsWpElement[burstsWpElement$TATA_Inr %in% c(\"TATA_Inr\", \"No_TATA_Inr\"),];\n",
    "ptest = ks.test(burst_rate ~ TATA_Inr, tata_inr_df, alternative = \"less\", exact=F);\n",
    "ptest\n",
    "tatainrenrich = tata_inr_df %>%\n",
    "    # filter(N_bursts >= 5 & N_cells_on >= 10) %>%\n",
    "    ggplot(aes(x=TATA_Inr, y = burst_rate, fill = TATA_Inr)) +\n",
    "    # geom_jitter(\n",
    "        # height = 0, width = 0.4, alpha=0.1, size=0.5, \n",
    "        # shape=21, color=\"grey\", position=position_jitter(0.2)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(0.01, 10)) +\n",
    "    ggtitle(\"Role of TATA+Inr element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Bursts per hour\");\n",
    "tatainrenrich\n",
    "ggsave(tatainrenrich, filename = \"../plots/BurstFrequency_TATA+Inr.pdf\", width=4, height=4);\n",
    "\n",
    "tatainrenrich = tata_inr_df %>%\n",
    "    # filter(N_bursts >= 5 & N_cells_on >= 10) %>%\n",
    "    ggplot(aes(x=TATA_Inr, y = 1/burst_rate, fill = TATA_Inr)) +\n",
    "    # geom_jitter(\n",
    "        # height = 0, width = 0.4, alpha=0.1, size=0.5, \n",
    "        # shape=21, color=\"grey\", position=position_jitter(0.2)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    scale_fill_manual(\n",
    "        values=c(\"#39568CFF\",\"#1F968BFF\"),\n",
    "        name=\"Promoter\\nStatus\") +\n",
    "    scale_y_log10(limits = c(0.1, 100)) +\n",
    "    ggtitle(\"Role of TATA+Inr element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"time until burst\");\n",
    "tatainrenrich\n",
    "ggsave(tatainrenrich, filename = \"../plots/TimeBetweenBursts_TATA+Inr.pdf\", width=4, height=4);\n",
    "\n",
    "tata_inr_df %>%\n",
    "    ggplot(aes(x=burst_rate, color=TATA_Inr)) +\n",
    "    scale_x_log10(limits = c(0.01, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of TATA+Inr element in Bursts\") +\n",
    "    xlab(\"Bursts per hour\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b931d65-6105-4204-85bd-931a35b4bd2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # r function to cbind various length columns:\n",
    "# cbind.fill <- function(...){\n",
    "#     nm <- list(...)\n",
    "#     nm <- lapply(nm, as.matrix)\n",
    "#     n <- max(sapply(nm, nrow))\n",
    "#     do.call(cbind, lapply(nm, function (x)\n",
    "#     rbind(x, matrix(, n-nrow(x), ncol(x)))))\n",
    "# }\n",
    "# comb = cbind.fill(tata$burst_rate, notata$burst_rate, tatainr$burst_rate, notatainr$burst_rate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c410b0ca-4907-44b4-a46e-eac88403b7d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_inr_df = burstsWpElement[burstsWpElement$TATA_Inr %in% c(\"TATA_Inr\", \"No_TATA_Inr\"),];\n",
    "ptest = ks.test(burst_size ~ TATA_Inr, tata_inr_df, alternative = \"greater\", exact=F);\n",
    "ptest\n",
    "\n",
    "tatainrenrich = tata_inr_df %>%\n",
    "    filter(burst_size > 1) %>%\n",
    "    ggplot(aes(x=TATA_Inr, y = burst_size, fill = TATA_Inr)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    # geom_jitter(\n",
    "    #     height = 0, width = 0.4, alpha=0.1, size=0.5, shape=21, color=\"red\") +\n",
    "    scale_fill_manual(values=c(\"#39568CFF\",\"#1F968BFF\"), name=\"Promoter\\nStatus\") +\n",
    "    # geom_text(data = median, aes(x = TATA_Inr, y = med, label = med), size = 3, vjust = -1.5) +\n",
    "    scale_y_log10(limits = c(1, 5)) +\n",
    "    ggtitle(\"Role of TATA+Inr element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Burst size\");\n",
    "tatainrenrich\n",
    "ggsave(tatainrenrich, filename = \"../plots/BurstSize_TATA+Inr.pdf\", width=4, height=4);\n",
    "\n",
    "tata_inr_df %>%\n",
    "    filter(burst_size > 1) %>%\n",
    "    ggplot(aes(x=burst_size, color=TATA_Inr)) +\n",
    "    scale_x_log10(limits = c(1, 5)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of TATA+Inr element in Bursts\") +\n",
    "    xlab(\"Burst size\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "505deee4-5bf9-4987-963b-2aecc8cd34a2",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataMedian =  summarise(group_by(tata_inr_df, TATA_Inr), MD = mean(round(burst_size, digits=2)))\n",
    "dataMedian"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ada9fd68-ff1b-4358-9f87-fe9dc13d37c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "tata_inr_df = burstsWpElement[burstsWpElement$TATA_Inr %in% c(\"TATA_Inr\", \"No_TATA_Inr\"),];\n",
    "ptest = ks.test(meanExpPerKb ~ TATA_Inr, tata_inr_df, alternative = \"greater\", exact=F);\n",
    "ptest\n",
    "\n",
    "tata_inr_df %>%\n",
    "    # filter(burst_size > 1) %>%\n",
    "    ggplot(aes(x=TATA_Inr, y = meanExpPerKb, fill = TATA_Inr)) +\n",
    "    geom_violin(show.legend = F, draw_quantiles = c(0.5)) +\n",
    "    # geom_jitter(\n",
    "    #     height = 0, width = 0.4, alpha=0.1, size=0.5, shape=21, color=\"red\") +\n",
    "    scale_fill_manual(values=c(\"#39568CFF\",\"#1F968BFF\"), name=\"Promoter\\nStatus\") +\n",
    "    # geom_text(data = median, aes(x = TATA_Inr, y = med, label = med), size = 3, vjust = -1.5) +\n",
    "    scale_y_log10(limits = c(0.001, 1)) +\n",
    "    ggtitle(\"Role of TATA+Inr element in Bursts\") +\n",
    "    xlab(\"\") +\n",
    "    ylab(\"Mean Expression\");\n",
    "\n",
    "tata_inr_df %>%\n",
    "    # filter(burst_size > 1) %>%\n",
    "    ggplot(aes(x=meanExpPerKb, color=TATA_Inr)) +\n",
    "    scale_x_log10(limits = c(1, 10)) +\n",
    "    guides(color=guide_legend(title=\"\")) +\n",
    "    scale_color_manual( values = c(\"#39568CFF\",\"#1F968BFF\") ) +\n",
    "    stat_ecdf(geom=\"step\", linewidth=1.5) +\n",
    "    ggtitle(\"Role of TATA+Inr element in Bursts\") +\n",
    "    xlab(\"Mean Expression\") +\n",
    "    ylab(\"Fraction of genes\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "553334d4-2de9-4c44-8d6a-718cc5a52817",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(tata_inr_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
