{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "637be83a-ef96-4ec1-93cf-0051dbbc4dc0",
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths(\"/home/mahat/.conda/envs/r422/lib/R/library\")\n",
    ".libPaths()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b159ff73-dadd-4aa4-beec-a733593eaa4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "    library(tidyverse)\n",
    "    library(dplyr)\n",
    "    library(matrixStats)\n",
    "    library(foreach)\n",
    "    library(doParallel)\n",
    "    library(plyranges)\n",
    "    library(viridis)\n",
    "    library(ggpointdensity)\n",
    "    library(Matrix)\n",
    "    library(data.table)\n",
    "    library(Hmisc)\n",
    "    library(gplots)\n",
    "    library(rstatix)\n",
    "    library(scattermore)\n",
    "    library(igraph)\n",
    "    library(ggraph)\n",
    "    library(clusterProfiler)\n",
    "    library(stringr)\n",
    "    library(org.Mm.eg.db)\n",
    "    library(circlize)\n",
    "    library(gplots)\n",
    "    library(RIdeogram)\n",
    "    library(enrichplot)\n",
    "    library(GOSemSim)\n",
    "    library(dynamicTreeCut)\n",
    "    library(WGCNA)\n",
    "    library(pheatmap)\n",
    "    library(seriation)\n",
    "    library(dendextend)\n",
    "    library(Seurat)\n",
    "    library(DESeq2)\n",
    "    library(lattice)\n",
    "    # library(ComplexHeatmap)\n",
    "});\n",
    "source(\"./scGRO_functions.r\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "104b0744-9e76-43bb-b44f-a1210e78db0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "registerDoParallel(16);\n",
    "options(\n",
    "    repr.plot.width=8,\n",
    "    repr.plot.height=6,\n",
    "    jupyter.plot_mimetypes = \"image/svg+xml\",\n",
    "    digits=5\n",
    ");\n",
    "theme_set(theme_classic() +\n",
    "    theme(\n",
    "        axis.title.x = element_text(color=\"black\", size=14, face=\"bold\"), \n",
    "        axis.title.y = element_text(color=\"black\", size=14, face=\"bold\"),\n",
    "        axis.text = element_text(color=\"black\", size=12, face=\"bold\"),\n",
    "        plot.title = element_text(face=\"bold\", size=14, hjust = 0.5),\n",
    "        axis.line = element_blank(),\n",
    "        # axis.ticks = element_blank()\n",
    "        panel.border = element_rect(colour = \"grey\", fill=NA, linewidth=1)\n",
    "    )\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d45707e4-6833-487d-8b0d-354cc9215a02",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load genes with dREG status for expression filter\n",
    "genesWithdREGstatus = readRDS(\"../data/groHMM_dREG_refinedGenes_mES_mm10.rds\");\n",
    "genesWithdREGstatus\n",
    "# load groHMM-extended genes and enhancers\n",
    "# features = read_bed(\"../data/groHMM_mES_BRsComb_LP-50_UTS10_features_customized_v2.bed\");\n",
    "\n",
    "# load features\n",
    "features = read_bed(\"../data/groHMM_dREG_refinedFeatures_mES_mm10_OSNcustomEnhancers_SEs.bed\");\n",
    "# features = read_bed(\"../data/mES_BRsComb_dREGfiltered_features_customized_OSNenhancersPlusSEs_v1.bed\");\n",
    "# features = read_bed(\"../data/groHMM_mES_BRsComb_LP-50_UTS10_features_customized_OSNenhancersPlusSEs_v2.bed\");\n",
    "names(features) = features$name;\n",
    "# features$name=NULL;\n",
    "features$score=NULL;\n",
    "summary(width(features))\n",
    "features\n",
    "\n",
    "# truncate long features to 20 kb:\n",
    "longf = which( width(features) > 20000 );\n",
    "features[longf] = features[longf] %>%\n",
    "    # filter( substr(names, 0, 3) == \"GN=\" ) %>%\n",
    "    resize( width = 20000, fix=\"start\" );\n",
    "summary(width(features))\n",
    "features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c251dbb-11f5-4c0c-b26d-81ce5f15bd48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# OSNK = read.table(\"~/group/genes_enhancer_list/Custom_mES_OSN_v4.bed\");\n",
    "# OSNK\n",
    "# test = OSNK[OSNK[,3] - OSNK[,2] < 0, ]\n",
    "# test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94d8a436-f910-4ec8-b481-7535aa6a9752",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Oct4, Sox2, Nanog, Klf4 5k bins:\n",
    "OSNK = read_bed(\"~/group/genes_enhancer_list/Custom_mES_OSN_v4.bed\") %>%\n",
    "    # filter out bins that are larger than 4:\n",
    "    filter( as.numeric(sub(\"^.*_\", \"\", name)) <= 4 ) %>%\n",
    "    sort();\n",
    "names(OSNK) = OSNK$name;\n",
    "# customOSN$score = NULL;\n",
    "\n",
    "name = c(\"INT_STITCHED_7784\", \"INT_STITCHED_1482\", \"INT_STITCHED_1973\",\n",
    "         \"INT_STITCHED_3347\", \"INT_STITCHED_3348\", \"INT_STITCHED_3349\")\n",
    "assignedGene = c(\"GN-Pou5f1\", \"GN-Sox2\", \"GN-Klf4\",\n",
    "                 \"GN-Nanog\", \"GN-Nanog\", \"GN-Nanog\")\n",
    "SElocation = c(\"Upstream\", \"Downstream\", \"Upstream\",\n",
    "               \"Upstream\", \"Upstream\", \"Downstream\")\n",
    "OSNKassignment = data.frame(name, assignedGene, SElocation);\n",
    "OSNKassignment\n",
    "\n",
    "## SEs that were CRISPRed in Moorthy 2017 Genome Research paper:\n",
    "vSE = read_bed(\"~/group/genes_enhancer_list/CRISPR_verified_SEs_Moorthy_2017.bed\") %>%\n",
    "    # filter out bins that are larger than 4:\n",
    "    filter( as.numeric(sub(\"^.*_\", \"\", name)) <= 4 ) %>%\n",
    "    sort();\n",
    "names(vSE) = vSE$name;\n",
    "# verifiedSE$score = NULL;\n",
    "\n",
    "name = c(\"INT_STITCHED_4179\", \"INT_STITCHED_5044\", \n",
    "         \"INT_STITCHED_2745\", \"INT_STITCHED_2175\", \"INT_STITCHED_5427\", \n",
    "         \"INT_STITCHED_1210\", \"INT_STITCHED_6113\", \"INT_STITCHED_6112\", \n",
    "         \"INT_STITCHED_4657\", \"INT_STITCHED_1626\", \"INT_STITCHED_746\")\n",
    "assignedGene = c(\"GN-Sall1\", \"GN-Tet1\", \n",
    "                 \"GN-Med13l\", \"GN-Macf1\", \"GN-Ranbp17\", \n",
    "                \"GN-Cbfa2t2\", \"GN-Esrrb\", \"GN-Esrrb\", \n",
    "                 \"GN-Ooep\", \"GN-Mcl1\",  \"GN-Etl4\")\n",
    "SElocation = c(\"Downstream\", \"Upstream\", \n",
    "               \"Upstream\", \"Intron\", \"Upstream\", \n",
    "               \"Upstream\", \"Intron\", \"Intron\", \n",
    "               \"Upstream\", \"Upstream\", \"Intron\")\n",
    "vSEassignment = data.frame(name, assignedGene, SElocation);\n",
    "vSEassignment\n",
    "\n",
    "# rbinding OSNK and vSE assignments\n",
    "SEassignment = rbind(OSNKassignment, vSEassignment)\n",
    "SEassignment\n",
    "\n",
    "# SEs and Genes that do not have SEs in gene intron:\n",
    "nonIntronSEs = SEassignment[SEassignment$SElocation != \"Intron\", 1]\n",
    "width(features[nonIntronSEs])\n",
    "unique(nonIntronSEs)\n",
    "nonIntronGenes = SEassignment[SEassignment$SElocation != \"Intron\", 2]\n",
    "width(features[nonIntronGenes])\n",
    "unique(nonIntronGenes)\n",
    "\n",
    "# Remove SEs and associated Genes where SE lies in Intron:\n",
    "verifiedSE = suppressWarnings(c(OSNK, vSE)) %>%\n",
    "    filter( sub(\"_[^_]+$\", \"\", names) %in% c(nonIntronSEs, nonIntronGenes));\n",
    "summary(width(verifiedSE))\n",
    "verifiedSE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82b2e448-be1a-4adf-a94f-ac4e1db2e8f3",
   "metadata": {},
   "outputs": [],
   "source": [
    "unique(seqnames(verifiedSE))\n",
    "table(unique(sub(\"_[^_]+$\", \"\", names(verifiedSE))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9637515e-5077-438f-8ef0-f14ff41ba8b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# select only SEs, not genes:\n",
    "SEonly = verifiedSE %>%\n",
    "    filter( substr(names, 0, 3) == \"INT\" );\n",
    "length(SEonly)\n",
    "\n",
    "# get all features in the neighborhood of SEs\n",
    "SEneighborsAll = features %>%\n",
    "    # filter genes with dREG peaks\n",
    "    filter(names %in% names(genesWithdREGstatus[genesWithdREGstatus$dREG == T]) | \n",
    "           !names %in% names(genesWithdREGstatus) ) %>%\n",
    "    # ignoring the Gm genes:\n",
    "    filter( !substr(names, 0, 5) == \"GN-Gm\" ) %>%\n",
    "    subsetByOverlaps(SEonly, maxgap=5000000);\n",
    "length(SEneighborsAll)\n",
    "\n",
    "# get the distance between SEneighbors and SE\n",
    "hits = distanceToNearest( SEneighborsAll, SEonly, ignore.strand=T );\n",
    "# assign the SE to SEneighbors genes by taking name of hits@to\n",
    "SEneighborsAll$SE = sub(\"_[^_]+$\", \"\", names(SEonly)[hits@to]); \n",
    "# assign the distance between SEneighbors and SE with hits@elementMetadata$distance\n",
    "SEneighborsAll$distance = hits@elementMetadata$distance;\n",
    "\n",
    "SEneighbors = SEneighborsAll %>%\n",
    "    filter(substr(names, 0, 3) == \"GN-\") %>%\n",
    "    # filter SE assigned genes so that the bins don't run into these genes\n",
    "    filter( !names %in% nonIntronGenes ) %>%\n",
    "    # 20 kb for dividing into 4 5kb bins\n",
    "    filter( width >= 20000 ) %>%\n",
    "    filter( distance < 1000000 | distance > 4000000 ) %>%\n",
    "    # add a column indicating if the gene is within TAD or outside TAD:\n",
    "    mutate( TADstatus = ifelse(distance <= 1000000, \"IN\", \"OUT\"));\n",
    "\n",
    "# filter for non-overlapping features:\n",
    "SEneighbors <- SEneighbors %>%\n",
    "    # reduce merges overlapping ranges and subsetting with \"equal\" only return that did not get merged\n",
    "    subsetByOverlaps( reduce(SEneighbors), type = \"equal\" );\n",
    "\n",
    "SEneighborsOthers = SEneighborsAll %>%\n",
    "    filter( !names %in% names(SEneighbors) );\n",
    "length(SEneighborsOthers)\n",
    "\n",
    "summary(as.factor(SEneighbors$TADstatus))\n",
    "length(unique(names(SEneighbors)))\n",
    "length(unique(SEneighbors$SE))\n",
    "SEneighbors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2bcda1d-8630-4f6f-8dd0-d3c631820ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = readRDS(\"../data/scGROv2p8_mapq3qc_max20kbp_filtered_counts.rds\");\n",
    "#counts = counts[,colSums(counts) >= 1000 ];\n",
    "dim(counts)\n",
    "allcells = colnames(counts);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "41ab20aa-1c09-479c-a4b8-73ba3f916f49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the range of expression of vefifiedSEGenes:\n",
    "verifiedSEgenesON = counts[ nonIntronGenes, ]\n",
    "dim(verifiedSEgenesON)\n",
    "summary(rowMeans(verifiedSEgenesON > 0))\n",
    "summary(as.factor(colSums(verifiedSEgenesON > 0)))\n",
    "# how many cells have at least 1 S-phase genes expressed:\n",
    "verifiedSEgenesExpressingCells = ncol(verifiedSEgenesON[ ,colSums(verifiedSEgenesON) > 0])\n",
    "verifiedSEgenesExpressingCells/ncol(counts)*100\n",
    "\n",
    "# Subset SEneighbors Genes such that their expression is similar to verifiedSEgenes\n",
    "SEneighborsON = counts[ names(SEneighbors), ]\n",
    "dim(SEneighborsON)\n",
    "summary(rowMeans(SEneighborsON > 0))\n",
    "# filter SEneighbors with expression within the verfiedSEgenes:\n",
    "SEneighborsON = SEneighborsON[rowMeans(SEneighborsON) >= min(rowMeans(verifiedSEgenesON > 0)) & \n",
    "                     rowMeans(SEneighborsON) <= max(rowMeans(verifiedSEgenesON > 0)), ]\n",
    "dim(SEneighborsON)\n",
    "summary(rowMeans(SEneighborsON > 0))\n",
    "\n",
    "# how many SEneighbor genes inside and outside of TAD survive the expression filter?\n",
    "test = SEneighbors %>%\n",
    "    filter( names %in% rownames(SEneighborsON) );\n",
    "summary(as.factor(test$TADstatus))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7511caa-bfd1-44f0-80c1-07d1cda63078",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create 5kbp time-bins of SEneighbors:\n",
    "# remove bins if they overlap with SEneighborOthers and the previous bins\n",
    "SEneighbors1 = resize(SEneighbors[rownames(SEneighborsON)], width=5000, fix=\"start\") %>%\n",
    "    subsetByOverlaps( SEneighborsOthers, invert = TRUE );\n",
    "names(SEneighbors1) = paste0(SEneighbors1$name, \"_1\");\n",
    "length(SEneighbors1)\n",
    "# # 5 kb upstream region - divergent transcription\n",
    "# SEneighborsDiv1 = shift_upstream(SEneighbors1, 5000) %>%\n",
    "#     mutate( strand = ifelse(strand == \"+\", \"-\", \"+\") ) %>%\n",
    "#     subsetByOverlaps( c(SEneighbors1), invert = TRUE );\n",
    "# names(SEneighborsDiv1) = paste0(SEneighborsDiv1$name, \"_1\")\n",
    "\n",
    "SEneighbors2 = shift_downstream(SEneighbors1, 5000) %>%\n",
    "    subsetByOverlaps( c(SEneighborsOthers, SEneighbors1), invert = TRUE );\n",
    "names(SEneighbors2) = paste0(SEneighbors2$name, \"_2\")\n",
    "length(SEneighbors2)\n",
    "SEneighbors3 = shift_downstream(SEneighbors2, 5000) %>%\n",
    "    subsetByOverlaps( c(SEneighborsOthers, SEneighbors2), invert = TRUE );\n",
    "names(SEneighbors3) = paste0(SEneighbors3$name, \"_3\")\n",
    "length(SEneighbors3)\n",
    "SEneighbors4 = shift_downstream(SEneighbors3, 5000) %>%\n",
    "    subsetByOverlaps( c(SEneighborsOthers, SEneighbors3), invert = TRUE );\n",
    "names(SEneighbors4) = paste0(SEneighbors4$name, \"_4\")\n",
    "length(SEneighbors4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c62c6e14-3aa9-461e-bea4-bc70459810c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://github.com/hbc/tinyatlas\n",
    "# cell cycle marker genes:\n",
    "cc_markers = fread(\"../data/Cell_cycle_marker_genes.csv\");\n",
    "# cc_markers[1:4, ]\n",
    "\n",
    "s_genes = features %>%\n",
    "    filter( sub(\"GN-\", \"\", names) %in% cc_markers[cc_markers$phase == \"S\"]$gene_name ) %>%\n",
    "    mutate( phase = \"S\" );\n",
    "length(s_genes);\n",
    "\n",
    "g2m_genes = features %>%\n",
    "    filter( sub(\"GN-\", \"\", names) %in% cc_markers[cc_markers$phase == \"G2/M\"]$gene_name ) %>%\n",
    "    mutate( phase = \"G2/M\" );\n",
    "length(g2m_genes);\n",
    "\n",
    "ccMarkers = c(s_genes, g2m_genes)\n",
    "summary(width(ccMarkers))\n",
    "summary(as.factor(ccMarkers$phase))\n",
    "ccMarkers\n",
    "\n",
    "# Get the range of expression of s_phase genes:\n",
    "s_genesON = as.matrix(counts[ names(s_genes), ])\n",
    "summary(rowMeans(s_genesON > 0))\n",
    "table(colSums(s_genesON > 0))\n",
    "# filter cells that have at least 10% of S-phase genes expressed:\n",
    "s_genesON = s_genesON[ ,colSums(s_genesON > 0) >= length(s_genes)/10 ]\n",
    "# how many cells have at least 10% of S-phase genes expressed:\n",
    "ncol(s_genesON)/ncol(counts)*100\n",
    "\n",
    "# # converting into a binary matrix:\n",
    "# s_genesON[s_genesON > 0] = 1;\n",
    "\n",
    "# converting less than 2 to 0:\n",
    "s_genesON[s_genesON < 2] = 0;\n",
    "\n",
    "pdf('../plots/S_phase_genes_heatmap.pdf',width=20,height=20)\n",
    "heatmap.2(s_genesON,\n",
    "    dendrogram='column', scale='none',\n",
    "    # Rowv=FALSE, Colv=FALSE,\n",
    "    # col=gray.colors(1,0,n=2), col=cm.colors(2),\n",
    "    col=colorRampPalette(c(\"white\", \"red\"))(n = 2),\n",
    "    trace='none', keysize=2, density.info='none',\n",
    "    labRow=FALSE, labCol=FALSE)\n",
    "dev.off()\n",
    "s_genesONhm = pheatmap(s_genesON,\n",
    "         cluster_cols = TRUE, \n",
    "         cluster_rows = TRUE,\n",
    "         scale = \"none\",\n",
    "         show_rownames = TRUE,\n",
    "         show_colnames = FALSE,\n",
    "         color = colorRampPalette(c(\"white\", \"red\"))(15),\n",
    "         border_color = \"blue\",\n",
    "         cellwidth = 0.1,\n",
    "         cellheight = 7.5,\n",
    "         # cutree_cols = 6,\n",
    "         # fontsize_col = 8,\n",
    "         fontsize_row = 8)\n",
    "# ggsave(s_genesONhm, filename = \"../plots/S_phase_genes_heatmap.pdf\", width=20, height=20);   \n",
    "\n",
    "# Get the range of expression of g2/m_phase genes:\n",
    "g2m_genesON = as.matrix(counts[ names(g2m_genes), ])\n",
    "summary(rowMeans(g2m_genesON > 0))\n",
    "table(colSums(g2m_genesON > 0))\n",
    "# filter cells that have at least 10% of G2M-phase genes expressed:\n",
    "# how many cells have at least 10% of G2M-phase genes expressed:\n",
    "g2m_genesON = g2m_genesON[ ,colSums(g2m_genesON > 0) >= length(g2m_genes)/10 ]\n",
    "\n",
    "# # converting into a binary matrix:\n",
    "# g2m_genesON[g2m_genesON > 0] = 1;\n",
    "\n",
    "# converting less than 2 to 0:\n",
    "g2m_genesON[g2m_genesON < 2] = 0;\n",
    "\n",
    "pdf('../plots/G2M_phase_genes_heatmap.pdf',width=20,height=20)\n",
    "heatmap.2(g2m_genesON,\n",
    "    dendrogram='column', scale='none',\n",
    "    # Rowv=FALSE, Colv=FALSE,\n",
    "    # col=gray.colors(1,0,n=2), col=cm.colors(2),\n",
    "    col=colorRampPalette(c(\"white\", \"red\"))(n = 10),\n",
    "    trace='none', keysize=2, density.info='none',\n",
    "    labRow=FALSE, labCol=FALSE)\n",
    "dev.off()\n",
    "g2m_genesONhm = pheatmap(g2m_genesON,\n",
    "         cluster_cols = TRUE, \n",
    "         cluster_rows = TRUE,\n",
    "         scale = \"none\",\n",
    "         show_rownames = TRUE,\n",
    "         show_colnames = FALSE,\n",
    "         color = colorRampPalette(c(\"white\", \"red\"))(15),\n",
    "         border_color = \"blue\",\n",
    "         cellwidth = 0.1,\n",
    "         cellheight = 5,\n",
    "         cutree_cols = 6,\n",
    "         # fontsize_col = 8,\n",
    "         fontsize_row = 8)\n",
    "# ggsave(g2m_genesONhm, filename = \"../plots/G2M_phase_genes_heatmap.pdf\", width=20, height=20);\n",
    "\n",
    "# from Long Cai intron seqFISH:\n",
    "# \"cells were sorted by G2/M marker gene mRNA levels (Aurka and Plk1)\"\n",
    "g2mCai_genes = features[ c(\"GN-Aurka\", \"GN-Plk1\") ]\n",
    "\n",
    "# # get all features in the neighborhood of ccMarker genes:\n",
    "# ccNeighbors = features %>%\n",
    "#     # filter genes with dREG peaks\n",
    "#     filter(names %in% names(genesWithdREGstatus[genesWithdREGstatus$dREG == T]) | \n",
    "#            !names %in% names(genesWithdREGstatus) ) %>%\n",
    "#     # ignoring the Gm genes:\n",
    "#     filter( !substr(names, 0, 5) == \"GN-Gm\" ) %>%\n",
    "#     # only find neighbors within 20kb region:\n",
    "#     subsetByOverlaps(ccMarkers, maxgap=20000) %>%\n",
    "#     filter(substr(names, 0, 3) == \"GN-\") %>%\n",
    "#     # filter SE assigned genes so that the bins don't run into these genes\n",
    "#     filter( !names %in% names(ccMarkers) );\n",
    "# ccNeighbors\n",
    "\n",
    "\n",
    "# # create 5kbp time-bins of ccMarkers:\n",
    "# # remove bins if they overlap with ccneighbors and the previous bins\n",
    "# ccMarkers1 = resize( ccMarkers, width=5000, fix=\"start\") %>%\n",
    "#     subsetByOverlaps( ccNeighbors, invert = TRUE );\n",
    "# names(ccMarkers1) = paste0(ccMarkers1$name, \"_1\");\n",
    "# length(ccMarkers1)\n",
    "# # # 5 kb upstream region - divergent transcription\n",
    "# # SEneighborsDiv1 = shift_upstream(SEneighbors1, 5000) %>%\n",
    "# #     mutate( strand = ifelse(strand == \"+\", \"-\", \"+\") ) %>%\n",
    "# #     subsetByOverlaps( c(SEneighbors1), invert = TRUE );\n",
    "# # names(SEneighborsDiv1) = paste0(SEneighborsDiv1$name, \"_1\")\n",
    "\n",
    "# ccMarkers2 = shift_downstream(ccMarkers1, 5000) %>%\n",
    "#     subsetByOverlaps( ccNeighbors, invert = TRUE );\n",
    "# names(ccMarkers2) = paste0(ccMarkers2$name, \"_2\")\n",
    "# length(ccMarkers2)\n",
    "# ccMarkers3 = shift_downstream(ccMarkers2, 5000) %>%\n",
    "#     subsetByOverlaps( ccNeighbors, invert = TRUE );\n",
    "# names(ccMarkers3) = paste0(ccMarkers3$name, \"_3\")\n",
    "# length(ccMarkers3)\n",
    "# ccMarkers4 = shift_downstream(ccMarkers3, 5000) %>%\n",
    "#     subsetByOverlaps( ccNeighbors, invert = TRUE );\n",
    "# names(ccMarkers4) = paste0(ccMarkers4$name, \"_4\")\n",
    "# length(ccMarkers4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4103e5e-67f3-4adb-84fc-3d76adfa4e17",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# histone genes:\n",
    "histGenes = features %>%\n",
    "    filter( substr(names, 0, 7) == \"GN-Hist\") #%>%\n",
    "    # filter( names %in% rownames(counts[rowSums(counts) > 0,]) );\n",
    "    # filter( names %in% rownames(counts[rowMeans(counts) > 0.001,]) );\n",
    "summary(as.factor(seqnames(histGenes)))\n",
    "summary(width(histGenes))\n",
    "\n",
    "# Get the range of expression of histone genes:\n",
    "histGenesON = as.matrix(counts[ names(histGenes), ])\n",
    "dim(histGenesON)\n",
    "summary(rowMeans(histGenesON > 0))\n",
    "table(colSums(histGenesON > 0))\n",
    "# how many cells have at least 1 histone genes expressed:\n",
    "histoneExpressingCells = ncol(histGenesON[ ,colSums(histGenesON > 0) > 0])\n",
    "histoneExpressingCells/ncol(counts)*100\n",
    "\n",
    "# converting into a binary matrix:\n",
    "histGenesON[histGenesON > 1] = 1;\n",
    "\n",
    "# heatmap:\n",
    "pdf('../plots/histoneGenes_heatmap.pdf',width=20,height=20)\n",
    "heatmap.2(histGenesON,\n",
    "    dendrogram='column', scale='none',\n",
    "    # Rowv=FALSE, Colv=FALSE,\n",
    "    # col=gray.colors(1,0,n=2), col=cm.colors(2),\n",
    "    col=colorRampPalette(c(\"white\", \"red\"))(n = 2),\n",
    "    trace='none', keysize=2, density.info='none',\n",
    "    labRow=FALSE, labCol=FALSE)\n",
    "dev.off()\n",
    "\n",
    "# truncate histone genes to 5 kb or 1st bin:\n",
    "longf = which( width(histGenes) > 5000 );\n",
    "histGenes[longf] = histGenes[longf] %>%\n",
    "    resize( width = 5000, fix=\"start\" );\n",
    "summary(width(histGenes))\n",
    "sum(width(histGenes))\n",
    "histGenes"
   ]
  },
  {
   "cell_type": "raw",
   "id": "28a2ffbb-4f86-498e-93a3-0b411372a989",
   "metadata": {},
   "source": [
    "histone genes width without groHMM:\n",
    "Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
    "359     444     470     573     526    2984 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4bf79ac2-bdae-44a6-a572-47cae2338326",
   "metadata": {},
   "outputs": [],
   "source": [
    "# NEW cell-cyle marker genes:\n",
    "\n",
    "# https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2600614/#pone.0003943.s001\n",
    "G1_Beyrouthy = c(\"GN-Fos\", \"GN-Egr1\", \"GN-Atf3\", \"GN-Myadm\") # \"GN-Ctgf\"\n",
    "# https://www.nature.com/articles/s41467-022-30545-8\n",
    "G1S_Riba = c(\"GN-Orc1\", \"GN-Ccne1\", \"GN-Ccne2\", \"GN-Mcm6\")\n",
    "G2M_Riba = c(\"GN-Wee1\", \"GN-Cdk1\", \"GN-Ccnf\", \"GN-Nusap1\", \"GN-Aurka\", \"GN-Ccna2\") # \"GN-Ccnb2\"\n",
    "\n",
    "ccMarkers = features[ c(G1_Beyrouthy, G1S_Riba, G2M_Riba) ]\n",
    "width(ccMarkers)\n",
    "\n",
    "# # https://www.molbiolcell.org/doi/epdf/10.1091/mbc.02-02-0030\n",
    "# G1S_Whitfield = c(\"GN-Ccne1\", \"GN-Ccne2\", \"GN-Cdc25a\", \"GN-Cdc6\", \"GN-E2f1\", \"GN-Mcm2\", \"GN-Mcm6\", \"GN-Npat\", \"GN-Pcna\", \"GN-Slbp\")\n",
    "# S_Whitfield = c(\"GN-Brca1\", \"GN-Dhfr\", \"GN-Msh2\", \"GN-Rrm1\", \"GN-Rrm2\", \"GN-Tyms\")\n",
    "# # G2_Whitfield = c(\"GN-Ccna2\", \"GN-Ccnf\", \"GN-Top2a\")\n",
    "# G2M_Whitfield = c(\"GN-Birc5\", \"GN-Bub1\", \"GN-Bub1b\", \"GN-Ccnb1\", \"GN-Ccnb2\", \"GN-Cdc20\", \"GN-Cdc25b\", \"GN-Cenpa\", \"GN-Cks2\", \"GN-Plk1\") #  \"GN-Cdkn2d\" - short and lost due to overlap,  \"GN-Stk15\" -- not present in mouse\n",
    "\n",
    "# ccMarkers = features[ c(G1S_Whitfield, S_Whitfield, G2M_Whitfield) ]\n",
    "# width(ccMarkers)\n",
    "\n",
    "# get all features in the neighborhood of ccMarker genes:\n",
    "ccNeighbors = features %>%\n",
    "    # filter genes with dREG peaks\n",
    "    filter(names %in% names(genesWithdREGstatus[genesWithdREGstatus$dREG == T]) | \n",
    "           !names %in% names(genesWithdREGstatus) ) %>%\n",
    "    # ignoring the Gm genes:\n",
    "    filter( !substr(names, 0, 5) == \"GN-Gm\" ) %>%\n",
    "    # only find neighbors within 20kb region:\n",
    "    subsetByOverlaps(ccMarkers, maxgap=20000) %>%\n",
    "    filter(substr(names, 0, 3) == \"GN-\") %>%\n",
    "    # filter SE assigned genes so that the bins don't run into these genes\n",
    "    filter( !names %in% names(ccMarkers) );\n",
    "\n",
    "# create 5kbp time-bins of ccMarkers:\n",
    "# remove bins if they overlap with ccneighbors and the previous bins\n",
    "ccMarkers1 = resize( ccMarkers, width=5000, fix=\"start\") %>%\n",
    "    subsetByOverlaps( ccNeighbors, invert = TRUE, minoverlap = 1000 );\n",
    "names(ccMarkers1) = paste0(ccMarkers1$name, \"_1\");\n",
    "length(ccMarkers1)\n",
    "names(ccMarkers1)\n",
    "ccMarkers1\n",
    "# # 5 kb upstream region - divergent transcription\n",
    "# SEneighborsDiv1 = shift_upstream(SEneighbors1, 5000) %>%\n",
    "#     mutate( strand = ifelse(strand == \"+\", \"-\", \"+\") ) %>%\n",
    "#     subsetByOverlaps( c(SEneighbors1), invert = TRUE );\n",
    "# names(SEneighborsDiv1) = paste0(SEneighborsDiv1$name, \"_1\")\n",
    "\n",
    "ccMarkers2 = shift_downstream(ccMarkers1, 5000) %>%\n",
    "    subsetByOverlaps( ccNeighbors, invert = TRUE );\n",
    "names(ccMarkers2) = paste0(ccMarkers2$name, \"_2\")\n",
    "length(ccMarkers2)\n",
    "ccMarkers3 = shift_downstream(ccMarkers2, 5000) %>%\n",
    "    subsetByOverlaps( ccNeighbors, invert = TRUE );\n",
    "names(ccMarkers3) = paste0(ccMarkers3$name, \"_3\")\n",
    "length(ccMarkers3)\n",
    "ccMarkers4 = shift_downstream(ccMarkers3, 5000) %>%\n",
    "    subsetByOverlaps( ccNeighbors, invert = TRUE );\n",
    "names(ccMarkers4) = paste0(ccMarkers4$name, \"_4\")\n",
    "length(ccMarkers4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ef0813d-8597-49b6-af8e-c08199713269",
   "metadata": {},
   "outputs": [],
   "source": [
    "sum(width(features[ G1_Beyrouthy ]))\n",
    "sum(width(features[ G1S_Riba ]))\n",
    "sum(width(features[ G2M_Riba ]))\n",
    "sum(width(histGenes))\n",
    "\n",
    "# sum(width(features[ G1S_Whitfield ]))\n",
    "# sum(width(features[ S_Whitfield ]))\n",
    "# sum(width(features[ G2M_Whitfield ]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b14e64d0-4c13-4942-b5dd-30805a7f85bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine features without histone genes:\n",
    "# select longer than 1 kb and truncate 1 kb:\n",
    "allfeatures = c(verifiedSE, \n",
    "                SEneighbors1, \n",
    "                SEneighbors2, \n",
    "                SEneighbors3, \n",
    "                SEneighbors4, \n",
    "                ccMarkers1,\n",
    "                ccMarkers2,\n",
    "                ccMarkers3,\n",
    "                ccMarkers4) %>%\n",
    "    filter( width > 1000 );\n",
    "allfeatures\n",
    "\n",
    "# truncate 1st kb of bin-1s:\n",
    "bin1 = which( as.numeric(sub(\"^.*_\", \"\", allfeatures$name)) == 1 );\n",
    "allfeatures[bin1] = allfeatures[bin1] %>%\n",
    "    anchor_3p() %>%\n",
    "    mutate( width = width - 1000 );\n",
    "summary(width(allfeatures))\n",
    "\n",
    "# add histone genes, without truncating\n",
    "allfeatures = c(allfeatures, histGenes)\n",
    "summary(width(allfeatures))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c94826c-6623-473d-aef8-23660a5d6c9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "scGRO = readRDS(\"../data/scGROv2p8_consolidated.rds\");\n",
    "# counts = readRDS(\"../data/scGROv2p8_mapq3qc_max20kbp_filtered_counts.rds\");\n",
    "# #counts = counts[,colSums(counts) >= 1000 ];\n",
    "# dim(counts)\n",
    "# allcells = colnames(counts);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb4475ed-5fd3-4b06-9aef-7643be24d56c",
   "metadata": {},
   "outputs": [],
   "source": [
    "table(substr(allcells, 0, 6));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "287bf640-cf2f-4b91-a83e-500d4681571d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge experiment ID and cell barcode to create unique cell ID across experiments\n",
    "scGRO = scGRO %>%\n",
    "    filter( mapq >= 3 & countQC & umiQC & plateQC & cellQC & miRQC ) %>%\n",
    "    # filter( !(Exp == \"Exp236\" & Plate %in% paste0(\"c0\", 5:8)) ) %>%\n",
    "    # filter( !(Exp == \"Exp260b\") ) %>%\n",
    "    subsetByOverlaps( allfeatures ) %>%\n",
    "    # subsetByOverlaps( verifiedSE ) %>%\n",
    "    mutate( cellID = factor( paste( Exp, Plate, Cell, sep=\"-\") ) ) %>%\n",
    "    filter( cellID %in% allcells ) #%>%\n",
    "    # select( cellID );\n",
    "scGRO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "59028396-df52-4f20-89b6-c92fa23d0ac2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "counts = foreach(\n",
    "    id = allcells,\n",
    "    .combine=\"cbind2\"\n",
    ") %dopar% {\n",
    "    reads = scGRO %>% filter(cellID == id);\n",
    "    # count reads in each feature for this cellID\n",
    "    counts1 = countOverlaps( allfeatures, reads );\n",
    "    # counts1 = countOverlaps( verifiedSE, reads );\n",
    "\n",
    "    counts = cbind(counts1) %>%\n",
    "        Matrix(sparse=T);\n",
    "\n",
    "    return( counts );\n",
    "}\n",
    "# columns = cells = cellIDs\n",
    "colnames(counts) = allcells;\n",
    "rownames(counts) = names(allfeatures);\n",
    "# rownames(counts) = names(verifiedSE);\n",
    "dim(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3fd9bd41-fb97-4e9e-acb4-3f53a930f54a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to sum rows with same rownames\n",
    "# the enhancers at various bins have same names\n",
    "counts = rowsum(counts, row.names(counts))\n",
    "dim(counts)\n",
    "# rowSums(counts)\n",
    "# colSums(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "286be97d-f3ad-4cc3-9518-48bc6d9cc129",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "newCounts = counts\n",
    "# remove bins ID in row names, so that the counts in the bins of same genes can be added:\n",
    "row.names(newCounts) = sub(\"_[^_]+$\", \"\", row.names(newCounts))\n",
    "dim(newCounts)\n",
    "# add counts from the same gene but different bins together:\n",
    "newCounts = rowsum(newCounts, row.names(newCounts))\n",
    "dim(newCounts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28c128c2-bfc8-4657-b559-2816cb89e9ef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# trying to see if G1S, Histone genes, and G2M genes expressing cells are distinct:\n",
    "G1S_Hist_G2M = as.matrix(newCounts[ c(G1_Beyrouthy, G1S_Riba, names(histGenes), G2M_Riba), ])\n",
    "# converting into a binary matrix:\n",
    "G1S_Hist_G2M[G1S_Hist_G2M > 1] = 1;\n",
    "dim(G1S_Hist_G2M)\n",
    "# Create the heatmap using pheatmap\n",
    "# https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/\n",
    "hmap = pheatmap(G1S_Hist_G2M,\n",
    "         cluster_cols = TRUE, \n",
    "         cluster_rows = FALSE,\n",
    "         scale = \"none\",\n",
    "         show_rownames = TRUE,\n",
    "         show_colnames = FALSE,\n",
    "         color = colorRampPalette(c(\"white\", \"red\"))(2),\n",
    "         border_color = NA,\n",
    "         # cellwidth = 3,\n",
    "         # cellheight = 3\n",
    "         # fontsize_col = 8,\n",
    "         fontsize_row = 3)\n",
    "ggsave(hmap, filename = \"../plots/G1S_HistoneGenes_G2M_heatmap.pdf\", width=20, height=20);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b289b22-4c2a-4e5b-ab43-bc0f17765e5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "emptyCells = G1S_Hist_G2M[ ,colSums(G1S_Hist_G2M) == 0]\n",
    "dim(emptyCells)\n",
    "\n",
    "emptyCells = counts[ ,colnames(emptyCells)]\n",
    "dim(emptyCells)\n",
    "\n",
    "new = c(allCells = colSums(counts), emptyCells = colSums(emptyCells))\n",
    "dim(new)\n",
    "head(new)\n",
    "\n",
    "summary(colSums(counts))\n",
    "summary(colSums(emptyCells))\n",
    "\n",
    "hist(colSums(counts))\n",
    "hist(colSums(emptyCells))\n",
    "# new %>%\n",
    "#     # filter(N_bursts >= 10 & burst_size > 1) %>%\n",
    "#     ggplot(aes(x=Type, y = burst_rate, fill = Type)) +\n",
    "#     geom_violin(show.legend = F, draw_quantiles = c(0.25, 0.5, 0.75)) +\n",
    "#     #geom_jitter(height = 0, width = 0.4, alpha=0.1, size=0.5) +\n",
    "#     scale_y_log10(limits = c(0.05, 5)) +\n",
    "#     ggtitle(\"Bursts from 10kb gene body\") +\n",
    "#     ylab(\"Bursts per hour\") +\n",
    "#     guides(fill=guide_legend(title=\"\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "66e66577-abbd-4097-8abe-00bf3927ee3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # trying to see if G1S and G2M genes expressing cells are distinct:\n",
    "# G1S_SplusHist_G2M = as.matrix(newCounts[ c(G1S_Whitfield, S_Whitfield, names(histGenes), G2M_Whitfield), ])\n",
    "\n",
    "# # provide rownames reflecting their origin for all rows\n",
    "# rownames(G1S_SplusHist_G2M) = c(rep(\"G1S\", each=length(G1S_Whitfield)), rep(\"S\", each=(length(S_Whitfield) + length(names(histGenes)))), rep(\"G2M\", each=length(G2M_Whitfield)));\n",
    "# dim(G1S_SplusHist_G2M)\n",
    "\n",
    "# # # converting into a binary matrix:\n",
    "# # G1S_Hist_G2M[G1S_Hist_G2M > 1] = 1;\n",
    "\n",
    "# # perform rowsums by rownames\n",
    "# G1S_SplusHist_G2M = rowsum(G1S_SplusHist_G2M, row.names(G1S_SplusHist_G2M))\n",
    "# dim(G1S_SplusHist_G2M)     \n",
    "\n",
    "# # R script that filters columns in a matrix so that \n",
    "# # the only columns with a value of more than 1 in any rows are retained:\n",
    "# G1S_SplusHist_G2M = G1S_SplusHist_G2M[, apply(G1S_SplusHist_G2M, 2, function(x) any(x > 1)) ]\n",
    "                                    \n",
    "# # change the row order of the matrix:\n",
    "# G1S_SplusHist_G2M = G1S_SplusHist_G2M[c(1,3,2),]\n",
    "# dim(G1S_SplusHist_G2M)\n",
    "# G1S_SplusHist_G2M[1:3,1:10]\n",
    "                                                    \n",
    "# # # converting into a binary matrix:\n",
    "# # G1S_Hist_G2M_binary = G1S_Hist_G2M[G1S_Hist_G2M > 1] = 1;\n",
    "# # G1S_Hist_G2M_binary[1:4,1:10] \n",
    "\n",
    "# # # Convert cells with less than 2 counts in 3rd row [ HIS ], to 0:\n",
    "# # # Can't perform this operation for 1 row, so did (3,3)\n",
    "# # G1S_Hist_G2M[c(3,3), ] = apply(G1S_Hist_G2M[c(3,3), ], 2, function(x) ifelse(x < 3, 0, x))\n",
    "# # dim(G1S_Hist_G2M)\n",
    "# # G1S_Hist_G2M[1:3,1:10]\n",
    "                           \n",
    "# # Create the heatmap using pheatmap\n",
    "# # https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/\n",
    "# hm = pheatmap(G1S_SplusHist_G2M,\n",
    "#          clustering_distance_cols=\"euclidean\",\n",
    "#          clustering_method=\"complete\",\n",
    "#          cluster_cols = TRUE, \n",
    "#          cluster_rows = FALSE,\n",
    "#          scale = \"none\",\n",
    "#          show_rownames = TRUE,\n",
    "#          show_colnames = FALSE,\n",
    "#          color = colorRampPalette(c(\"white\", \"red\"))(10),\n",
    "#          border_color = \"blue\",\n",
    "#          cellwidth = 0.3,\n",
    "#          cellheight = 20,\n",
    "#          # cutree_cols = 6,\n",
    "#          # fontsize_col = 8,\n",
    "#          fontsize_row = 8)\n",
    "# ggsave(hm, filename = \"../plots/G1S_SplusHistoneGenes_G2M_combined_heatmap.pdf\", width=20, height=20);   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74c1c272-8c33-45b1-8c65-bce78c23c97f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# trying to see if G1S and G2M genes expressing cells are distinct:\n",
    "G1_G1S_Hist_G2M = as.matrix(newCounts[ c(G1_Beyrouthy, G1S_Riba, names(histGenes), G2M_Riba), ])\n",
    "\n",
    "# provide rownames reflecting their origin for all rows\n",
    "rownames(G1_G1S_Hist_G2M) = c(rep(\"G1\", each=length(G1_Beyrouthy)), rep(\"G1/S\", each=length(G1S_Riba)), rep(\"S\", each=length(names(histGenes))), rep(\"G2/M\", each=length(G2M_Riba)));\n",
    "dim(G1_G1S_Hist_G2M)\n",
    "\n",
    "# # converting into a binary matrix:\n",
    "# G1S_Hist_G2M[G1S_Hist_G2M > 1] = 1;\n",
    "\n",
    "# perform rowsums by rownames\n",
    "G1_G1S_Hist_G2M = rowsum(G1_G1S_Hist_G2M, row.names(G1_G1S_Hist_G2M))\n",
    "dim(G1_G1S_Hist_G2M)     \n",
    "\n",
    "# R script that filters columns in a matrix so that \n",
    "# the only columns with a value of more than 1 in any rows are retained:\n",
    "G1_G1S_Hist_G2M = G1_G1S_Hist_G2M[, apply(G1_G1S_Hist_G2M, 2, function(x) any(x > 1)) ]\n",
    "                                    \n",
    "# change the row order of the matrix:\n",
    "G1_G1S_Hist_G2M = G1_G1S_Hist_G2M[c(1,2,4,3),]\n",
    "dim(G1_G1S_Hist_G2M)\n",
    "G1_G1S_Hist_G2M[1:4,1:10]\n",
    "                                                    \n",
    "# # converting into a binary matrix:\n",
    "# G1S_Hist_G2M_binary = G1S_Hist_G2M[G1S_Hist_G2M > 1] = 1;\n",
    "# G1S_Hist_G2M_binary[1:4,1:10] \n",
    "\n",
    "# # Convert cells with less than 2 counts in 3rd row [ HIS ], to 0:\n",
    "# # Can't perform this operation for 1 row, so did (3,3)\n",
    "# G1S_Hist_G2M[c(3,3), ] = apply(G1S_Hist_G2M[c(3,3), ], 2, function(x) ifelse(x < 3, 0, x))\n",
    "# dim(G1S_Hist_G2M)\n",
    "# G1S_Hist_G2M[1:3,1:10]\n",
    "                           \n",
    "# Create the heatmap using pheatmap\n",
    "# https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/\n",
    "hm = pheatmap(G1_G1S_Hist_G2M,\n",
    "         clustering_distance_cols=\"euclidean\",\n",
    "         clustering_method=\"complete\",\n",
    "         cluster_cols = TRUE, \n",
    "         cluster_rows = FALSE,\n",
    "         scale = \"none\",\n",
    "         show_rownames = TRUE,\n",
    "         show_colnames = FALSE,\n",
    "         color = colorRampPalette(c(\"white\", \"red\"))(10),\n",
    "         border_color = \"blue\",\n",
    "         cellwidth = 0.5,\n",
    "         cellheight = 20,\n",
    "         # cutree_cols = 6,\n",
    "         # fontsize_col = 8,\n",
    "         fontsize_row = 8)\n",
    "ggsave(hm, filename = \"../plots/G1_G1S_HistoneGenes_G2M_combined_heatmap.pdf\", width=20, height=20);   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1370c5d-01ed-47c0-a4e2-352f6d9aeac0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# trying to see if G1S and G2M genes expressing cells are distinct:\n",
    "G1S_Hist_G2M = as.matrix(newCounts[ c(G1S_Riba, names(histGenes), G2M_Riba), ])\n",
    "\n",
    "# provide rownames reflecting their origin for all rows\n",
    "rownames(G1S_Hist_G2M) = c(rep(\"G1/S\", each=length(G1S_Riba)), rep(\"S\", each=length(names(histGenes))), rep(\"G2/M\", each=length(G2M_Riba)));\n",
    "dim(G1S_Hist_G2M)\n",
    "\n",
    "# # converting into a binary matrix:\n",
    "# G1S_Hist_G2M[G1S_Hist_G2M > 1] = 1;\n",
    "\n",
    "# perform rowsums by rownames\n",
    "G1S_Hist_G2M = rowsum(G1S_Hist_G2M, row.names(G1S_Hist_G2M))\n",
    "dim(G1S_Hist_G2M)     \n",
    "\n",
    "# R script that filters columns in a matrix so that \n",
    "# the only columns with a value of more than 1 in any rows are retained:\n",
    "G1S_Hist_G2M = G1S_Hist_G2M[, apply(G1S_Hist_G2M, 2, function(x) any(x > 1)) ]\n",
    "                                    \n",
    "# change the row order of the matrix:\n",
    "G1S_Hist_G2M = G1S_Hist_G2M[c(1,3,2),]\n",
    "dim(G1S_Hist_G2M)\n",
    "G1S_Hist_G2M[1:3,1:10]\n",
    "                                                    \n",
    "# # converting into a binary matrix:\n",
    "# G1S_Hist_G2M_binary = G1S_Hist_G2M[G1S_Hist_G2M > 1] = 1;\n",
    "# G1S_Hist_G2M_binary[1:4,1:10] \n",
    "\n",
    "# # Convert cells with less than 2 counts in 3rd row [ HIS ], to 0:\n",
    "# # Can't perform this operation for 1 row, so did (3,3)\n",
    "# G1S_Hist_G2M[c(3,3), ] = apply(G1S_Hist_G2M[c(3,3), ], 2, function(x) ifelse(x < 3, 0, x))\n",
    "# dim(G1S_Hist_G2M)\n",
    "# G1S_Hist_G2M[1:3,1:10]\n",
    "                           \n",
    "# Create the heatmap using pheatmap\n",
    "# https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/\n",
    "hm = pheatmap(G1S_Hist_G2M,\n",
    "         clustering_distance_cols=\"euclidean\",\n",
    "         clustering_method=\"complete\",\n",
    "         cluster_cols = TRUE, \n",
    "         cluster_rows = FALSE,\n",
    "         scale = \"none\",\n",
    "         show_rownames = TRUE,\n",
    "         show_colnames = FALSE,\n",
    "         color = colorRampPalette(c(\"white\", \"red\"))(10),\n",
    "         border_color = \"blue\",\n",
    "         cellwidth = 0.5,\n",
    "         cellheight = 20,\n",
    "         # cutree_cols = 6,\n",
    "         # fontsize_col = 8,\n",
    "         fontsize_row = 8)\n",
    "ggsave(hm, filename = \"../plots/G1S_HistoneGenes_G2M_combined_heatmap.pdf\", width=20, height=20);   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c35a429-8ce6-4949-b2c0-be518d834b52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# # circular heatmap:\n",
    "# # ordering the column by the clustered hm, since circlize does not cluster using columns:\n",
    "# hmCirc = G1S_Hist_G2M[ ,hm$tree_col$order];\n",
    "# hmCirc[1:3, 1:4]\n",
    "# col_fun1 = colorRamp2(c(0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15), c('white','#FFFFFF','#FFECEC','#FFDADA','#FFC8C8','#FFB6B6','#FFA3A3','#FF9191','#FF7F7F','#FF6D6D','#FF5B5B','#FF4848','#FF3636','#FF2424','#FF1212','#FF0000'))\n",
    "# circos.heatmap(hmCirc, col = col_fun1)\n",
    "# circos.clear()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a74af58-7308-4211-925d-dd4db7291c1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to reverse the order of columns in heatmap & mirror the heatmap above:\n",
    "col_dend = hm[[2]]\n",
    "# col_dend = rotate(col_dend, order = rev(colnames(G1_G1S_Hist_G2M)[get_order(col_dend)]))\n",
    "col_dend = rotate(col_dend, order = rev(colnames(G1S_Hist_G2M)[get_order(col_dend)]))\n",
    "\n",
    "# # change the row order of the matrix:\n",
    "# G1_G1S_Hist_G2M = G1_G1S_Hist_G2M[c(2,3,4,1),]\n",
    "# dim(G1_G1S_Hist_G2M)\n",
    "# G1_G1S_Hist_G2M[1:4,1:10]\n",
    "                               \n",
    "hmMirror = pheatmap(G1S_Hist_G2M, #G1_G1S_Hist_G2M,\n",
    "         cluster_cols = as.hclust(col_dend),\n",
    "         cluster_rows = FALSE,\n",
    "         scale = \"none\",\n",
    "         show_rownames = TRUE,\n",
    "         show_colnames = FALSE,\n",
    "         color = colorRampPalette(c(\"white\", \"red\"))(10),\n",
    "         border_color = \"none\",\n",
    "         cellwidth = 0.5,\n",
    "         cellheight = 20,\n",
    "         cutree_cols = 6.25,\n",
    "         # fontsize_col = 8,\n",
    "         fontsize_row = 8)\n",
    "ggsave(hmMirror, filename = \"../plots/G1S_HistoneGenes_G2M_combined_mirror_heatmap.pdf\", width=8, height=2); "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76f7a11b-dcd6-4571-9cfe-605fb2b02dae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot dendogram:\n",
    "plot(hmMirror$tree_col)\n",
    "abline(h=6.25, col=\"red\", lty=2, lwd=2)\n",
    "\n",
    "# cut hm column dendogram at height=7 (which results in 6 clusters)\n",
    "clusters = cutree(hmMirror$tree_col, h=6.25)\n",
    "table(clusters)\n",
    "# column IDs that belong to cluster 1\n",
    "clust1 = clusters[clusters == 1]\n",
    "str(clust1)\n",
    "clust1[1:4]\n",
    "length(clust1)\n",
    "clust2 = clusters[clusters == 2]\n",
    "length(clust2)\n",
    "clust3 = clusters[clusters == 3]\n",
    "length(clust3)\n",
    "clust4 = clusters[clusters == 4]\n",
    "length(clust4)\n",
    "# clust4\n",
    "clust5 = clusters[clusters == 5]\n",
    "length(clust5)\n",
    "# clust5\n",
    "clust6 = clusters[clusters == 6]\n",
    "length(clust6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d7708f51-22f3-41da-97fd-7cc4774bbdeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Examine reads per cell in G1S, S and G2M cells:\n",
    "# Use original count matrix:\n",
    "fullLengthCounts = readRDS(\"../data/scGROv2p8_mapq3qc_filtered_counts.rds\");\n",
    "\n",
    "G2Mcells = as.matrix(fullLengthCounts[ ,names(clust1) ])\n",
    "dim(G2Mcells)\n",
    "G1Scells = as.matrix(fullLengthCounts[ ,names(clust2) ])\n",
    "dim(G1Scells)\n",
    "Scells = as.matrix(fullLengthCounts[ ,names(clust3) ])\n",
    "dim(Scells)\n",
    "\n",
    "# dataframe for plotting:\n",
    "CCreadCounts = data.frame(\n",
    "    class = c(rep(\"G2/M\", ncol(G2Mcells)), rep(\"G1/S\", ncol(G1Scells)), rep(\"S\", ncol(Scells))),\n",
    "    ReadCounts = c(colSums(G2Mcells), colSums(G1Scells), colSums(Scells)));\n",
    "CCreadCounts[1:4, ]\n",
    "# t-test:\n",
    "t.test(ReadCounts ~ class, CCreadCounts[CCreadCounts$class %in% c(\"G1/S\",\"G2/M\"), ]);\n",
    "t.test(ReadCounts ~ class, CCreadCounts[CCreadCounts$class %in% c(\"G1/S\",\"S\"), ]);\n",
    "t.test(ReadCounts ~ class, CCreadCounts[CCreadCounts$class %in% c(\"G2/M\",\"S\"), ]);\n",
    "\n",
    "wilcox.test(ReadCounts ~ class, CCreadCounts[CCreadCounts$class %in% c(\"G1/S\",\"G2/M\"), ]);\n",
    "wilcox.test(ReadCounts ~ class, CCreadCounts[CCreadCounts$class %in% c(\"G1/S\",\"S\"), ]);\n",
    "wilcox.test(ReadCounts ~ class, CCreadCounts[CCreadCounts$class %in% c(\"G2/M\",\"S\"), ]);\n",
    "\n",
    "# I reorder the groups order : I change the order of the factor data$names\n",
    "CCreadCounts$class = factor(CCreadCounts$class, levels=c(\"G1/S\", \"S\", \"G2/M\"))\n",
    "\n",
    "# plot boxplot\n",
    "ggplot(CCreadCounts, aes(x=class, y=ReadCounts)) +\n",
    "    geom_boxplot(outlier.shape = NA, color=c(\"#ebc14c\", \"#265369\", \"#545756\"), lwd=1.5) + #, rep(\"#818664\", 2))) + #333333\n",
    "    labs(x=\"\", y=\"Reads per cell\", ) +\n",
    "       # title=\"Percentage of Cells in Each Phase\",\n",
    "    coord_cartesian(ylim = c(0, 10000))\n",
    "ggsave(filename=\"../plots/CellCycle_readCounts_boxplot.pdf\", width=4, height=4, units=\"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40170284-cbcc-475d-8233-3214923e697d",
   "metadata": {},
   "outputs": [],
   "source": [
    "median(CCreadCounts[CCreadCounts$class == \"G1/S\", ]$ReadCounts)\n",
    "median(CCreadCounts[CCreadCounts$class == \"S\", ]$ReadCounts)\n",
    "median(CCreadCounts[CCreadCounts$class == \"G2/M\", ]$ReadCounts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95fa62ea-e56e-44a0-8af2-f1568d12382b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# donut plot to show the number of cells in various cell cycles:\n",
    "CCnumbers = data.frame(count = c(length(clust2), length(clust3), length(clust1)), \n",
    "                       category = c(\"G1/S\", \"S\", \"G2/M\"))\n",
    "                    \n",
    "# Compute percentages\n",
    "CCnumbers$fraction = CCnumbers$count / sum(CCnumbers$count)\n",
    "# Compute the cumulative percentages (top of each rectangle)\n",
    "CCnumbers$ymax = cumsum(CCnumbers$fraction)\n",
    "# Compute the bottom of each rectangle\n",
    "CCnumbers$ymin = c(0, head(CCnumbers$ymax, n=-1))\n",
    "CCnumbers\n",
    "\n",
    "# I reorder the groups order : I change the order of the factor data$names\n",
    "CCnumbers$category = factor(CCnumbers$category, levels=c(\"G1/S\", \"S\", \"G2/M\"))\n",
    "\n",
    "# Make the plot\n",
    "ggplot(CCnumbers, aes(ymax=ymax, ymin=ymin, xmax=4, xmin=3, fill=category)) +\n",
    "    geom_rect() +\n",
    "    # Try to remove that to understand how the chart is built initially\n",
    "    coord_polar(theta=\"y\") + \n",
    "    # Try to remove that to see how to make a pie chart\n",
    "    xlim(c(2, 4)) +\n",
    "    theme_void() +\n",
    "    theme(legend.position = c(0.5, 0.5), \n",
    "          legend.title=element_blank(),\n",
    "          legend.text = element_text(size=14, face=\"bold\"),\n",
    "          legend.key.size = unit(1, 'cm')) +\n",
    "    scale_fill_manual(values = c(\"#ebc14c\", \"#265369\", \"#545756\"))\n",
    "\n",
    "ggsave(filename=\"../plots/CellCycle_fraction_of_cells_in_different_phases.pdf\", width=4, height=4, units=\"in\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3e31f01-a3de-4a42-80d4-b0e4f1744701",
   "metadata": {},
   "source": [
    "## Find differentially expressed genes in cell cycles:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2aec0e45-d260-49a9-bf78-06095129349d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a matrix with cells from \n",
    "CCgeneExp = as.matrix(cbind(fullLengthCounts[ ,names(clust3) ], fullLengthCounts[ ,names(clust2) ], fullLengthCounts[ ,names(clust1) ]));\n",
    "\n",
    "# # binarize the data\n",
    "# CCgeneExp[CCgeneExp>1] = 1;\n",
    "\n",
    "# select genes only\n",
    "CCgeneExp = CCgeneExp[ substr(rownames(CCgeneExp), 0, 3) == \"GN-\", ]\n",
    "# remove \"GN-\"\n",
    "rownames(CCgeneExp) = sub(\"GN-\", \"\", rownames(CCgeneExp))\n",
    "rownames(CCgeneExp)[1:4]\n",
    "\n",
    "# find rownames that contain \"Hist\" as part of the names and convert all of them to \"Histones\n",
    "rownames(CCgeneExp)[grepl(\"Hist\", rownames(CCgeneExp))] <- \"Histones\"\n",
    "dim(CCgeneExp)\n",
    "\n",
    "# sum rows that have the same names, in this case Histones\n",
    "CCgeneExp = rowsum(CCgeneExp, row.names(CCgeneExp))\n",
    "dim(CCgeneExp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57e5f19a-ebbf-4ec5-875a-62f76178df99",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create Seurat object:\n",
    "s_obj = CreateSeuratObject( counts = CCgeneExp)\n",
    "s_obj@assays\n",
    "s_obj\n",
    "\n",
    "# change the orig.ident from \"SeuratObject' to cell cycle stages\n",
    "s_obj@meta.data$orig.ident = as.factor(c(rep(\"G1/S\", length(clust3)), rep(\"S\", length(clust2)), rep(\"G2/M\", length(clust1))))\n",
    "# Set the identity of your cells to the desired column\n",
    "Idents(object = s_obj) = s_obj@meta.data$'orig.ident'\n",
    "s_obj@meta.data[1:4, ]\n",
    "\n",
    "# normalize the data\n",
    "s_obj_norm = ScaleData(s_obj)\n",
    "    \n",
    "# markers differentially expressed in each identity group by comparing it to all of the others\n",
    "CCdiffGenes = FindAllMarkers(s_obj_norm, min.pct = 0.1, test.use = \"poisson\", min.diff.pct = 0.1, logfc.threshold = 0.2, only.pos = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "699a7675-9dd6-439e-bbf7-03cc4cf59e4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "CCdiffGenesF = CCdiffGenes %>%\n",
    "    # filter on log2 fold change or the genes used to classify cell cycles in the first place:\n",
    "    filter( avg_log2FC >= 0.33 | rownames(CCdiffGenes) %in% c(G1S_Riba, G2M_Riba));\n",
    "dim(CCdiffGenesF)\n",
    "CCdiffGenesF[1:4]\n",
    "summary(CCdiffGenesF$cluster)\n",
    "fwrite(CCdiffGenesF, file=\"../data/CellCycle_genes_discovered_by_Seurat_FindAllFeatures.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2ed6b0ab-911e-4a9c-9454-1eb44551fe1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# z-score plot of differentially expressed genes:\n",
    "\n",
    "CCgeneMeanExp = data.frame(\n",
    "    name = rownames(G2Mcells),\n",
    "    G1Scells = rowMeans(G1Scells > 0),\n",
    "    Scells = rowMeans(Scells > 0),\n",
    "    G2Mcells = rowMeans(G2Mcells > 0));\n",
    "\n",
    "CCgeneMeanExp = CCgeneMeanExp %>%\n",
    "    filter( substr(name, 0, 3) == \"GN-\" )\n",
    "rownames(CCgeneMeanExp) = sub(\"GN-\", \"\", CCgeneMeanExp$name)\n",
    "CCgeneMeanExp$name = NULL\n",
    "dim(CCgeneMeanExp)\n",
    "CCgeneMeanExp[1:4, ]\n",
    "\n",
    "# convert into a matrix for changing rownames and summing by rownames \n",
    "CCgeneMeanExp = as.matrix(CCgeneMeanExp)\n",
    "# find rownames that contain \"Hist\" as part of the names and convert all of them to \"Histones\n",
    "rownames(CCgeneMeanExp)[grepl(\"Hist\", rownames(CCgeneMeanExp))] <- \"Histones\"\n",
    "# sum rows that have the same names, in this case Histones\n",
    "CCgeneMeanExp = rowsum(CCgeneMeanExp, row.names(CCgeneMeanExp))\n",
    "dim(CCgeneMeanExp)\n",
    "\n",
    "# Filter CCgeneMeanExp for genes differentially expressed genes in cell-cycle:\n",
    "CCgeneMeanExp = CCgeneMeanExp[rownames(CCgeneMeanExp) %in% rownames(CCdiffGenesF), ]\n",
    "dim(CCgeneMeanExp)\n",
    "\n",
    "# calculate z-score\n",
    "CCgeneMeanExp = t(apply(CCgeneMeanExp, 1, scale))\n",
    "colnames(CCgeneMeanExp) = colnames(CCgeneMeanExp)\n",
    "CCgeneMeanExp\n",
    "\n",
    "mycol = colorRampPalette(c('#4098b5', '#72cbe8', 'white', '#e88f72', '#b55c40'))(39) #\n",
    "mybreaks = c(seq(-2, -1.1, length=10), seq(-1, 1, length=20), seq(1.1, 2, length=10))\n",
    "\n",
    "pdf(\"../plots/cell_cycle_FinadAllMarkers_MeanExpressingCells_zscore_hm2.pdf\", width = 4, height = 16)\n",
    "heatmap.2(CCgeneMeanExp, \n",
    "          dendrogram=c(\"row\"), \n",
    "          Rowv=T, \n",
    "          Colv=F, \n",
    "          scale=\"none\", \n",
    "          sepwidth=c(0.01,0.01), \n",
    "          sepcolor=\"grey\", \n",
    "          rowsep=0:nrow(CCgeneMeanExp), \n",
    "          colsep=0:ncol(CCgeneMeanExp), \n",
    "          key=T, \n",
    "          density.info=\"none\", \n",
    "          trace=c(\"none\"), \n",
    "          cexCol=0.9, \n",
    "          labRow=rownames(CCgeneMeanExp), \n",
    "          labCol=colnames(CCgeneMeanExp), \n",
    "          breaks = mybreaks, \n",
    "          col = mycol,  \n",
    "          margins = c(12, 8))\n",
    "dev.off() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76dab742-b487-443f-94cc-1509b1f83c41",
   "metadata": {},
   "outputs": [],
   "source": [
    "G1S_Riba = c(\"Orc1\", \"Ccne1\", \"Ccne2\", \"Mcm6\")\n",
    "G2M_Riba = c(\"Wee1\", \"Cdk1\", \"Ccnf\", \"Nusap1\", \"Aurka\", \"Ccna2\", \"Ccnb2\")\n",
    "rownames(CCgeneMeanExp[rownames(CCgeneMeanExp) %in% G1S_Riba, ])\n",
    "rownames(CCgeneMeanExp[rownames(CCgeneMeanExp) %in% G2M_Riba, ])\n",
    "\n",
    "# https://github.com/hbc/tinyatlas\n",
    "cc_markers = fread(\"../data/Cell_cycle_marker_genes.csv\");\n",
    "length(cc_markers$gene_name)\n",
    "cc_markers[cc_markers$gene_name %in% rownames(CCgeneMeanExp)]$gene_name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "513194bb-a3f0-4643-a34c-fd22e471ca77",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load libraries needed for this script\n",
    "\n",
    "# test = cbind(A1[,9], A2[,9], B1[,9], B2[,9])\n",
    "# # Create a talbe with six columns which contains the gene body read counts.\n",
    "\n",
    "# row.names(test) = A1[,5]\n",
    "# # gene names are assigned as row names.\n",
    "\n",
    "# colnames(test) = c(\"A1\", \"A2\", \"B1\", \"B2\")\n",
    "# # specifying column names.\n",
    "\n",
    "\n",
    "# testDesign = data.frame(row.names = colnames(test), condition = as.factor(c(rep(\"untreated\",2),rep(\"treated\",2))))\n",
    "# # make the metadata table (this is well explained in the PDF)\n",
    "\n",
    "# print(testDesign)\n",
    "# # provides information about the metadata.\n",
    "\n",
    "# print(\"this is how the table looks before normalization\")\n",
    "# print(head(test))\n",
    "# # shows how the test file looks like.\n",
    "\n",
    "# cds = DESeqDataSetFromMatrix(countData = test, colData = testDesign, design = ~ condition)\n",
    "# # since test is a matrix of counts and the column information, we can construct a DESeqDataSet using DESeqDataSetFromMatrix function\n",
    "\n",
    "# print(cds)\n",
    "# # gives info about cds\n",
    "\n",
    "# cds$condition = relevel(cds$condition, \"untreated\")\n",
    "# # It is important to supply levels (otherwise the levels are chosen inalphabetical order) and to put the “control” or “untreated” level as the first element (”base level”), so that the log2 fold changes produced by default will be the expected comparison against the base level\n",
    "\n",
    "# sizeFactors(cds) = c(nf1, nf2, nf3, nf4)\n",
    "# # instead of using the estimateSizeFactors to allow DESeq to normalize the data, I input my normalization factor calculated from the reads that map to the drosophila genome.\n",
    "\n",
    "# print(\"the normalization factors are:\")\n",
    "# print(sizeFactors(cds))\n",
    "\n",
    "# print(\"this is how the table looks after normalization\")\n",
    "# print(head(counts(cds,normalized=TRUE)))\n",
    "# # to visualize the normalized read counts.\n",
    "\n",
    "\n",
    "# DESeq_cds = DESeq(cds)\n",
    "# # this is the function that performs DESeq. Using the supplied sizeFactors, it estimates dispersions, gene-wise dispersion estimates, mean-dispersion relationship, final dispersion estimates, and performs fitting model and testing\n",
    "\n",
    "# res = results(DESeq_cds)\n",
    "# # table with the necessary informations like: baseMean log2FoldChange     lfcSE       stat      pvalue      padj\n",
    "\n",
    "# sigGenes = res[(res$padj < padjusted) & !is.na(res$padj),]\n",
    "# # this subsets the rows based on the supplied adjusted pvalue and the absence of NA values\n",
    "\n",
    "# print(\" the adjuseted p value to be used to filter significantly changed genes from DESeq2 is:\")\n",
    "# print(padjusted)\n",
    "\n",
    "# print(\"# of significantly changed genes with given adj-pValue:\")\n",
    "# print(nrow(sigGenes))\n",
    "\n",
    "# pdf(file = paste(\"PC_KPC001S_inVitro_vs_PC_HY19636_inVitro_Volcanoplot\", \".pdf\", sep = ''), width = 8, height = 12, bg = 'white')\n",
    "# par(font.axis = 2)\n",
    "# par(font.lab=2)\n",
    "# par(cex.lab=1.3)\n",
    "# par(cex.axis=1.3)\n",
    "# par(cex=2)\n",
    "# par(cex.main=1.3)\n",
    "# par(mar=c(5,5,4,2))\n",
    "# plot(res[,2], -log10(res[,6]), pch=20, main=\"PC_KPC001S_inVitro_vs_PC_HY19636_inVitro\", xlab=\"Log2 [ Fold Change ]\", ylab=\"-Log10 [ p_adj ]\")\n",
    "# points(sigGenes[,2], -log10(sigGenes[,6]), pch=20, col=\"red\")\n",
    "# dev.off()\n",
    "# # making volcano plot\n",
    "\n",
    "# write.table(sigGenes, file = \"DESeq2-output.txt\", row.names=T, sep='\\t', quote=F, na=\"\")\n",
    "# # write the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "caf642da-2daa-434d-a60f-49c0e87bd39e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# trying to see if SEs and SEgenes are expressed in G1S, S, or G2M phase:\n",
    "SEsCollapsed = as.matrix(newCounts[ c(\"INT_STITCHED_4179\", \"INT_STITCHED_5044\", \"INT_STITCHED_2745\", \n",
    "                                      \"INT_STITCHED_5427\", \"INT_STITCHED_1210\", \"INT_STITCHED_4657\", \n",
    "                                      \"INT_STITCHED_1626\",\n",
    "                                      \"GN-Sall1\", \"GN-Tet1\", \"GN-Med13l\", \n",
    "                                      \"GN-Ranbp17\", \"GN-Cbfa2t2\", \"GN-Ooep\", \n",
    "                                      \"GN-Mcl1\"), ])\n",
    "# SEsCollapsed = as.matrix(newCounts[ c(vSEassignment$name, vSEassignment$assignedGene), ])\n",
    "\n",
    "\n",
    "# provide rownames reflecting their origin for all rows\n",
    "rownames(SEsCollapsed) = c(rep(\"SEs\", each=7),\n",
    "                           rep(\"SEsgenes\", each=7));\n",
    "dim(SEsCollapsed)\n",
    "# # converting into a binary matrix:\n",
    "# SEsCollapsed[SEsCollapsed > 1] = 1;\n",
    "\n",
    "# perform rowsums by rownames\n",
    "SEsCollapsed = rowsum(SEsCollapsed, row.names(SEsCollapsed))\n",
    "dim(SEsCollapsed)\n",
    "# order the columns by the column order of hm above\n",
    "SEsCollapsed = SEsCollapsed[ ,hmMirror$tree_col$order];\n",
    "dim(SEsCollapsed)\n",
    "SEsCollapsed[1:2, 1:10]\n",
    "\n",
    "# convert all values below 2 to 0:\n",
    "SEsCollapsed[SEsCollapsed < 2] = 0;\n",
    "\n",
    "# Create the heatmap using pheatmap\n",
    "# https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/\n",
    "hm2 = pheatmap(SEsCollapsed,\n",
    "         cluster_cols = FALSE, \n",
    "         cluster_rows = FALSE,\n",
    "         scale = \"none\",\n",
    "         show_rownames = TRUE,\n",
    "         show_colnames = FALSE,\n",
    "         color = colorRampPalette(c(\"white\", \"red\"))(15),\n",
    "         border_color = NA,\n",
    "         cellwidth = 0.5,\n",
    "         cellheight = 20,\n",
    "         # fontsize_col = 8,\n",
    "         fontsize_row = 8)\n",
    "ggsave(hm2, filename = \"../plots/verifiedSEs_heatmap.pdf\", width=20, height=20);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c13e7673-98b2-413f-9faf-bdd1dbb356df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# trying to see if SEs and SEgenes are expressed in G1S, S, or G2M phase:\n",
    "OSNKCollapsed = as.matrix(newCounts[ c(\"INT_STITCHED_7784\", \"INT_STITCHED_1482\", \"INT_STITCHED_3347\", \n",
    "                                       \"INT_STITCHED_3348\", \"INT_STITCHED_3349\", \"INT_STITCHED_1973\",\n",
    "                                      \"GN-Pou5f1\", \"GN-Sox2\", \"GN-Nanog\", \"GN-Klf4\"), ])\n",
    "\n",
    "# provide rownames reflecting their origin for all rows\n",
    "rownames(OSNKCollapsed) = c(rep(\"OSNKSEs\", each=6),\n",
    "                               rep(\"OSNKsgenes\", each=4));\n",
    "dim(OSNKCollapsed)\n",
    "# # converting into a binary matrix:\n",
    "# SEsCollapsed[SEsCollapsed > 1] = 1;\n",
    "\n",
    "# perform rowsums by rownames\n",
    "OSNKCollapsed = rowsum(OSNKCollapsed, row.names(OSNKCollapsed))\n",
    "dim(OSNKCollapsed)\n",
    "# order the columns by the column order of hm above\n",
    "OSNKCollapsed = OSNKCollapsed[ ,hmMirror$tree_col$order];\n",
    "dim(OSNKCollapsed)\n",
    "OSNKCollapsed[1:2, 1:10]\n",
    "\n",
    "# convert all values below 2 to 0:\n",
    "OSNKCollapsed[OSNKCollapsed < 2] = 0;\n",
    "\n",
    "# Create the heatmap using pheatmap\n",
    "# https://davetang.org/muse/2018/05/15/making-a-heatmap-in-r-with-the-pheatmap-package/\n",
    "hm2 = pheatmap(OSNKCollapsed,\n",
    "         cluster_cols = FALSE, \n",
    "         cluster_rows = FALSE,\n",
    "         scale = \"none\",\n",
    "         show_rownames = TRUE,\n",
    "         show_colnames = FALSE,\n",
    "         color = colorRampPalette(c(\"white\", \"red\"))(15),\n",
    "         border_color = NA,\n",
    "         cellwidth = 0.5,\n",
    "         cellheight = 20,\n",
    "         # fontsize_col = 8,\n",
    "         fontsize_row = 8)\n",
    "ggsave(hm2, filename = \"../plots/OSNKSEs_heatmap.pdf\", width=20, height=20);"
   ]
  },
  {
   "cell_type": "raw",
   "id": "410bce3a-1619-43ea-8d7d-21aa9c8d5b62",
   "metadata": {},
   "source": [
    "# Get the range of expression of vefifiedSEGenes:\n",
    "verifiedSEgenesON = counts[ sub(\"CGN\", \"GN\", SEassignment[SEassignment$Location != \"Intron\", 2]), ]\n",
    "dim(verifiedSEgenesON)\n",
    "summary(rowMeans(verifiedSEgenesON > 0))\n",
    "\n",
    "# Subset SEneighbors Genes such that their expression is similar to verifiedSEgenes\n",
    "SEneighborsON = counts[ names(SEneighbors), ]\n",
    "dim(SEneighborsON)\n",
    "summary(rowMeans(SEneighborsON > 0))\n",
    "# filter SEneighbors with expression within the verfiedSEgenes:\n",
    "SEneighborsON = SEneighborsON[rowMeans(SEneighborsON) > min(rowMeans(verifiedSEgenesON > 0)) & \n",
    "                     rowMeans(SEneighborsON) < max(rowMeans(verifiedSEgenesON > 0)), ]\n",
    "dim(SEneighborsON)\n",
    "summary(rowMeans(SEneighborsON > 0))\n",
    "\n",
    "# how many SEneighbor genes inside and outside of TAD survive the expression filter?\n",
    "test = SEneighbors %>%\n",
    "    filter( names %in% rownames(SEneighborsON) );\n",
    "summary(as.factor(test$TADstatus))"
   ]
  },
  {
   "cell_type": "raw",
   "id": "c4808ba0-10cd-49aa-a412-a5ea5063d0e4",
   "metadata": {},
   "source": [
    "# exclude SEneighbors features with expression beyond the range of \n",
    "counts = counts[ rownames(counts) %in% c(names(verifiedSE),\n",
    "                                       names(histGenes),\n",
    "                                       # verifiedGenes only so that they are not dropped off from verifiedSEgenesON:\n",
    "                                       rownames(verifiedSEgenesON),\n",
    "                                       rownames(SEneighborsON)\n",
    "                                       ), ]\n",
    "dim(counts)\n",
    "rownames(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b33cd1c4-af73-464f-92cb-3fc902d08035",
   "metadata": {},
   "outputs": [],
   "source": [
    "data.frame(\n",
    "    x=rowMeans(counts>0)\n",
    ") %>%\n",
    "    ggplot( aes(x=x) ) +\n",
    "    geom_histogram(binwidth=0.05) +\n",
    "    scale_x_log10() +\n",
    "    ggtitle(\"Reads per Feature\") +\n",
    "    xlab(\"Reads per cell\") +\n",
    "    ylab(\"Number of features\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6755b9bb-c508-4de1-b833-09f5e4156b7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only retain features with counts in 0.5% or more cells\n",
    "observed = counts[ rowMeans(counts>0) >= 0.005, ];\n",
    "dim(observed)\n",
    "\n",
    "test = substr(rownames(observed), 0, 3)\n",
    "summary(as.factor(test))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4db2fef8-d2b6-43bb-9297-d0a1193a4ba5",
   "metadata": {},
   "source": [
    "## ExG Correlation:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "530e3f09-7ef9-4fd6-ba07-adbcf4698763",
   "metadata": {},
   "outputs": [],
   "source": [
    "# transforming the matrix to calculate corr, hmisc package reqires genes in columns and cells in rows\n",
    "obsx = as.matrix(t(observed))\n",
    "dim(obsx)\n",
    "obsx[1:5,1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ead3d130-8597-4c27-9b6f-c3f3dd6d075d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ++++++++++++++++++++++++++++\n",
    "# flattenCorrMatrix\n",
    "# ++++++++++++++++++++++++++++\n",
    "# cormat : matrix of the correlation coefficients\n",
    "# pmat : matrix of the correlation p-values\n",
    "flattenCorrMatrix = function(cormat, pmat) {\n",
    "  ut = upper.tri(cormat)\n",
    "  data.frame(\n",
    "    Gene = rownames(cormat)[row(cormat)[ut]],\n",
    "    Enhancer = rownames(cormat)[col(cormat)[ut]],\n",
    "    corr = (cormat)[ut],\n",
    "    pVal = pmat[ut]\n",
    "    )\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1571614d-a10e-4d54-b5d9-b648f4ccb1ff",
   "metadata": {},
   "source": [
    "#### Pearson correlation on non-binarized data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db9ea1f1-544b-4a9e-8cc2-af6f598119b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use hmisc package to calculate correlation and p-value:\n",
    "corrP = rcorr(obsx, type = \"pearson\")\n",
    "\n",
    "# corrPflat = flattenCorrMatrix( corrP$r, corrP$P ) %>%\n",
    "#     filter( substr(Gene, 0, 3) == \"GN-\" & substr(Enhancer, 0, 3) != \"GN-\" ) %>%\n",
    "#     # filter rows such that both Gene and Enhancer were assigned pairs (based on same row in SEcenter):\n",
    "#     filter( paste0(sub(\"_[^_]+$\", \"\", Gene), sub(\"_[^_]+$\", \"\", Enhancer)) %in% paste0(SEassignment$assignedGene, SEassignment$name) ) %>%\n",
    "#     # filter(as.character(seqnames(SEcenter[Gene])) == as.character(seqnames(SEcenter[Enhancer])));\n",
    "#     mutate( pAdj = p.adjust(pVal, method = \"fdr\") ) %>%\n",
    "#     mutate( corrRank = rank(corr) ) %>%\n",
    "#     mutate( pValRank = rank(pVal) );\n",
    "\n",
    "# fwrite(corrPflat, file=\"../data/scGROv2p8_SE_CRISPRverified_5Kbins_pearson_correlation.csv.gz\")\n",
    "# dim(corrPflat)\n",
    "# corrPflat[1:4, ]\n",
    "\n",
    "# corrPlot = corrPflat %>%\n",
    "#     ggplot(aes(x = corrRank, y = corr)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     ylim(-0.1, 0.3) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG Pearson Corr\") +\n",
    "#     xlab(\"Rank\") +\n",
    "#     ylab(\"Pearson corr. coeff.\");\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_pearson_corrCoeff_rank.png\", width=4, height=4);\n",
    "\n",
    "# corrPlot = corrPflat %>%\n",
    "#     ggplot(aes(x = pValRank, y = pVal)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG Pearson Corr\") +\n",
    "#     xlab(\"Rank\") +\n",
    "#     ylab(\"Pearson corr. p-value\");\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_pearson_corrPvalue_rank.png\", width=4, height=4);\n",
    "\n",
    "# corrPlot = corrPflat %>%\n",
    "#     ggplot(aes( x = corr, y = pVal)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG Pearson Corr\") +\n",
    "#     xlab(\"Pearson corr. coeff.\") +\n",
    "#     ylab(\"Pearson corr. p-value\") ;\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_pearson_corrCoeff_vs_pvalue.png\", width=4, height=4)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "aa04b4d6-89f2-4454-8a26-bcd240eb290c",
   "metadata": {},
   "source": [
    "pdf('../plots/scGRO_SE_CRISPRverified_5Kbins_pearson_corrCoeff_heatmap.pdf',width=20,height=20)\n",
    "heatmap.2(corrP$r,\n",
    "    dendrogram='both',\n",
    "    scale='none',\n",
    "    col=colorRampPalette(c(\"blue\", \"white\", \"red\"))(n = 100),\n",
    "    trace='none',\n",
    "    keysize=2,\n",
    "    density.info='none',\n",
    "    labRow=FALSE,\n",
    "    labCol=FALSE)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63fb3203-0719-4a99-9810-de8293556cf1",
   "metadata": {},
   "source": [
    "#### Spearman correlation on non-binarized data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "94a7a89d-8911-43e8-9081-f8fcb927be01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use hmisc package to calculate correlation and p-value:\n",
    "corrS = rcorr(obsx, type = \"spearman\")\n",
    "\n",
    "# corrSflat = flattenCorrMatrix( corrS$r, corrS$P ) %>%\n",
    "#   filter( substr(Gene, 0, 3) == \"GN-\" & substr(Enhancer, 0, 3) != \"GN-\" ) %>%\n",
    "#     # filter rows such that both Gene and Enhancer were assigned pairs (based on same row in SEcenter):\n",
    "#     filter( paste0(sub(\"_[^_]+$\", \"\", Gene), sub(\"_[^_]+$\", \"\", Enhancer)) %in% paste0(SEassignment$assignedGene, SEassignment$name) ) %>%\n",
    "#     # filter(as.character(seqnames(SEcenter[Gene])) == as.character(seqnames(SEcenter[Enhancer])));\n",
    "#     mutate( pAdj = p.adjust(pVal, method = \"fdr\") ) %>%\n",
    "#     mutate( corrRank = rank(corr) ) %>%\n",
    "#     mutate( pValRank = rank(pVal) );\n",
    "\n",
    "# fwrite(corrSflat, file=\"../data/scGROv2p8_SE_CRISPRverified_5Kbins_spearman_correlation.csv.gz\")\n",
    "# dim(corrSflat)\n",
    "# corrSflat[1:4, ]\n",
    "\n",
    "# corrPlot = corrSflat %>%\n",
    "#     ggplot(aes(x = corrRank, y = corr)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     ylim(-0.1, 0.3) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG spearman Corr\") +\n",
    "#     xlab(\"Rank\") +\n",
    "#     ylab(\"Spearman corr. coeff.\");\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_spearman_corrCoeff_rank.png\", width=4, height=4);\n",
    "\n",
    "# corrPlot = corrSflat %>%\n",
    "#     ggplot(aes(x = pValRank, y = pVal)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG spearman Corr\") +\n",
    "#     xlab(\"Rank\") +\n",
    "#     ylab(\"Spearman corr. p-value\");\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_spearman_corrPvalue_rank.png\", width=4, height=4);\n",
    "\n",
    "# corrPlot = corrSflat %>%\n",
    "#     ggplot(aes( x = corr, y = pVal)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG spearman Corr\") +\n",
    "#     xlab(\"Spearman corr. coeff.\") +\n",
    "#     ylab(\"Spearman corr. p-value\") ;\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_spearman_corrCoeff_vs_pvalue.png\", width=4, height=4)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "739c7fbf-d2b9-4db0-b999-1d2d479608c5",
   "metadata": {},
   "source": [
    "pdf('../plots/scGRO_SE_CRISPRverified_5Kbins_spearman_corrCoeff_heatmap.pdf',width=20,height=20)\n",
    "heatmap.2(corrS$r,\n",
    "    dendrogram='both',\n",
    "    scale='none',\n",
    "    col=colorRampPalette(c(\"blue\", \"white\", \"red\"))(n = 100),\n",
    "    trace='none',\n",
    "    keysize=2,\n",
    "    density.info='none',\n",
    "    labRow=FALSE,\n",
    "    labCol=FALSE)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07cd5025-d508-44f2-9490-4a8319587dcf",
   "metadata": {},
   "source": [
    "#### Pearson/Spearman on binary matrix\n",
    "##### Pearson and Spearman on binary data gives same result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e189f2e-4667-45c2-9eb7-f1d0313cf9ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Binarize observed counts\n",
    "obsb = observed;\n",
    "obsb[obsb>1] = 1;\n",
    "# transforming the matrix to calculate corr, hmisc package reqires genes in columns and cells in rows\n",
    "obsb = as.matrix(t(obsb))\n",
    "dim(obsb)\n",
    "obsb[1:5,1:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e7c4d8c-7c1c-4340-a470-60727473d9db",
   "metadata": {},
   "outputs": [],
   "source": [
    "# use hmisc package to calculate correlation and p-value:\n",
    "corrB = rcorr(obsb, type = \"pearson\")\n",
    "\n",
    "# corrBflat = flattenCorrMatrix( corrB$r, corrB$P ) %>%\n",
    "#     filter( substr(Gene, 0, 3) == \"GN-\" & substr(Enhancer, 0, 3) != \"GN-\" ) %>%\n",
    "#     # filter rows such that both Gene and Enhancer were assigned pairs (based on same row in SEcenter):\n",
    "#     filter( paste0(sub(\"_[^_]+$\", \"\", Gene), sub(\"_[^_]+$\", \"\", Enhancer)) %in% paste0(SEassignment$assignedGene, SEassignment$name) ) %>%\n",
    "#     # filter(as.character(seqnames(SEcenter[Gene])) == as.character(seqnames(SEcenter[Enhancer])));\n",
    "#     # filter genes that have SE in intorn:\n",
    "#     mutate( pAdj = p.adjust(pVal, method = \"fdr\") ) %>%\n",
    "#     mutate( corrRank = rank(corr) ) %>%\n",
    "#     mutate( pValRank = rank(pVal) );\n",
    "\n",
    "# fwrite(corrBflat, file=\"../data/scGROv2p8_SE_CRISPRverified_5Kbins_binary_correlation.csv.gz\")\n",
    "# dim(corrBflat)\n",
    "# corrBflat[1:4, ]\n",
    "\n",
    "# corrPlot = corrBflat %>%\n",
    "#     ggplot(aes(x = corrRank, y = corr)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     ylim(-0.1, 0.3) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG Binary Corr\") +\n",
    "#     xlab(\"Rank\") +\n",
    "#     ylab(\"Binary corr. coeff.\");\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_binary_corrCoeff_rank.png\", width=4, height=4);\n",
    "\n",
    "# corrPlot = corrBflat %>%\n",
    "#     ggplot(aes(x = pValRank, y = pVal)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG Binary Corr\") +\n",
    "#     xlab(\"Rank\") +\n",
    "#     ylab(\"Binary corr. p-value\");\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_binary_corrPvalue_rank.png\", width=4, height=4);\n",
    "\n",
    "# corrPlot = corrBflat %>%\n",
    "#     ggplot(aes( x = corr, y = pVal)) +\n",
    "#     geom_point(pch = 16, aes(color = pAdj)) +\n",
    "#     scale_color_viridis() +\n",
    "#     ggtitle(\"ExG Binary Corr\") +\n",
    "#     xlab(\"Binary corr. coeff.\") +\n",
    "#     ylab(\"Binary corr. p-value\") ;\n",
    "# corrPlot\n",
    "# # ggsave(corrPlot, filename = \"../plots/scGRO_SE_CRISPRverified_5Kbins_binary_corrCoeff_vs_pvalue.png\", width=4, height=4)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "fb9cd7ff-efbf-469d-a531-ceec7132f06a",
   "metadata": {},
   "source": [
    "pdf('../plots/scGRO_SE_CRISPRverified_5Kbins_binary_corrCoeff_heatmap.pdf',width=20,height=20)\n",
    "heatmap.2(corrB$r,\n",
    "    dendrogram='both',\n",
    "    scale='none',\n",
    "    col=colorRampPalette(c(\"blue\", \"white\", \"red\"))(n = 100),\n",
    "    trace='none',\n",
    "    keysize=2,\n",
    "    density.info='none',\n",
    "    labRow=FALSE,\n",
    "    labCol=FALSE)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e3dec38a-45ba-4bad-b5f6-d42d2bf21c0c",
   "metadata": {},
   "source": [
    "### Without filtering for assigned SE to gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b960b4e1-ccce-4aab-ac38-451e6190f5c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "corrPflatRaw = flattenCorrMatrix( corrP$r, corrP$P )\n",
    "corrSflatRaw = flattenCorrMatrix( corrS$r, corrS$P )\n",
    "corrBflatRaw = flattenCorrMatrix( corrB$r, corrB$P )\n",
    "# join two tables using two columns\n",
    "corrRaw = left_join(   corrPflatRaw, corrSflatRaw, by=c(\"Gene\", \"Enhancer\"), suffix = c(\".P\", \".S\") ) %>%\n",
    "          left_join(., corrBflatRaw, by=c(\"Gene\", \"Enhancer\") ) %>%\n",
    "          mutate( corrRank = rank(corr) ) %>%\n",
    "          mutate( pValRank = rank(pVal) ) %>%\n",
    "          mutate( GeneBin     = sub(\"^.*_\", \"\", Gene) ) %>%\n",
    "          mutate( EnhancerBin = sub(\"^.*_\", \"\", Enhancer) ) %>%\n",
    "          mutate( Enh_Gene = as.numeric(EnhancerBin) - as.numeric(GeneBin) ) %>%\n",
    "          mutate( pAdj = p.adjust(pVal, method = \"fdr\") ) %>%\n",
    "          # na.omit();\n",
    "          arrange( Gene );\n",
    "    \n",
    "dim(corrRaw)\n",
    "table(substr(corrRaw$Gene, 0, 3));\n",
    "table(substr(corrRaw$Enhancer, 0, 3));\n",
    "corrRaw[1:4, ]\n",
    "\n",
    "fwrite(corrRaw, file=\"../data/scGROv2p8_ExG_CRISPRverified_5Kbins_ALL_correlation.csv.gz\")\n",
    "\n",
    "# # lookup gene attributes\n",
    "# Gene = allfeatures[ corrRaw$Gene ];\n",
    "# Enhancer = allfeatures[ corrRaw$Enhancer ];\n",
    "# corrRaw$lengthG = width(Gene);\n",
    "# corrRaw$lengthE = width(Enhancer);\n",
    "# corrRaw$strG    = as.character(strand(Gene));\n",
    "# corrRaw$strE    = as.character(strand(Enhancer));\n",
    "\n",
    "# # compute distance between gene promoters\n",
    "# proG = resize(Gene, width=1, fix=\"start\");\n",
    "# proE = resize(Enhancer, width=1, fix=\"center\");\n",
    "# strand(proG) = \"*\";\n",
    "# strand(proE) = \"*\";\n",
    "# # corrRaw$distance = width(pgap( proG, proE ));\n",
    "# corrRaw$distance = ifelse(as.character(seqnames(proG)) == as.character(seqnames(proE)), width(pgap( proG, proE )), \"NA\");\n",
    "# corrRaw[1:4, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28b8cd6e-af30-44b0-815c-6572711def45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CRISPR verified genes and thier SEs bins:\n",
    "corrRawF = corrRaw %>%\n",
    "    # filter for GN- and thier SEs:\n",
    "    filter( substr(Gene, 0, 3) == \"GN-\" & substr(Enhancer, 0, 3) == \"INT\" ) %>%\n",
    "    # Ensure that the CGN and SEs are proper pairs:\n",
    "    filter( paste0(sub(\"_[^_]+$\", \"\", Gene), sub(\"_[^_]+$\", \"\", Enhancer)) %in% paste0(SEassignment$assignedGene, SEassignment$name) ) %>%\n",
    "    filter( pAdj < 0.05 ) %>%\n",
    "    arrange( Gene );\n",
    "dim(corrRawF)\n",
    "corrRawF\n",
    "corrRawF$Enh_Gene\n",
    "\n",
    "SE_gene_bin_hist = corrRawF %>%\n",
    "    ggplot(aes(x = Enh_Gene)) +\n",
    "    geom_histogram() +\n",
    "    geom_histogram(bins = 13, position = \"identity\", color = \"white\") +\n",
    "    geom_density() +\n",
    "    xlim(-6, 6) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    ggtitle(\"Binary (SE-Gene)\") +\n",
    "    xlab(\"Bin difference\") +\n",
    "    ylab(\"Number of pairs\") +\n",
    "    scale_fill_manual(values=c(\"#39568CFF\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 14, face=\"bold\"))\n",
    "SE_gene_bin_hist\n",
    "\n",
    "summary_bin_hist = data.frame(\n",
    "    Neg = sum(corrRawF$Enh_Gene < 0)/length(corrRawF$Enh_Gene),\n",
    "    Zero = sum(corrRawF$Enh_Gene == 0)/length(corrRawF$Enh_Gene),\n",
    "    Pos = sum(corrRawF$Enh_Gene > 0)/length(corrRawF$Enh_Gene)\n",
    "    ) %>%\n",
    "    pivot_longer(1:3, names_to = 'Category', values_to = 'Incidence') %>%\n",
    "    # use fct_inorder from forcats (part of tidyverse) to keep the original order of the data:\n",
    "    ggplot(aes(x = fct_inorder(Category), y = Incidence)) +\n",
    "    # geom_bar(stat=\"identity\", fill=c(\"#9e9a75\",\"gray\", \"#41533b\")) +\n",
    "    geom_col(fill=c(\"#9e9a75\",\"gray\", \"#41533b\")) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    ggtitle(\"Binary (SE-Gene)\") +\n",
    "    xlab(\"Class\") +\n",
    "    ylab(\"Fraction of pairs\") +\n",
    "    # scale_fill_manual(values=c(\"#9e9a75\",\"gray\", \"#41533b\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 14, face=\"bold\"))\n",
    "summary_bin_hist"
   ]
  },
  {
   "cell_type": "raw",
   "id": "32bab0b5-8304-4212-bda0-2880fbea1c05",
   "metadata": {},
   "source": [
    "Gene\tEnhancer\tcorr.P\tpVal.P\tcorr.S\tpVal.S\tcorr\tpVal\tcorrRank\tpValRank\tGeneBin\tEnhancerBin\tEnh_Gene\tpAdj\n",
    "<chr>\t<chr>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<chr>\t<chr>\t<dbl>\t<dbl>\n",
    "GN-Cbfa2t2_1\tINT_STITCHED_1210_1\t0.059338\t2.3145e-03\t0.063212\t1.1708e-03\t0.063316\t1.1489e-03\t159932\t9140.5\t1\t1\t0\t2.1250e-02\n",
    "GN-Nanog_1\tINT_STITCHED_3349_2\t0.056944\t3.4614e-03\t0.061288\t1.6499e-03\t0.061419\t1.6124e-03\t158822\t10250.0\t1\t2\t1\t2.6588e-02\n",
    "GN-Ooep_1\tINT_STITCHED_4657_1\t0.078581\t5.4063e-05\t0.075576\t1.0340e-04\t0.075328\t1.0898e-04\t164644\t4427.5\t1\t1\t0\t4.1609e-03\n",
    "GN-Ooep_2\tINT_STITCHED_4657_2\t0.125999\t8.6190e-11\t0.094905\t1.0638e-06\t0.094252\t1.2615e-06\t167816\t1255.5\t2\t2\t0\t1.6960e-04\n",
    "GN-Pou5f1_1\tINT_STITCHED_7784_3\t0.054350\t5.2688e-03\t0.059838\t2.1240e-03\t0.060118\t2.0235e-03\t158074\t10998.5\t1\t3\t2\t3.1086e-02\n",
    "GN-Ranbp17_1\tINT_STITCHED_5427_2\t0.106515\t4.2620e-08\t0.106515\t4.2620e-08\t0.106515\t4.2620e-08\t168552\t520.0\t1\t2\t1\t1.3857e-05\n",
    "GN-Ranbp17_2\tINT_STITCHED_5427_4\t0.071411\t2.4452e-04\t0.071411\t2.4452e-04\t0.071411\t2.4452e-04\t163439\t5633.0\t2\t4\t2\t7.3378e-03\n",
    "GN-Sall1_1\tINT_STITCHED_4179_2\t0.084541\t1.3954e-05\t0.072323\t2.0330e-04\t0.071302\t2.4992e-04\t163392\t5680.0\t1\t2\t1\t7.4390e-03\n",
    "GN-Sall1_1\tINT_STITCHED_4179_4\t0.081367\t2.9032e-05\t0.089027\t4.7388e-06\t0.089372\t4.3517e-06\t167306\t1766.0\t1\t4\t3\t4.1642e-04\n",
    "GN-Sall1_2\tINT_STITCHED_4179_2\t0.059641\t2.1971e-03\t0.069537\t3.5505e-04\t0.069930\t3.2853e-04\t162942\t6130.5\t2\t2\t0\t9.0598e-03\n",
    "GN-Sox2_1\tINT_STITCHED_1482_4\t0.052705\t6.8190e-03\t0.057161\t3.3393e-03\t0.057327\t3.2485e-03\t156215\t12857.0\t1\t4\t3\t4.2715e-02\n",
    "GN-Sox2_3\tINT_STITCHED_1482_5\t0.067775\t5.0008e-04\t0.070287\t3.0615e-04\t0.070319\t3.0420e-04\t163052\t6020.5\t3\t5\t2\t8.5392e-03\n",
    "GN-Sox2_4\tINT_STITCHED_1482_2\t0.061696\t1.5353e-03\t0.065631\t7.5068e-04\t0.065698\t7.4127e-04\t161102\t7969.5\t4\t2\t-2\t1.5725e-02\n",
    "GN-Sox2_5\tINT_STITCHED_1482_6\t0.072493\t1.9635e-04\t0.072493\t1.9635e-04\t0.072493\t1.9635e-04\t163810\t5261.5\t5\t6\t1\t6.3077e-03\n",
    "GN-Sox2_6\tINT_STITCHED_1482_6\t0.061123\t1.6984e-03\t0.063251\t1.1625e-03\t0.063279\t1.1566e-03\t159908\t9164.0\t6\t6\t0\t2.1331e-02\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "afad5d3f-f743-4007-bc7a-ad5bb2c1f054",
   "metadata": {},
   "outputs": [],
   "source": [
    "# CRISPR verified genes and cross SEs bins:\n",
    "corrRawCross = corrRaw %>%\n",
    "    # filter for GN- and thier SEs:\n",
    "    filter( substr(Gene, 0, 3) == \"GN-\" & substr(Enhancer, 0, 3) == \"INT\" ) %>%\n",
    "    filter( sub(\"_[^_]+$\", \"\", Gene) %in% SEassignment$assignedGene & sub(\"_[^_]+$\", \"\", Enhancer) %in% SEassignment$name ) %>%\n",
    "    # Ensure that the CGN and SEs are proper pairs:\n",
    "    filter( !paste0(sub(\"_[^_]+$\", \"\", Gene), sub(\"_[^_]+$\", \"\", Enhancer)) %in% paste0(SEassignment$assignedGene, SEassignment$name) ) %>%\n",
    "    filter( pAdj < 0.05 ) %>%\n",
    "    arrange( Gene );\n",
    "dim(corrRawCross)\n",
    "corrRawCross[1:4, ]\n",
    "corrRawCross$Enh_Gene\n",
    "\n",
    "SE_gene_bin_hist = corrRawCross %>%\n",
    "    ggplot(aes(x = Enh_Gene)) +\n",
    "    geom_histogram() +\n",
    "    geom_histogram(bins = 13, position = \"identity\", color = \"white\") +\n",
    "    geom_density() +\n",
    "    xlim(-6, 6) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    ggtitle(\"Binary (SE-Gene)\") +\n",
    "    xlab(\"Bin difference\") +\n",
    "    ylab(\"Number of pairs\") +\n",
    "    scale_fill_manual(values=c(\"#39568CFF\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 14, face=\"bold\"))\n",
    "SE_gene_bin_hist\n",
    "\n",
    "summary_bin_hist = data.frame(\n",
    "    Neg = sum(corrRawCross$Enh_Gene < 0)/length(corrRawCross$Enh_Gene),\n",
    "    Zero = sum(corrRawCross$Enh_Gene == 0)/length(corrRawCross$Enh_Gene),\n",
    "    Pos = sum(corrRawCross$Enh_Gene > 0)/length(corrRawCross$Enh_Gene)\n",
    "    ) %>%\n",
    "    pivot_longer(1:3, names_to = 'Category', values_to = 'Incidence') %>%\n",
    "    # use fct_inorder from forcats (part of tidyverse) to keep the original order of the data:\n",
    "    ggplot(aes(x = fct_inorder(Category), y = Incidence)) +\n",
    "    # geom_bar(stat=\"identity\", fill=c(\"#9e9a75\",\"gray\", \"#41533b\")) +\n",
    "    geom_col(fill=c(\"#9e9a75\",\"gray\", \"#41533b\")) +\n",
    "    theme(legend.position=\"none\") +\n",
    "    ggtitle(\"Binary (SE-Gene)\") +\n",
    "    xlab(\"Class\") +\n",
    "    ylab(\"Fraction of pairs\") +\n",
    "    # scale_fill_manual(values=c(\"#9e9a75\",\"gray\", \"#41533b\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 14, face=\"bold\"))\n",
    "summary_bin_hist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da083066-3602-421c-add2-5cdccee102fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "t.test(corrRawF$Enh_Gene, corrRawCross$Enh_Gene, alternative = \"two.sided\")\n",
    "# wilcox.test(corrRawF$Enh_Gene, corrRawCross$Enh_Gene)\n",
    "# ks.test(corrRawF$Enh_Gene, corrRawCross$Enh_Gene, alternative = \"less\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97252b6c-f15b-4d59-926a-9a8d247af5f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# overlapping histogram:\n",
    "corrRawF = corrRaw %>%\n",
    "        # filter for GN- and thier SEs:\n",
    "    filter( substr(Gene, 0, 3) == \"GN-\" & substr(Enhancer, 0, 3) == \"INT\" ) %>%\n",
    "    filter( sub(\"_[^_]+$\", \"\", Gene) %in% SEassignment$assignedGene & sub(\"_[^_]+$\", \"\", Enhancer) %in% SEassignment$name ) %>%\n",
    "    filter( pAdj < 0.05 );\n",
    "dim(corrRawF)\n",
    "corrRawF[1:4, ]\n",
    "\n",
    "# make a new column class to plot all vs significant as overlapping histogram:\n",
    "corrRawF$class = ifelse(paste0(sub(\"_[^_]+$\", \"\", corrRawF$Gene), sub(\"_[^_]+$\", \"\", corrRawF$Enhancer)) %in% paste0(SEassignment$assignedGene, SEassignment$name), \"CRISPR verified pairs\", \"aScrambled random pairs\")\n",
    "summary(as.factor(corrRawF$class))\n",
    "corrRawF[1:4, ]\n",
    "\n",
    "corrRawF %>%\n",
    "    ggplot( aes(x=Enh_Gene, fill=class)) +\n",
    "    geom_histogram(binwidth = 1, position = \"identity\", \n",
    "                   alpha = 0.6, \n",
    "                   # color = \"white\",\n",
    "                   # mapping = aes(y = after_stat(count))) +\n",
    "                   # mapping = aes(y = after_stat(ncount))) +\n",
    "                   # mapping = aes(y = after_stat(count/sum(count)))) +\n",
    "                   mapping = aes(y = after_stat(density))) +\n",
    "    scale_fill_manual(values=c(\"#7c6a77\", \"#011a0c\")) +\n",
    "    xlim(-4, 4) +\n",
    "    theme(legend.position = c(0.8, 0.875), \n",
    "          legend.text = element_text(size=10, face=\"bold\"),\n",
    "          legend.key.size = unit(0.5, 'cm')) +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "    # scale_x_log10() +\n",
    "    # ggtitle(\"Co-expressed genes\") +\n",
    "    xlab(\"Difference in correlated bins (Enhancer - Gene)\") +\n",
    "    ylab(\"Density\")\n",
    "ggsave(filename=\"../plots/ExG_Difference_in_correlated_bins_E-G.pdf\", width=4, height=4, units=\"in\")\n",
    "\n",
    "corrRawF %>%\n",
    "    ggplot( aes(x=Enh_Gene, fill=class)) +\n",
    "    geom_histogram(binwidth = 1, position = \"identity\", alpha = 0.6) +\n",
    "    xlim(-4, 4) +\n",
    "    theme(legend.position = \"none\") +\n",
    "    guides(fill=guide_legend(title=\"\")) +\n",
    "\n",
    "    facet_wrap(~class, scales = \"free\") +\n",
    "    scale_fill_manual(values=c(\"#7c6a77\", \"#011a0c\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 14, face=\"bold\")) +\n",
    "\n",
    "    # ggtitle(\"Co-expressed genes\") +\n",
    "    xlab(\"Difference in correlated bin position \\n from TSS (Enhancer - Gene)\") +\n",
    "    ylab(\"Number of pairs\")\n",
    "ggsave(filename=\"../plots/ExG_Difference_in_correlated_bins_E-G_facets.pdf\", width=4, height=4, units=\"in\")\n",
    "\n",
    "corrRawF %>%\n",
    "    ggplot( aes(x=Enh_Gene, fill=class)) +\n",
    "    geom_histogram(binwidth = 1, position = \"identity\", alpha = 0.6, mapping = aes(y = after_stat(density))) +\n",
    "    xlim(-4, 4) +\n",
    "    theme(legend.position = \"none\") +\n",
    "    # theme(legend.position = c(0.8, 0.875), \n",
    "    #       legend.text = element_text(size=10, face=\"bold\"),\n",
    "    #       legend.key.size = unit(0.5, 'cm')) +\n",
    "    # guides(fill=guide_legend(title=\"\")) +\n",
    "\n",
    "    facet_wrap(~class) + # , scales = \"free\"\n",
    "    scale_fill_manual(values=c(\"#7c6a77\", \"#011a0c\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 14, face=\"bold\")) +\n",
    "\n",
    "    # ggtitle(\"Co-expressed genes\") +\n",
    "    xlab(\"Difference in correlated bin position from TSS \\n (Enhancer - Gene)\") +\n",
    "    ylab(\"Density\")\n",
    "ggsave(filename=\"../plots/ExG_Difference_in_correlated_bins_E-G_facets_density.pdf\", width=6, height=4, units=\"in\")\n",
    "\n",
    "t.test(Enh_Gene ~ class, corrRawF);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4dd52473-4851-4f5a-8e94-3e7b04a1fc01",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c63f4b85-52b9-43d5-9bf1-f4afdab5bc46",
   "metadata": {},
   "outputs": [],
   "source": [
    "# SEneighbors genes and SEs bins:\n",
    "corrSEneighbors = corrRaw %>%\n",
    "    filter( substr(Gene, 0, 3) == \"GN-\" & substr(Enhancer, 0, 3) == \"INT\" ) %>%\n",
    "    filter( sub(\"_[^_]+$\", \"\", Gene) %in% names(SEneighbors[SEneighbors$TADstatus == \"IN\", ])) %>%\n",
    "    filter( as.character(seqnames(allfeatures[Gene])) == as.character(seqnames(allfeatures[Enhancer])) );\n",
    "nrow(corrSEneighbors)\n",
    "\n",
    "corrSEneighborsSig = corrSEneighbors %>%\n",
    "    filter( pAdj < 0.05 );\n",
    "nrow(corrSEneighborsSig)\n",
    "\n",
    "SE_gene_bin_hist = corrSEneighborsSig %>%\n",
    "    ggplot(aes(x = Enh_Gene)) +\n",
    "    geom_histogram() +\n",
    "    geom_histogram(binwidth = 1, position = \"identity\", color = \"white\") +\n",
    "    geom_density() +\n",
    "    theme(legend.position=\"none\") +\n",
    "    ggtitle(\"Binary (SE-Gene)\") +\n",
    "    xlab(\"Bin difference\") +\n",
    "    ylab(\"Number of pairs\") +\n",
    "    scale_fill_manual(values=c(\"#39568CFF\")) +\n",
    "    theme(strip.background = element_blank(),\n",
    "        strip.text = element_text(size = 14, face=\"bold\"))\n",
    "SE_gene_bin_hist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a6fe5b76-ae8b-4c32-bdde-6010cce432fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "corrSEout = corrRaw %>%\n",
    "    filter( substr(Gene, 0, 3) == \"GN-\" & substr(Enhancer, 0, 3) == \"INT\" ) %>%\n",
    "    filter( sub(\"_[^_]+$\", \"\", Gene) %in% names(SEneighbors[SEneighbors$TADstatus == \"OUT\", ])) %>%\n",
    "    filter( as.character(seqnames(allfeatures[Gene])) == as.character(seqnames(allfeatures[Enhancer])) );\n",
    "nrow(corrSEout)\n",
    "\n",
    "corrSEoutSig = corrSEout %>%\n",
    "    filter( pAdj < 0.05 );\n",
    "nrow(corrSEoutSig)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ba18d4d1-1591-446f-9996-dc57c85f1dd5",
   "metadata": {},
   "source": [
    "SE neighbors - IN (within 1 Mb) 108 genes\n",
    "122/1118 = 10.91\n",
    "\n",
    "SE neighbors - OUT: (4 -5 Mb) 100 genes\n",
    "70/742 = 9.43"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "991e83e6-d8d0-42a3-bc00-2c74faf04abc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Histone genes and SEs bins:\n",
    "corrHistones = corrRaw %>%\n",
    "    filter( substr(Gene, 0, 7) == \"GN-Hist\" & substr(Enhancer, 0, 3) == \"INT\" );\n",
    "nrow(corrHistones)\n",
    "\n",
    "corrHistonesSig = corrHistones %>%\n",
    "    filter( pAdj < 0.05 );\n",
    "nrow(corrHistonesSig)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "2480befc-9967-48f8-b221-9c5a0f50a945",
   "metadata": {},
   "source": [
    "Hist genes enrichment:\n",
    "133/2475 = 5.37"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "515a1ee8-72c6-4149-b1af-e91e7d2816fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# G1 genes and SEs bins:\n",
    "corrG1 = corrRaw %>%\n",
    "    filter( sub(\"_[^_]+$\", \"\", Gene) %in% G1_Beyrouthy & substr(Enhancer, 0, 3) == \"INT\" );\n",
    "nrow(corrG1)\n",
    "\n",
    "corrG1Sig = corrG1 %>%\n",
    "    filter( pAdj < 0.05 );\n",
    "nrow(corrG1Sig)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "0d95299c-fe34-43c3-84b5-1e612dfbf400",
   "metadata": {},
   "source": [
    "G1 genes enrichment:\n",
    "23/270 = 8.51"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "85049256-1f18-41ba-8b9b-5babc9dba2dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# G1S and SEs bins:\n",
    "corrG1S = corrRaw %>%\n",
    "    filter( sub(\"_[^_]+$\", \"\", Gene) %in% G1S_Riba & substr(Enhancer, 0, 3) == \"INT\" );\n",
    "nrow(corrG1S)\n",
    "\n",
    "corrG1SSig = corrG1S %>%\n",
    "    filter( pAdj < 0.05 );\n",
    "nrow(corrG1SSig)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87061974-a05a-4bc8-88d2-06441338300d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# G1S and SEs bins:\n",
    "corrG2M = corrRaw %>%\n",
    "    filter( sub(\"_[^_]+$\", \"\", Gene) %in% G2M_Riba & substr(Enhancer, 0, 3) == \"INT\" );\n",
    "nrow(corrG2M)\n",
    "\n",
    "corrG2MSig = corrG2M %>%\n",
    "    filter( pAdj < 0.05 );\n",
    "nrow(corrG2MSig)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "42d83cd7-18a4-44ac-b033-2a10cc9aad99",
   "metadata": {},
   "source": [
    "G1S genes enrichment:\n",
    "94/675 = 13.92\n",
    "\n",
    "G2M genes enrichment:\n",
    "88/1035 = 8.50"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87d9494e-a097-4f1f-9aa1-1ac4c171a0ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# barplot showing the expression of SEs in various cell cycles and TADs:\n",
    "SEcellcycle = data.frame(\n",
    "    class = c(\"G1S\", \n",
    "              \"S\", \n",
    "              \"G2M\"),\n",
    "    percentage = c(nrow(corrG1SSig)/nrow(corrG1S)*100,\n",
    "                   nrow(corrHistonesSig)/nrow(corrHistones)*100,\n",
    "                   nrow(corrG2MSig)/nrow(corrG2M)*100\n",
    "                  ));\n",
    "SEcellcycle\n",
    "ggplot(SEcellcycle, aes(x=fct_inorder(class), y=percentage)) +\n",
    "  geom_bar(stat=\"identity\", fill=c(\"#ebc14c\", \"#265369\", \"#545756\")) + #333333\n",
    "  labs(x=\"Cell Cycle Phase\", \n",
    "       # title=\"Percentage of Cells in Each Phase\",\n",
    "       y=\"Percentage\", )+\n",
    "    ylim(0, 15)\n",
    "    # scale_fill_manual(values=c(\"#7c6a77\", \"#011a0c\")) +\n",
    "    # geom_density() +\n",
    "    # xlim(0, 2500000) +\n",
    "    # theme(legend.position = c(0.8, 0.9), \n",
    "          # legend.text = element_text(size=10, face=\"bold\"),\n",
    "          # legend.key.size = unit(0.7, 'cm')) +\n",
    "    # guides(fill=guide_legend(title=\"\")) +\n",
    "    # scale_x_log10() +\n",
    "    # ggtitle(\"Co-expressed genes\") +\n",
    "    # xlab(\"\") +\n",
    "ggsave(filename=\"../plots/ExG_crisprVerifiedSEs_expression_in_cellCycle.pdf\", width=4, height=4, units=\"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7776f427-4835-454d-88f1-dba9e2172540",
   "metadata": {},
   "outputs": [],
   "source": [
    "# barplot showing the expression of SEs in various cell cycles and TADs:\n",
    "SEcellcycle = data.frame(\n",
    "    class = c(\"G1S\", \n",
    "              \"S\", \n",
    "              \"G2M\", \n",
    "              \"Inside 1 Mb\", \n",
    "              \"Between 4-5 Mb\"),\n",
    "    percentage = c(nrow(corrG1SSig)/nrow(corrG1S)*100,\n",
    "                   nrow(corrHistonesSig)/nrow(corrHistones)*100,\n",
    "                   nrow(corrG2MSig)/nrow(corrG2M)*100,\n",
    "                   nrow(corrSEneighborsSig)/nrow(corrSEneighbors)*100,\n",
    "                   nrow(corrSEoutSig)/nrow(corrSEout)*100\n",
    "                  ));\n",
    "SEcellcycle\n",
    "ggplot(SEcellcycle, aes(x=fct_inorder(class), y=percentage)) +\n",
    "  geom_bar(stat=\"identity\", fill=c(\"#ebc14c\", \"#265369\", \"#545756\", rep(\"#818664\", 2))) + #333333\n",
    "  labs(x=\"Cell Cycle Phase\", \n",
    "       # title=\"Percentage of Cells in Each Phase\",\n",
    "       y=\"Percentage\", )+\n",
    "    ylim(0, 15)\n",
    "    # scale_fill_manual(values=c(\"#7c6a77\", \"#011a0c\")) +\n",
    "    # geom_density() +\n",
    "    # xlim(0, 2500000) +\n",
    "    # theme(legend.position = c(0.8, 0.9), \n",
    "          # legend.text = element_text(size=10, face=\"bold\"),\n",
    "          # legend.key.size = unit(0.7, 'cm')) +\n",
    "    # guides(fill=guide_legend(title=\"\")) +\n",
    "    # scale_x_log10() +\n",
    "    # ggtitle(\"Co-expressed genes\") +\n",
    "    # xlab(\"\") +\n",
    "ggsave(filename=\"../plots/ExG_crisprVerifiedSEs_expression_in_cellCycle_plus_TADs.pdf\", width=4, height=4, units=\"in\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8906db4-9c00-4677-a303-78d8ed45e405",
   "metadata": {},
   "outputs": [],
   "source": [
    "FAIL"
   ]
  },
  {
   "cell_type": "raw",
   "id": "18195ff4-211f-4bdd-a231-ff11137b527a",
   "metadata": {},
   "source": [
    "#### Read Chi-square ExG correlation"
   ]
  },
  {
   "cell_type": "raw",
   "id": "bb892c94-e8a4-486f-95b1-cda401b0b08c",
   "metadata": {},
   "source": [
    "corrChiSquare = fread(file=\"../data/scGROv2p8_Gene10kb_Enh_genes_10Kpermuted_chisq_empp.csv.gz\");\n",
    "corrChiSquare[1:4,]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "003af356-befd-40d5-9957-55f1c9cf6e09",
   "metadata": {},
   "source": [
    "corrChiSquare = fread(  file=\"../data/scGROv2p8_Gene10kb_Enh_genes_10Kpermuted_chisq_empp.csv.gz\" ) %>%\n",
    "                mutate( Gene.chr = as.character( seqnames(features[paste0(\"GN-\",Gene)]) ) ) %>%\n",
    "                mutate( Enh.chr  = as.character( seqnames(features[Enhancer]) ) ) %>%\n",
    "                mutate( Gene.tss = start(features[paste0(\"GN-\",Gene)]) ) %>%\n",
    "                mutate( Enh.mid  = start(features[Enhancer])+width(features[Enhancer])/2 ) %>%\n",
    "                mutate( distance = ifelse(Gene.chr == Enh.chr, abs(Enh.mid-Gene.tss), NA) );\n",
    "\n",
    "# diff_chr = corrChiSquare %>%\n",
    "#     filter(!Gene %in% blacklist) %>%\n",
    "#     filter(as.character(seqnames(features[paste0(\"GN-\",Gene)])) != as.character(seqnames(features[Enhancer])));\n",
    "\n",
    "# # remove blacklist genes and\n",
    "# # keep gene-enhancer pairs on same chromosome\n",
    "# same_chr = chr_corr %>%\n",
    "#     filter(!Gene %in% blacklist) %>%\n",
    "#     filter(as.character(seqnames(features[paste0(\"GN-\",Gene)])) == as.character(seqnames(features[Enhancer])));"
   ]
  },
  {
   "cell_type": "raw",
   "id": "31b493aa-1bf4-4839-a3d0-629d2c788fb7",
   "metadata": {},
   "source": [
    "### Merge the 4 ExG tests: three type of correlations and a chi-square"
   ]
  },
  {
   "cell_type": "raw",
   "id": "aa374a91-f528-436a-94d3-b030244456c3",
   "metadata": {},
   "source": [
    "# join two tables using two columns\n",
    "corr = left_join(   corrPflat[ ,c(1:3,5)], corrSflat[ ,c(1:3,5)], by=c(\"Gene\", \"Enhancer\"), suffix = c(\".P\", \".S\") ) %>%\n",
    "       left_join(., corrBflat[ ,c(1:3,5)], by=c(\"Gene\", \"Enhancer\") ) #%>%\n",
    "       # left_join(., corrChiSquare, by=c(\"Gene\", \"Enhancer\") ) #%>%\n",
    "    # na.omit();\n",
    "\n",
    "dim(corr)\n",
    "corr[1:4, ]\n",
    "fwrite(corr, file=\"../data/scGROv2p8_SE_CRISPRverified_5Kbins_ALL_correlation.csv.gz\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
