{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab9df20e-4f8e-4839-8090-89d9c72b377a",
   "metadata": {},
   "outputs": [],
   "source": [
    ".libPaths(\"/home/mahat/.conda/envs/r422/lib/R/library\")\n",
    ".libPaths()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b159ff73-dadd-4aa4-beec-a733593eaa4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages({\n",
    "    library(tidyverse)\n",
    "    library(ggplot2)\n",
    "    library(dplyr)\n",
    "    library(matrixStats)\n",
    "    library(foreach)\n",
    "    library(doParallel)\n",
    "    library(plyranges)\n",
    "    library(viridis)\n",
    "    library(ggpointdensity)\n",
    "    library(Matrix)\n",
    "    library(data.table)\n",
    "    library(TxDb.Mmusculus.UCSC.mm10.knownGene)\n",
    "    library(org.Mm.eg.db)\n",
    "    library(rtracklayer)\n",
    "});"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "104b0744-9e76-43bb-b44f-a1210e78db0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "registerDoParallel(16);\n",
    "options(\n",
    "    repr.plot.width=4,\n",
    "    repr.plot.height=3,\n",
    "    jupyter.plot_mimetypes = \"image/svg+xml\"\n",
    ");\n",
    "theme_set(theme_classic() +\n",
    "    theme(\n",
    "        axis.title.x = element_text(color=\"black\", size=14, face=\"bold\"), \n",
    "        axis.title.y = element_text(color=\"black\", size=14, face=\"bold\"),\n",
    "        axis.text = element_text(color=\"black\", size=12, face=\"bold\"),\n",
    "        plot.title = element_text(face=\"bold\", size=14, hjust = 0.5),\n",
    "        axis.line = element_blank(),\n",
    "        # axis.ticks = element_blank()\n",
    "        panel.border = element_rect(colour = \"grey\", fill=NA, linewidth=1)\n",
    "    )\n",
    ");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ab14f7ba-9850-4412-8ac2-6ba9fa5b8f0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify maximum gene length to be used for G-E correlation:\n",
    "geneLength = 10000;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "589d9607-c20a-4356-81f2-76e6c3fda3a8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load genes with dREG status for expression filter\n",
    "genesWithdREGstatus = readRDS(\"../data/groHMM_dREG_refinedGenes_mES_mm10.rds\");\n",
    "genesWithdREGstatus\n",
    "# load groHMM-extended genes and enhancers\n",
    "# features = read_bed(\"../data/groHMM_mES_BRsComb_LP-50_UTS10_features_customized_v2.bed\");\n",
    "\n",
    "# load features\n",
    "features = read_bed(\"../data/groHMM_dREG_refinedFeatures_mES_mm10_OSNcustomEnhancers_SEs.bed\");\n",
    "# features = read_bed(\"../data/mES_BRsComb_dREGfiltered_features_customized_OSNenhancersPlusSEs_v1.bed\");\n",
    "# features = read_bed(\"../data/groHMM_mES_BRsComb_LP-50_UTS10_features_customized_OSNenhancersPlusSEs_v2.bed\");\n",
    "names(features) = features$name;\n",
    "# features$name=NULL;\n",
    "features$score=NULL;\n",
    "features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "116f613e-6fa9-4fa4-b56d-b00c8d17db5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes shorter than 1500 are: 2931\n",
    "# genes shorter than 1000 are: 2209\n",
    "# genes shorter than  500 are:  484\n",
    "# genes shorter than  200 are:  147\n",
    "\n",
    "# short Gm### genes\n",
    "shortGm = features %>%\n",
    "    filter( substr(names, 0, 3) == \"GN-\" ) %>%\n",
    "    filter( width > 200 ) %>%\n",
    "    filter( width < 1000 & substr(names, 0, 5) == \"GN-Gm\" );\n",
    "length(shortGm)\n",
    "summary(width(shortGm))\n",
    "names(shortGm)[1:5]\n",
    "\n",
    "# define highly expressed \"blacklist\" genes to be\n",
    "# removed from correlations\n",
    "blacklist = grep( \"GN-Rp[ls]\" , names(features) );\n",
    "blacklist = grep( \"GN-Sn[oh]\", names(features) ) %>%\n",
    "    append(blacklist);\n",
    "\n",
    "blacklist = names(features)[ blacklist ];\n",
    "length(blacklist)\n",
    "summary(width(features[ blacklist ]))\n",
    "blacklist[c(1:5, 110:115)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f31c9b5b-f3de-437a-9f6c-b0c8549708c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter genes\n",
    "genes = features %>%\n",
    "    filter( substr(names, 0, 3) == \"GN-\" ) %>%\n",
    "    # remove genes less than 1500 nt\n",
    "    filter( width >= 1500 ) %>%\n",
    "    # filter genes with dREG peaks\n",
    "    filter( names %in% names(genesWithdREGstatus[genesWithdREGstatus$dREG == T]) | \n",
    "           !names %in% names(genesWithdREGstatus) ) %>%\n",
    "    # filter highly expressed genes\n",
    "    filter(!names %in% blacklist ) %>%\n",
    "    # filter Gm genes shorter than 1 kb\n",
    "    filter(!names %in% names(shortGm) ) %>%\n",
    "    # remove 500 nt at the 5' end\n",
    "    anchor_3p() %>%\n",
    "    mutate( width = width - 500 );\n",
    "summary(width(genes))\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b29626a4-b3dc-4d13-abcf-a88c0586e511",
   "metadata": {},
   "outputs": [],
   "source": [
    "enhancers = features %>%\n",
    "    filter( substr(names, 0, 3) != \"GN-\" ) %>%\n",
    "    filter( width >= 3000 );\n",
    "summary(width(enhancers))\n",
    "table(substr(enhancers$name, 0, 3));\n",
    "enhancers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "578f5038-21e2-43ad-9412-b497c23b7961",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load groHMM-extended genes and enhancers\n",
    "dREG = read.table(\"../data/PROseq_mES_BRsComb.dREG.peak.full.bed\", header=F, stringsAsFactors=F);\n",
    "colnames(dREG) = c(\"chr\", \"start\", \"end\", \"score\", \"pval\", \"center\");\n",
    "# use the \"center\" column as start\n",
    "# colnames(dREG) = c(\"chr\", \"leftEnd\", \"rightEnd\", \"score\", \"pval\", \"start\");\n",
    "# dREG$end = dREG$start;\n",
    "dREG = GRanges(dREG);\n",
    "summary(dREG$score)\n",
    " # dREG = dREG %>%\n",
    " #    # removing dREG peaks with score lower than mean\n",
    " #    filter( score > median(score) );\n",
    "\n",
    "# mark dREG peaks by the overlapping enhancers (from freatures):\n",
    "hits = findOverlaps( dREG, enhancers, ignore.strand=T );\n",
    "dREG$assignedEnhancer = NA;\n",
    "# assign ovelapping SE to dREG peaks\n",
    "dREG$assignedEnhancer[hits@from] = names(enhancers)[hits@to];\n",
    "\n",
    "# filter dREG peak to only leave one with highest score:\n",
    "dREGfilt = dREG %>%\n",
    "    # filter dREG peaks that do not overlap with an enhancer\n",
    "    na.omit() %>%\n",
    "    group_by( assignedEnhancer ) %>%\n",
    "    filter( score == max(score) ) %>%\n",
    "    ungroup() %>%\n",
    "    # make data table for left_join\n",
    "    as.data.table;\n",
    "dim(dREGfilt)\n",
    "table(substr(dREGfilt$assignedEnhancer, 0, 3));\n",
    "length(unique(dREGfilt$assignedEnhancer))\n",
    "dREGfilt[1:4, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be0e2d74-a2d2-48b8-bf4b-6805d3c755f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "Ecenter = left_join( as.data.table(enhancers), dREGfilt[,c(6:9)], by=c(\"name\"=\"assignedEnhancer\") ) %>%\n",
    "    na.omit() #%>%\n",
    "    # GRanges()\n",
    "# names(Ecenter) = Ecenter$name;\n",
    "Ecenter$pval = NULL;\n",
    "Ecenter$score = NULL;\n",
    "Ecenter[1:4, ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc85f54d-fb22-4c77-8718-9e0aa4af648f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create Enhancer bin in sense direction\n",
    "Edown = Ecenter %>%\n",
    "    mutate( strand = \"+\" ) %>%\n",
    "    #convert center to start\n",
    "    mutate( start = center ) %>%\n",
    "    # make sure the end is at least 1750 For > 1750, keep as it is\n",
    "    mutate( end = ifelse(end - center > 1750, end, center + 1750) ) %>%\n",
    "    GRanges() %>%\n",
    "    anchor_3p() %>%\n",
    "    mutate( width = width - 250 )\n",
    "    # # no need for this beacuse the genes will be filtered against enhancers:\n",
    "    # subsetByOverlaps( genes, invert = TRUE );\n",
    "names(Edown) = Edown$name;\n",
    "Edown$center = NULL;\n",
    "Edown\n",
    "# create Enhancer bin in anti-sense direction\n",
    "Eup = Ecenter %>%\n",
    "    mutate( strand = \"-\" ) %>%\n",
    "    mutate( end = center) %>%\n",
    "    mutate( start = ifelse(center - start > 1750, start, center - 1750) ) %>%\n",
    "    GRanges() %>%\n",
    "    anchor_3p() %>%\n",
    "    mutate( width = width - 250 );\n",
    "names(Eup) = Eup$name;\n",
    "Eup$center = NULL;\n",
    "Eup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b04ccc6-8b8a-435f-ad34-432a143d0c49",
   "metadata": {},
   "outputs": [],
   "source": [
    "# combine features:\n",
    "expressedFeatures = c(Eup, Edown, genes) %>%\n",
    "     sort();\n",
    "summary(width(expressedFeatures))\n",
    "expressedFeatures\n",
    "\n",
    "# trim the long genes to 10 kb\n",
    "longf = which( width(expressedFeatures) >= 10000 );\n",
    "expressedFeatures[longf] = expressedFeatures[longf] %>%\n",
    "    resize( width = 10000, fix=\"start\" );\n",
    "summary(width(expressedFeatures))\n",
    "expressedFeatures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "297721c9-d639-47b6-9d41-8c151e1d942e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter out features that overlap:\n",
    "hits = findOverlaps(expressedFeatures)\n",
    "expressedFeatures$overlappingFeature = NA;\n",
    "# assign ovelapping feature\n",
    "expressedFeatures$overlappingFeature[hits@from] = names(expressedFeatures)[hits@to];\n",
    "\n",
    "test = expressedFeatures %>%\n",
    "    na.omit() %>%\n",
    "    group_by( overlappingFeature ) %>%\n",
    "    filter( n_distinct(name) == 1 ) %>%\n",
    "    ungroup();\n",
    "test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0a65a46d-f720-48a9-8291-70760caaade0",
   "metadata": {},
   "outputs": [],
   "source": [
    "scGRO = readRDS(\"../data/scGROv2p8_consolidated.rds\");\n",
    "counts = readRDS(\"../data/scGROv2p8_mapq3qc_max10kbp_filtered_counts.rds\");\n",
    "#counts = counts[,colSums(counts) >= 1000 ];\n",
    "dim(counts)\n",
    "allcells = colnames(counts);\n",
    "table(substr(allcells, 0, 6));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3e472e8-686c-4ada-a741-b2f125312aac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# quick check for the expression of OSN genes:\n",
    "test = counts[rownames(counts) %in% c(\"GN-Pou5f1\",\"GN-Sox2\",\"GN-Nanog\",\"GN-Sox2long\"),]\n",
    "t(rowMeans(test))\n",
    "\n",
    "Pou5f1 = counts[rownames(counts) == c(\"GN-Pou5f1\"),]\n",
    "table(Pou5f1)\n",
    "Nanog = counts[rownames(counts) == c(\"GN-Nanog\"),]\n",
    "table(Nanog)\n",
    "Sox2 = counts[rownames(counts) == c(\"GN-Sox2\"),]\n",
    "table(Sox2)\n",
    "Sox2long = counts[rownames(counts) == c(\"GN-Sox2long\"),]\n",
    "table(Sox2long)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9875c9be-bdc0-4699-858c-0a9daa64c3e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge experiment ID and cell barcode to create unique cell ID across experiments\n",
    "scGRO = scGRO %>%\n",
    "    filter( mapq >= 3 & countQC & umiQC & plateQC & cellQC & miRQC ) %>%\n",
    "    subsetByOverlaps( expressedFeatures ) %>%\n",
    "    mutate( cellID = factor( paste( Exp, Plate, Cell, sep=\"-\") ) ) %>%\n",
    "    filter( cellID %in% allcells ) #%>%\n",
    "    # select( cellID );\n",
    "scGRO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a51c3e25-8cff-445f-bdae-89c81e45e5b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts = foreach(\n",
    "    id = allcells,\n",
    "    .combine=\"cbind2\"\n",
    ") %dopar% {\n",
    "    reads = scGRO %>% filter(cellID == id);\n",
    "    # count reads in each feature for this cellID\n",
    "    counts1 = countOverlaps( expressedFeatures, reads );\n",
    "\n",
    "    counts = cbind(counts1) %>%\n",
    "        Matrix(sparse=T);\n",
    "\n",
    "    return( counts );\n",
    "}\n",
    "# columns = cells = cellIDs\n",
    "colnames(counts) = allcells;\n",
    "rownames(counts) = names(expressedFeatures);\n",
    "dim(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b88fb4fa-f597-4253-9b7e-a8a820d43cde",
   "metadata": {},
   "outputs": [],
   "source": [
    "# to sum rows with same rownames\n",
    "# the sense and anti-sense bins of enhancers have same names\n",
    "counts = rowsum(counts, row.names(counts))\n",
    "dim(counts)\n",
    "# rowSums(counts)\n",
    "# colSums(counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5f33c8c-aa45-4ba3-bc96-9b7622d595c9",
   "metadata": {},
   "outputs": [],
   "source": [
    "# only retain features with counts in 0.1% or more cells\n",
    "observed = counts[ rowMeans(counts>0) >= 0.001, ];\n",
    "dim(observed)\n",
    "\n",
    "test = substr(rownames(observed), 0, 3)\n",
    "summary(as.factor(test))\n",
    "\n",
    "genes = genes %>%\n",
    "    filter(names %in% rownames(observed));\n",
    "length(genes);\n",
    "enhancers = enhancers %>%\n",
    "    filter(names %in% rownames(observed));\n",
    "length(enhancers);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9e2099a-f20d-41c1-92a5-470fe81086d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "summary(colSums(observed))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "345d8032-bb66-40f5-b039-3789aad195a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert read counts into probability -- each cell column sums to 1\n",
    "pmatrix = t(t(observed) / colSums(observed));\n",
    "# average across cells\n",
    "pvector = rowMeans(pmatrix);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "95316360-0425-4463-8d26-461101ea17bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define important dimensions of our data\n",
    "Niters = 1000;\n",
    "Ncells = ncol(observed);\n",
    "Ngenes = length(genes);\n",
    "Nenh   = length(enhancers);\n",
    "Nreads = colSums(observed);\n",
    "samplesize = sum(Nreads);\n",
    "\n",
    "# repeat each cellID by its read count\n",
    "cell_index = rep(1:Ncells, times=Nreads);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e21f2cfd-e46f-43f2-bf1e-193f2dbfc15f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# clear unsed memory\n",
    "gc()\n",
    "# memory.size() ### Checking your memory size\n",
    "# memory.limit() ## Checking the set limit\n",
    "# memory.limit(size=56000) ### expanding your memory _ here it goes beyond to your actually memory. This 56000 is proposed for 64Bit. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d232958-d8fd-4e57-9785-7e129434e168",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Binarize observed counts\n",
    "obs_genes = observed[names(genes),];\n",
    "obs_genes[ obs_genes>1 ] = 1;\n",
    "obs_enh = observed[names(enhancers),];\n",
    "obs_enh[ obs_enh>1 ] = 1;\n",
    "\n",
    "# Multiply each Ngenes x Ncells matrix by its\n",
    "# transpose within each iteration.\n",
    "# Because data is binary, this counts co-occurence\n",
    "# of 1's among all genes pairs.\n",
    "#obsx = obs_genes %*% t(obs_enh) / Ncells;\n",
    "\n",
    "# implement chisq test\n",
    "# compute expected coexpression\n",
    "GxE = rowMeans(obs_genes) %*% t(rowMeans(obs_enh));\n",
    "obsGE = as.matrix(obs_genes %*% t(obs_enh)) / Ncells;\n",
    "obs_chisq = (obsGE-GxE)^2/GxE;\n",
    "obs_chisq = obs_chisq * ifelse(obsGE>GxE, 1, -1);\n",
    "dim(obsGE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e35179a8-dd99-4faf-89d3-1182ff8395d9",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# initialize matrix to compute empirical p-values\n",
    "emp_p_pos = matrix(0, nrow=Ngenes, ncol=Nenh);\n",
    "emp_p_neg = matrix(0, nrow=Ngenes, ncol=Nenh);\n",
    "# emp_p     = matrix(0, nrow=Ngenes, ncol=Nenh);\n",
    "\n",
    "for( n in 1:Niters ) {\n",
    "    # report progress every 100 iterations\n",
    "    if( n %% 100 == 0 ) {\n",
    "        message(n);\n",
    "    }\n",
    "    \n",
    "    # randomly sample from genes with replacement\n",
    "    simx = sample( Ngenes+Nenh, size=samplesize, replace=T, prob=pvector );\n",
    "    # assign sampled genes to cells based on read count\n",
    "    simx = cbind(simx, cell_index);\n",
    "    # binarize\n",
    "    simx = unique(simx);\n",
    "    simx = sparseMatrix(i=simx[,1], j=simx[,2], x=1, dims=c(Ngenes+Nenh, Ncells));\n",
    "    rownames(simx) = rownames(observed);\n",
    "    colnames(simx) = colnames(observed);\n",
    "    \n",
    "    # compute coexpression\n",
    "    #simx = simx[names(genes),] %*% t(simx[names(enhancers),]) / Ncells;\n",
    "    # compare to observed coexpression\n",
    "    #emp_p = emp_p + (simx >= obsx);\n",
    "    \n",
    "    # compute chisq and p-values\n",
    "    suppressWarnings({\n",
    "        GxE = rowMeans(simx[names(genes),]) %*% t(rowMeans(simx[names(enhancers),]));\n",
    "        simGE = as.matrix(simx[names(genes),] %*% t(simx[names(enhancers),])) / Ncells;\n",
    "        sim_chisq = (simGE-GxE)^2/GxE;\n",
    "        sim_chisq = sim_chisq * ifelse(simGE>GxE, 1, -1);\n",
    "        emp_p_pos = emp_p_pos + !(obs_chisq > sim_chisq);\n",
    "        emp_p_neg = emp_p_neg + !(obs_chisq < sim_chisq);\n",
    "        # emp_p_pos = emp_p_pos + (!(GxE > simGE & obs_chisq > sim_chisq));\n",
    "        # emp_p_neg = emp_p_neg + (!(GxE < simGE & obs_chisq > sim_chisq));\n",
    "        # also add the old method of p-value by just observing the incidence of obsx smaller than simx\n",
    "        # emp_p = emp_p + (simGE >= GxE);\n",
    "    });\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ff1d8e2-3aed-4099-8711-6404702fe752",
   "metadata": {},
   "outputs": [],
   "source": [
    "GxE = rowMeans(obs_genes) %*% t(rowMeans(obs_enh));\n",
    "chr_corr = data.frame(\n",
    "    Gene       = rep(1:Ngenes, Nenh),\n",
    "    Enhancer   = rep(1:Nenh  , each=Ngenes),\n",
    "    pGene      = rep(rowMeans(obs_genes), Nenh),\n",
    "    pEnh       = rep(rowMeans(obs_enh), each=Ngenes),\n",
    "    expCotrans = as.vector(GxE),\n",
    "    obsCotrans = as.vector(obsGE),\n",
    "    chisq      = as.vector(obs_chisq),\n",
    "    emp_p_pos  = as.numeric(emp_p_pos) / Niters, # correlated cotranscription\n",
    "    emp_p_neg  = as.numeric(emp_p_neg) / Niters  # anticorrelated cotranscription\n",
    "    # emp_p      = as.numeric(emp_p) / Niters  # anticorrelated cotranscription\n",
    ");\n",
    "chr_corr$Gene = as.factor(names(genes)[chr_corr$Gene]);\n",
    "chr_corr$Enhancer = as.factor(names(enhancers)[chr_corr$Enhancer]);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0dfe1919-4297-432c-b4b7-ff0f356de0be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fwrite(chr_corr, file=\"../data/scGROv2p8_Gene10kb_Enh_v1genes_1Kpermuted_chisq_empp.csv.gz\");\n",
    "# fwrite(chr_corr, file=\"../data/scGROv2p8_Gene10kb_Enh_genes_10Kpermuted_chisq_empp.csv.gz\");\n",
    "fwrite(chr_corr, file=\"../data/scGROv2p8_ExG_1Kpermuted_chisq_empp.csv.gz\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd19e89d-3b8f-4f47-ab0a-9ccd6fa9e6ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "dim(chr_corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "479a12ac-39ca-4cb4-b803-675e27b19bfa",
   "metadata": {},
   "outputs": [],
   "source": [
    "test = chr_corr %>%\n",
    "    filter(emp_p_pos != \"NA\")\n",
    "dim(test)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d078e84a-05a2-479d-bf08-b32dac05c73d",
   "metadata": {},
   "source": [
    "## Read file that has been written for analyses:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8472d18-3cd0-413e-8d62-e62b3f29f2d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr_corrInp = fread(file=\"../data/scGROv2p8_Gene10kb_Enh_v1genes_1Kpermuted_chisq_empp.csv.gz\");\n",
    "# chr_corrInp = fread(file=\"../data/scGROv2p8_Gene10kb_Enh_genes_10Kpermuted_chisq_empp.csv.gz\");\n",
    "chr_corrInp = fread(file=\"../data/scGROv2p8_ExG_1Kpermuted_chisq_empp.csv.gz\");\n",
    "dim(chr_corrInp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19a1acd4-1b90-44bf-b2dd-6c930a1eafda",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_corr = chr_corrInp %>%\n",
    "    filter( as.character(seqnames(expressedFeatures[Gene])) == as.character(seqnames(expressedFeatures[Enhancer] ))) %>%\n",
    "    filter( emp_p_pos <= 0.2 | emp_p_neg <= 0.1 );"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cc9b242-9e67-4caa-a1a8-bcaed52ead3c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# lookup gene attributes\n",
    "feat1 = expressedFeatures[ chr_corr$Gene ];\n",
    "feat2 = expressedFeatures[ chr_corr$Enhancer ];\n",
    "chr_corr$geneLength = width(feat1);\n",
    "chr_corr$enhsLength = width(feat2);\n",
    "chr_corr$geneStrand = as.character(strand(feat1));\n",
    "chr_corr$enhsStrand = as.character(strand(feat2));"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52026734-d8a9-4d87-a488-38076b59f8cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# compute distance between gene promoters\n",
    "pro1 = resize(feat1, width=1, fix=\"start\");\n",
    "pro2 = resize(feat2, width=1, fix=\"start\");\n",
    "strand(pro1) = \"*\";\n",
    "strand(pro2) = \"*\";\n",
    "chr_corr$distance = width(pgap( pro1, pro2 ));\n",
    "\n",
    "chr_corr = chr_corr %>% filter(distance < 2e6);\n",
    "chr_corr$fdr_pos = p.adjust(chr_corr$emp_p_pos, method = \"fdr\");\n",
    "nrow(chr_corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "075ef478-7ecd-4fff-a92a-5f40f5350e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "head(chr_corr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c5b88f5-7f41-49b3-9fa8-d80d56bafa03",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_corr %>%\n",
    "    ggplot( aes(x=distance/1000)) +\n",
    "    geom_histogram(binwidth=50) +\n",
    "    xlim(0, 2000) +\n",
    "    ggtitle(\"Correlated gene-enhancer pairs\") +\n",
    "    xlab(\"Distance between pair (kb)\") +\n",
    "    ylab(\"Number of feature pairs\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e900914a-7d2c-4957-bdea-8caf5e654a95",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_corr %>%\n",
    "    filter( distance <= 200000 & emp_p_pos <= 0.05 ) %>%\n",
    "    ggplot( aes(x=distance, y=-log10(emp_p_pos)) ) +\n",
    "    geom_point(size=0.5) +\n",
    "    xlim(0, 100000) +\n",
    "    scale_color_viridis() +\n",
    "    geom_smooth(method = \"lm\", color=\"red\", alpha=0.5, formula = y ~ x) +\n",
    "    ggtitle(\"Gene-enhancer pairs\") +\n",
    "    xlab(\"Distance between pair (bp)\") +\n",
    "    ylab(\"-log10 P\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d6cbc59-e172-43dc-81a0-3ec412831492",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_corr %>%\n",
    "    filter( distance <= 200000 & emp_p_neg <= 0.05 ) %>%\n",
    "    ggplot( aes(x=distance, y=-log10(emp_p_neg)) ) +\n",
    "    geom_smooth(method = \"lm\", color=\"red\", alpha=0.5, formula = y ~ x) +\n",
    "    geom_point(size=0.5) +\n",
    "    xlim(0, 100000) +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(\"Gene-enhancer pairs Negatively co-transcription\") +\n",
    "    xlab(\"Distance between pair (bp)\") +\n",
    "    ylab(\"-log10 P\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36b7e7bb-f65a-4842-a9ee-f4749c58bf65",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_corr %>%\n",
    "    filter( emp_p_pos <= 0.05 ) %>%\n",
    "    ggplot( aes(x=distance/1000, y=-log10(emp_p_pos)) ) +\n",
    "    geom_pointdensity(size=0.5) +\n",
    "    geom_smooth(method = \"lm\", color=\"red\", alpha=0.5, formula = y ~ x) +\n",
    "    xlim(0, 1000) +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(\"Feature pairs\") +\n",
    "    xlab(\"Distance between pair (kb)\") +\n",
    "    ylab(\"-log10 P\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "715c0ca9-5f56-4d7e-9ec5-61afb77a5b59",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_corr %>%\n",
    "    filter( emp_p_neg <= 0.05 ) %>%\n",
    "    ggplot( aes(x=distance/1000, y=-log10(emp_p_neg)) ) +\n",
    "    geom_pointdensity(size=0.5) +\n",
    "    geom_smooth(method = \"lm\", color=\"red\", alpha=0.5, formula = y ~ x) +\n",
    "    xlim(0, 1000) +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(\"Feature pairs Negatively co-transcribed\") +\n",
    "    xlab(\"Distance between pair (kb)\") +\n",
    "    ylab(\"-log10 P\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20e0a08f-9456-449d-985a-b9ed066e2b87",
   "metadata": {},
   "outputs": [],
   "source": [
    "query_genes = c(\"GN-Nanog\",\"GN-Sox2\",\"GN-Sox2long\", \"GN-Pou5f1\", \"GN-Klf4\");\n",
    "query_enh = c(\n",
    "    \"Sox2_200kbUp\",\n",
    "    \"Sox2_2kbDn\",\n",
    "    \"Sox2_SRR1_4kbUp\", \n",
    "    \"Sox2_SRR2_2kbDn\", \n",
    "    \"Sox2_18kbDn\", \n",
    "    \"Sox2_85kbDn\", \n",
    "    \"Sox2_95kbDn\",\n",
    "    \"Sox2_106kbDn\",\n",
    "    \"Sox2_107kbDn\",\n",
    "    \"Sox2_109kbDn\",\n",
    "    \"Sox2_111kbDn\",\n",
    "    \"Sox2_105_109kbDn\",\n",
    "    \"Sox2_109_112kbDn\",\n",
    "    \"Sox2_SCR_Dn\",\n",
    "    \"Sox2_105kbDn\",\n",
    "    \"Sox2_1MbDn\",\n",
    "    \"Klf4_E1_67kbUp\",\n",
    "    \"Klf4_E2_58kbUp\", \n",
    "    \"Klf4_E3_56kbUp\",\n",
    "    \"Klf4_E123_55kbUp\",\n",
    "    \"Klf4_50kbUp\",\n",
    "    \"Nanog_5kbUp\",\n",
    "    \"Nanog_45kbUp\",\n",
    "    \"Nanog_60kbDn\",\n",
    "    \"Pou5f1_25kbUp\",\n",
    "    \"Pou5f1_20kbUp\",\n",
    "    \"Pou5f1_3kbUp\",\n",
    "    \"Pou5f1_39kbDn\",\n",
    "    \"Pou5f1_80kbDn\"\n",
    ");\n",
    "\n",
    "query_enh = c(\n",
    "    query_enh,\n",
    "    paste0( query_enh, \"_pl\" ),\n",
    "    paste0( query_enh, \"_mn\" )\n",
    ");\n",
    "query_enh = query_enh[ query_enh %in% names(features) ];\n",
    "\n",
    "query = c(query_genes, query_enh);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54fbbe90-2c3e-4649-ab61-20169046cc04",
   "metadata": {},
   "outputs": [],
   "source": [
    "OSN_corr = chr_corrInp %>%\n",
    "    filter( as.character(seqnames(expressedFeatures[Gene])) == as.character(seqnames(expressedFeatures[Enhancer] ))) %>%\n",
    "    filter( Gene %in% query & Enhancer %in% query );\n",
    "dim(OSN_corr)\n",
    "OSN_corr$fdr_pos = p.adjust(OSN_corr$emp_p_pos, method = \"fdr\");\n",
    "\n",
    "\n",
    "\n",
    "OSN_corr %>%\n",
    "    filter( emp_p_pos <= 0.2 | emp_p_neg <= 0.1 );\n",
    "    # filter( fdr_pos >= 0.05 );\n",
    "    # filter( Gene %in% query & Enhancer %in% query );# %>%\n",
    "    # filter( Gene %in% query_genes                 );# %>%\n",
    "    # filter(                 Enhancer %in% query ) %>%\n",
    "    # filter( emp_p_pos <= 0.5 & distance < 1000000 );"
   ]
  },
  {
   "cell_type": "raw",
   "id": "0584d074-41d8-4e8c-9d5d-27d693d1e1c4",
   "metadata": {},
   "source": [
    "# 1K permuted:\n",
    "A data.table: 15 × 14 \n",
    "Gene\tEnhancer\tpGene\tpEnh\texpCotrans\tobsCotrans\tchisq\temp_p_pos\temp_p_neg\tgeneLength\tenhsLength\tgeneStrand\tenhsStrand\tdistance\n",
    "<chr>\t<chr>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<dbl>\t<int>\t<int>\t<chr>\t<chr>\t<int>\n",
    "GN-Sox2 \tSox2_2kbDn \t0.05555556\t0.010582011\t0.0005878895\t0.0026455026\t7.201646e-03\t0.005\t0.995\t2469\t1139\t+\t*\t3131\n",
    "GN-Sox2long\tSox2_2kbDn \t0.09183673\t0.010582011\t0.0009718173\t0.0071806500\t3.966754e-02\t0.000\t1.000\t10000\t1139\t+\t*\t3131\n",
    "GN-Sox2 \tSox2_SRR2_2kbDn \t0.05555556\t0.007558579\t0.0004199211\t0.0018896447\t5.144033e-03\t0.009\t0.991\t2469\t949\t+\t*\t3600\n",
    "GN-Sox2long\tSox2_SRR2_2kbDn \t0.09183673\t0.007558579\t0.0006941552\t0.0064247921\t4.730959e-02\t0.000\t1.000\t10000\t949\t+\t*\t3600\n",
    "GN-Sox2 \tSox2_105kbDn \t0.05555556\t0.067271353\t0.0037372974\t0.0034013605\t-3.019657e-05\t0.951\t0.049\t2469\t32990\t+\t*\t84867\n",
    "GN-Sox2long\tSox2_109_112kbDn\t0.09183673\t0.009448224\t0.0008676940\t0.0018896447\t1.203631e-03\t0.189\t0.811\t10000\t3023\t+\t*\t109387\n",
    "GN-Sox2 \tSox2_109kbDn \t0.05555556\t0.005291005\t0.0002939447\t0.0011337868\t2.399549e-03\t0.064\t0.936\t2469\t1438\t+\t*\t109604\n",
    "GN-Sox2long\tSox2_109kbDn \t0.09183673\t0.005291005\t0.0004859086\t0.0015117158\t2.165593e-03\t0.064\t0.936\t10000\t1438\t+\t*\t109604\n",
    "GN-Sox2long\tSox2_1MbDn \t0.09183673\t0.036281179\t0.0033319450\t0.0071806500\t4.445611e-03\t0.043\t0.957\t10000\t22236\t+\t*\t1062870\n",
    "GN-Sox2long\tSox2_1MbDn \t0.09183673\t0.036281179\t0.0033319450\t0.0071806500\t4.445611e-03\t0.043\t0.957\t10000\t22236\t+\t*\t1062870\n",
    "GN-Klf4 \tKlf4_E1_67kbUp \t0.03212396\t0.004913076\t0.0001578275\t0.0007558579\t2.266021e-03\t0.069\t0.932\t10000\t560\t-\t*\t68191\n",
    "GN-Nanog \tNanog_5kbUp \t0.05933485\t0.007558579\t0.0004484871\t0.0011337868\t1.047155e-03\t0.154\t0.846\t10000\t2850\t+\t*\t5788\n",
    "GN-Pou5f1 \tPou5f1_20kbUp \t0.10430839\t0.008692366\t0.0009066867\t0.0026455026\t3.334648e-03\t0.034\t0.966\t10000\t2887\t+\t*\t16640\n",
    "GN-Pou5f1 \tPou5f1_3kbUp \t0.10430839\t0.008692366\t0.0009066867\t0.0022675737\t2.042617e-03\t0.077\t0.923\t10000\t3627\t+\t*\t3919\n",
    "GN-Pou5f1 \tPou5f1_80kbDn \t0.10430839\t0.004157218\t0.0004336328\t0.0000000000\t-4.336328e-04\t0.951\t0.049\t10000\t2921\t+\t*\t87300"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fa5fa35-8bf2-40a5-986e-837a77ac0503",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEmm10 = read_bed(\"~/group/genes_enhancer_list/mm10_mESC_OSN_SE_nonOverlappingWithGenes_v2.bed\")\n",
    "length(SEmm10)\n",
    "SEmm10\n",
    "\n",
    "# get all features within 1MB of SEs\n",
    "SEneighbors = subsetByOverlaps(genes, SEmm10, maxgap=1000000);\n",
    "hits = distanceToNearest( SEneighbors, SEmm10, ignore.strand=T );\n",
    "#hits\n",
    "SEneighbors$SE = SEmm10$name[hits@to];\n",
    "SEneighbors$distance = hits@elementMetadata$distance;\n",
    "SEneighbors\n",
    "write.csv(SEneighbors, file=\"../data/SEs_neighbor_genes.csv\")\n",
    "\n",
    "SEs = SEneighbors$SE\n",
    "SEgenes = names(SEneighbors)\n",
    "query_SE = c(SEs, SEgenes);\n",
    "length(unique(SEgenes))\n",
    "length(unique(SEs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12475965-9982-44f6-8b96-53f7ef01984e",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEcorr = chr_corr %>%\n",
    "    filter( Gene %in% query_SE & Enhancer %in% query_SE ) %>%\n",
    "    filter( emp_p_pos < 0.05 & distance <= 100000 ) %>%\n",
    "    filter(pGene >= 0.025 & pEnh >= 0.025 )#; %>%\n",
    "    # filter( substr(Gene, 0, 5) != \"GN-Gm\");\n",
    "\n",
    "length(unique(SEcorr$Gene))\n",
    "length(unique(SEcorr$Enhancer))\n",
    "# % of SEs that have a protein coding gene within the distance limit:\n",
    "length(unique(SEcorr$Enhancer))/length(unique(SEs))*100\n",
    "\n",
    "SEcorr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2c28c36a-93ab-49e6-a950-62e9894be365",
   "metadata": {},
   "outputs": [],
   "source": [
    "## SE assigned genes by the original paper:\n",
    "SEmm9 = read.csv(\"~/group/mES_PROseq_RNAseq_ATACseq/SE_Whyte_Young_2013/SE_proximalGenes_mm9.csv\")\n",
    "SEmm9 = GRanges(SEmm9) %>%\n",
    "    filter( isSuper == \"YES\");\n",
    "SEmm9\n",
    "length(unique(SEmm9$proximal_gene))\n",
    "    \n",
    "glist = read.table(\"~/group/genes_enhancer_list/mm10RefFlat_glist.txt\");\n",
    "glist = glist[2:nrow(glist),]\n",
    "colnames(glist) = c(\"name\", \"NMname\", \"chr\", \"strand\", \"start\", \"end\");\n",
    "# write.table(glist[,c(3,5,6,1,2,4)], file=\"~/group/genes_enhancer_list/mm10RefFlat_glist.bed\", row.names=F, col.names=F, sep='\\t', quote=F)\n",
    "\n",
    "SEglist = GRanges(glist) %>%\n",
    "    filter( NMname %in% SEmm9$proximal_gene ) %>%\n",
    "    mutate( name = paste0(\"GN-\", name) ); \n",
    "\n",
    "SEs = SEmm9$ID;\n",
    "SEgenes = SEglist$name;\n",
    "query_SE = c(SEs, SEgenes);\n",
    "length(unique(SEgenes))\n",
    "length(unique(SEs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2805bc39-8038-4235-8ae7-e8fb130dcbb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEcorr = chr_corrInp %>%\n",
    "    filter( as.character(seqnames(expressedFeatures[Gene])) == as.character(seqnames(expressedFeatures[Enhancer] ))) %>%\n",
    "    filter( Gene %in% query_SE & Enhancer %in% query_SE ) %>%\n",
    "    filter(pGene >= 0.01 & pEnh >= 0.01 );\n",
    "SEcorr$fdr_pos = p.adjust(SEcorr$emp_p_pos, method = \"fdr\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a8bbf67f-f032-42c2-83ca-086ec760e4b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "SEcorr = SEcorr %>%\n",
    "    filter( Gene %in% query_SE & Enhancer %in% query_SE ) %>%\n",
    "    filter( emp_p_pos < 0.05 ) %>%\n",
    "    filter(pGene >= 0.01 & pEnh >= 0.01 )#; %>%\n",
    "    # filter( substr(Gene, 0, 5) != \"GN-Gm\");\n",
    "\n",
    "length(unique(SEcorr$Gene))\n",
    "length(unique(SEcorr$Enhancer))\n",
    "# % of SEs that have a protein coding gene within the distance limit:\n",
    "length(unique(SEcorr$Enhancer))/length(unique(SEs))*100\n",
    "\n",
    "SEcorr"
   ]
  },
  {
   "cell_type": "raw",
   "id": "af511657-b8d3-48e1-ac15-7957509decba",
   "metadata": {},
   "source": [
    "## GENES nor overlapped by their SE:\n",
    "Cbfa2t2\n",
    "Zfp704\n",
    "Klf3\n",
    "Ddit4\n",
    "Klf9\n",
    "Dmrt1\n",
    "\n",
    "Can be modified to trim overlap:\n",
    "Id1\n",
    "Sox2\n",
    "Phc1\n",
    "Nanog\n",
    "Zfp710\n",
    "Fgf4\n",
    "Vdac1\n",
    "Mycn\n",
    "Nfkbia\n",
    "Itpk1\n",
    "Jam2\n",
    "Spry4\t"
   ]
  },
  {
   "cell_type": "raw",
   "id": "f9724ea7-fafd-4b6d-9708-f8606503a331",
   "metadata": {},
   "source": [
    "Gene_3kb with v1 features:\n",
    "\n",
    "\n",
    "\n",
    "Gene_3kb with v4 features (With groHMM):\n",
    "Gene\tEnhancer\tCotranscription\temp_p\tlength1\tlength2\tstr1\tstr2\tdistance\n",
    "<chr>\t<chr>\t<dbl>\t<dbl>\t<int>\t<int>\t<chr>\t<chr>\t<int>\n",
    "Sox2\tSox2_2kbDn_pl\t0.001892506\t0.015\t3000\t1139\t+\t+\t3129\n",
    "Sox2\tSox2_1MbDn_pl\t0.002271007\t0.058\t3000\t22236\t+\t+\t1062868\n",
    "Sox2\tSox2_2kbDn\t0.002649508\t0.007\t3000\t1139\t+\t*\t3129\n",
    "Nanog\tNanog_60kbDn_mn\t0.003406510\t0.057\t3000\t19801\t+\t-\t73433\n",
    "Nanog\tNanog_60kbDn\t0.004542014\t0.061\t3000\t19801\t+\t*\t53633\n",
    "Pou5f1\tPou5f1_3kbUp_pl\t0.001514005\t0.001\t3000\t3471\t+\t+\t3764\n",
    "Pou5f1\tPou5f1_25kbUp\t0.003406510\t0.013\t3000\t16201\t+\t*\t22231\n",
    "Pou5f1\tPou5f1_3kbUp\t0.001514005\t0.017\t3000\t3627\t+\t*\t3920\n",
    "\n",
    "\n",
    "Gene_10kb with v1 features (Without groHMM):\n",
    "Gene\tEnhancer\tCotranscription\temp_p\tlength1\tlength2\tstr1\tstr2\tdistance\n",
    "<chr>\t<chr>\t<dbl>\t<dbl>\t<int>\t<int>\t<chr>\t<chr>\t<int>\n",
    "Sox2\tSox2_2kbDn_pl\t0.006434519\t0.000\t10000\t1139\t+\t+\t3129\n",
    "Sox2\tSox2_1MbDn_pl\t0.003028009\t0.015\t10000\t22236\t+\t+\t1062868\n",
    "Sox2\tSox2_2kbDn\t0.007191522\t0.000\t10000\t1139\t+\t*\t3129\n",
    "Sox2\tSox2_1MbDn\t0.005677517\t0.036\t10000\t22236\t+\t*\t1062868\n",
    "Nanog\tNanog_60kbDn_pl\t0.003785011\t0.063\t10000\t19801\t+\t+\t53710\n",
    "Nanog\tNanog_60kbDn_mn\t0.006434519\t0.051\t10000\t19801\t+\t-\t73510\n",
    "Nanog\tNanog_60kbDn\t0.010219531\t0.006\t10000\t19801\t+\t*\t53710\n",
    "Pou5f1\tPou5f1_3kbUp_pl\t0.001892506\t0.002\t10000\t3471\t+\t+\t3764\n",
    "Pou5f1\tPou5f1_80kbDn_pl\t0.001892506\t0.069\t10000\t12925\t+\t+\t84067\n",
    "Pou5f1\tPou5f1_39kbDn_mn\t0.001514005\t0.089\t10000\t11451\t+\t-\t52967\n",
    "Pou5f1\tPou5f1_25kbUp\t0.004542014\t0.064\t10000\t16201\t+\t*\t22231\n",
    "Pou5f1\tPou5f1_3kbUp\t0.002271007\t0.012\t10000\t3627\t+\t*\t3920\n",
    "\n",
    "Gene_10kb with v4 features (With groHMM):\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
